{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Purpose of this notebook is to get full text articles from PMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import numpy as np \n",
    "import pandas as pd \n",
    "import matplotlib.pyplot as plt \n",
    "import os, json\n",
    "import requests\n",
    "import time\n",
    "\n",
    "import sys\n",
    "sys.path.append('../pubmed_rag')\n",
    "\n",
    "from bioc import get_biocjson, passages_to_df\n",
    "\n",
    "# import nltk\n",
    "# from nltk.stem import WordNetLemmatizer\n",
    "# nltk.download('wordnet')\n",
    "\n",
    "#from spellchecker import SpellChecker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/test_pmid_list.csv', header=None)\n",
    "df.head()\n",
    "\n",
    "pmids = df[0].astype(str).to_list()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Current status \n",
    "\n",
    "Getting full texts via biocjson from [Pubtator](https://www.ncbi.nlm.nih.gov/research/pubtator3/api) instead of [BioC API for PMC](https://www.ncbi.nlm.nih.gov/research/bionlp/APIs/BioC-PMC/)\n",
    "\n",
    "Why?? \n",
    "- I only have pmids currently, and Pubtator automatically links pmid to pmcid to return full text if it is available, otherwise returns abstract.\n",
    "- BioC API for PMC is not compatible with pmid id even though it says it does\n",
    "\n",
    "Bioc files\n",
    "- it looks like each 'passage' is a sentence\n",
    "- and the passages have annotations \n",
    "- additionally, relevant to us, the passage has metadata such as what section its from\n",
    "\n",
    "\n",
    "Steps\n",
    "- config\n",
    "    - provide list of pmids \n",
    "    - output path\n",
    "- retrieve biocjson (full text or abstract) and save to output path\n",
    "- for each json parse for text + section\n",
    "\n",
    "Vertex database of sentences or sections?\n",
    "Option 1:\n",
    "- embedd each sentence using [PubMedBERT now called BiomedBERT](https://huggingface.co/microsoft/BiomedNLP-BiomedBERT-base-uncased-abstract-fulltext)\n",
    "- keeping metadata\n",
    "\n",
    "Option 2:\n",
    "- collapse the sentences into one section \n",
    "- embed each section using same model\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example usage\n",
    "sample = pmids[:15]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "for each in sample:\n",
    "    \n",
    "    result = get_biocjson(each, 'biocjson')\n",
    "\n",
    "    if result is not None:\n",
    "        passages_to_df(result, 'biocjson')\n",
    "\n",
    "    time.sleep(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>section</th>\n",
       "      <th>sentence</th>\n",
       "      <th>pmid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>date</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>title</td>\n",
       "      <td>Heterogeneous Network Edge Prediction: A Data ...</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>abstract</td>\n",
       "      <td>The first decade of Genome Wide Association St...</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>abstract</td>\n",
       "      <td>Author Summary.</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>abstract</td>\n",
       "      <td>For complex human diseases, identifying the ge...</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>intro</td>\n",
       "      <td>Introduction.</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   index   section                                           sentence  \\\n",
       "0      0     title  Heterogeneous Network Edge Prediction: A Data ...   \n",
       "1      1  abstract  The first decade of Genome Wide Association St...   \n",
       "2      2  abstract                                    Author Summary.   \n",
       "3      3  abstract  For complex human diseases, identifying the ge...   \n",
       "4      4     intro                                      Introduction.   \n",
       "\n",
       "       pmid       pmcid                      date  \\\n",
       "0  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "1  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "2  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "3  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "4  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "\n",
       "                           authors           journal  \n",
       "0  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "1  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "2  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "3  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "4  Himmelstein DS and Baranzini SE  PLoS Comput Biol  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cleaning?\n",
    "# read in example\n",
    "df_test = pd.read_csv('biocjson/df_26158728.csv')\n",
    "# lower case section names\n",
    "df_test['section'] = df_test['section'].str.lower().str.strip()\n",
    "# pmids to object\n",
    "df_test['pmid'] = df_test['pmid'].astype(str)\n",
    "df_test['date'] = pd.to_datetime(df_test['date'])\n",
    "# also stripping sentences in case?\n",
    "df_test['sentence'] = df_test['sentence'].str.strip()\n",
    "\n",
    "punctuations = ('!',',','.','?',',','\"', \"'\")\n",
    "# lol adding a . to the end for now?\n",
    "df_test['sentence'] = np.where(df_test['sentence'].str.endswith(punctuations), df_test['sentence'], df_test['sentence']+'.')\n",
    "\n",
    "# check\n",
    "df_test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "# which sections to keep? \n",
    "keep_sections = ['title', 'abstract', 'intro', 'results', 'discuss', 'methods']\n",
    "\n",
    "# filter \n",
    "df_filtered = df_test[df_test['section'].isin(keep_sections)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>section</th>\n",
       "      <th>sentence</th>\n",
       "      <th>pmid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>date</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>intro</td>\n",
       "      <td>Introduction</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>11</td>\n",
       "      <td>results</td>\n",
       "      <td>Results</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>60</td>\n",
       "      <td>discuss</td>\n",
       "      <td>Discussion</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>66</td>\n",
       "      <td>methods</td>\n",
       "      <td>Methods</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>73</td>\n",
       "      <td>methods</td>\n",
       "      <td>Nodes</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>75</td>\n",
       "      <td>methods</td>\n",
       "      <td>Associations</td>\n",
       "      <td>26158728</td>\n",
       "      <td>PMC4497619</td>\n",
       "      <td>2015-07-09 00:00:00+00:00</td>\n",
       "      <td>Himmelstein DS and Baranzini SE</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    index  section      sentence      pmid       pmcid  \\\n",
       "4       4    intro  Introduction  26158728  PMC4497619   \n",
       "11     11  results       Results  26158728  PMC4497619   \n",
       "60     60  discuss    Discussion  26158728  PMC4497619   \n",
       "66     66  methods       Methods  26158728  PMC4497619   \n",
       "73     73  methods         Nodes  26158728  PMC4497619   \n",
       "75     75  methods  Associations  26158728  PMC4497619   \n",
       "\n",
       "                        date                          authors  \\\n",
       "4  2015-07-09 00:00:00+00:00  Himmelstein DS and Baranzini SE   \n",
       "11 2015-07-09 00:00:00+00:00  Himmelstein DS and Baranzini SE   \n",
       "60 2015-07-09 00:00:00+00:00  Himmelstein DS and Baranzini SE   \n",
       "66 2015-07-09 00:00:00+00:00  Himmelstein DS and Baranzini SE   \n",
       "73 2015-07-09 00:00:00+00:00  Himmelstein DS and Baranzini SE   \n",
       "75 2015-07-09 00:00:00+00:00  Himmelstein DS and Baranzini SE   \n",
       "\n",
       "             journal  \n",
       "4   PLoS Comput Biol  \n",
       "11  PLoS Comput Biol  \n",
       "60  PLoS Comput Biol  \n",
       "66  PLoS Comput Biol  \n",
       "73  PLoS Comput Biol  \n",
       "75  PLoS Comput Biol  "
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_filtered[[len(x)==1 for x in df_filtered['sentence'].str.split(' ', )]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Heterogeneous Network Edge Prediction: A Data Integration Approach to Prioritize Disease-Associated Genes.\n",
      "The first decade of Genome Wide Association Studies (GWAS) has uncovered a wealth of disease-associated variants. Two important derivations will be the translation of this information into a multiscale understanding of pathogenic variants and leveraging existing data to increase the power of existing and future studies through prioritization. We explore edge prediction on heterogeneous networks:graphs with multiple node and edge types:for accomplishing both tasks. First we constructed a network with 18 node types:genes, diseases, tissues, pathophysiologies, and 14 MSigDB (molecular signatures database) collections:and 19 edge types from high-throughput publicly-available resources. From this network composed of 40,343 nodes and 1,608,168 edges, we extracted features that describe the topology between specific genes and diseases. Next, we trained a model from GWAS associations and predicted the probability of association between each protein-coding gene and each of 29 well-studied complex diseases. The model, which achieved 132-fold enrichment in precision at 10% recall, outperformed any individual domain, highlighting the benefit of integrative approaches. We identified pleiotropy, transcriptional signatures of perturbations, pathways, and protein interactions as influential mechanisms explaining pathogenesis. Our method successfully predicted the results (with AUROC = 0.79) from a withheld multiple sclerosis (MS) GWAS despite starting with only 13 previously associated genes. Finally, we combined our network predictions with statistical evidence of association to propose four novel MS genes, three of which (JAK2, REL, RUNX3) validated on the masked GWAS. Furthermore, our predictions provide biological support highlighting REL as the causal gene within its gene-rich locus. Users can browse all predictions online (http://het.io). Heterogeneous network edge prediction effectively prioritized genetic associations and provides a powerful new approach for data integration across multiple domains. Author Summary. For complex human diseases, identifying the genes harboring susceptibility variants has taken on medical importance. Disease-associated genes provide clues for elucidating disease etiology, predicting disease risk, and highlighting therapeutic targets. Here, we develop a method to predict whether a given gene and disease are associated. To capture the multitude of biological entities underlying pathogenesis, we constructed a heterogeneous network, containing multiple node and edge types. We built on a technique developed for social network analysis, which embraces disparate sources of data to make predictions from heterogeneous networks. Using the compendium of associations from genome-wide studies, we learned the influential mechanisms underlying pathogenesis. Our findings provide a novel perspective about the existence of pervasive pleiotropy across complex diseases. Furthermore, we suggest transcriptional signatures of perturbations are an underutilized resource amongst prioritization approaches. For multiple sclerosis, we demonstrated our ability to prioritize future studies and discover novel susceptibility genes. Researchers can use these predictions to increase the statistical power of their studies, to suggest the causal genes from a set of candidates, or to generate evidence-based experimental hypothesis.\n",
      "Introduction. In the last decade, genome-wide association studies (GWAS) have been established as the main strategy to map genetic susceptibility in dozens of complex diseases and phenotypes. Despite the success of this approach in mapping variation in thousands of loci to hundreds of complex phenotypes, researchers are now confronted with the challenge of maximizing the scientific contribution of existing GWAS datasets, whose undertakings represented a substantial investment of human and monetary resources from the community at large. A central assumption in GWAS is that every region in the genome (and hence every gene) is a-priori equally likely to be associated with the phenotype in question. As a result, small effect sizes and multiple comparisons limit the pace of discovery. However, rational prioritization approaches may afford an increase in study power while avoiding the constraints and expense related to expanded sampling. One such way forward is the current trend of analyzing the combined contribution of susceptibility variants in the context of biological pathways, rather than single SNPs. For example, Yaspan et al described an approach that aggregates variants of interest from a GWAS into biological pathways using genomic randomization to control for multiple testing and minimize type I error. The popular software PLINK also includes an option to evaluate groups of associations at the gene level, thus enabling pathway analysis by computing enriched gene sets. A less explored but potentially revealing strategy is the integration of diverse sources of data to build more accurate and comprehensive models of disease susceptibility. Several strategies have been attempted to identify the mechanisms underlying pathogenesis and use these insights to prioritize genes for genetic association analyses. Gene-set enrichment analyses identify prevalent biological functions amongst genes contained in disease-associated loci. Gene network approaches search for neighborhoods of genes where disease-associated loci aggregate. Jia et al. reported dmGWAS, a strategy to integrate association signals from GWAS into the human protein interaction network. A similar approach was developed by our group and tested in two large studies comprising more than 15,000 cases. Literature mining techniques aim to chronicle the relatedness of genes to identify a subset of highly-related associated genes. For example, Raychaudhuri et al. reported the Gene Relationships Among Implicated Loci (GRAIL) algorithm, an approach to assess relationships among genomic disease regions by text mining of PubMed abstracts. Prioritization strategies generally rely on user-provided loci as the sole input and do not incorporate broader disease-specific knowledge. Typically, the proportion of genome-wide significant discoveries in a given GWAS is low, thus leaving little high-confidence signal for seed-based approaches to build from. To overcome this limitation, here we aimed at characterizing the ability of various information domains to identify pathogenic variants across the entire compendium of complex disease associations. Using this multiscale approach, we developed a framework to prioritize both existing and future GWAS analyses and highlight candidate genes for further analysis. To approach this problem, we resorted to a method that integrated diverse information domains naturally. Heterogeneous (or multipartite) networks are a class of networks which contain multiple types of entities (nodes) and relationships (edges or links), and provide a data structure capable of expressing diversity in an intuitive and scalable fashion. Most existing techniques available for network analysis have been developed for homogeneous networks and are not directly extensible to heterogenous networks. Accordingly, the early network analyses in disease biology concentrated on homogeneous networks. However in the last half-decade, the complexity of biological systems has spurred interest in heterogeneous approaches. While still a developing field, network-based biological data integration has been pursued using a variety of techniques. Approaches such as GeneMANIA, weight and then project individual data sources onto a single dimension, enabling homogeneous network algorithms to be used to characterize the resulting graphs. Other techniques operate on multi-relational (single node type, multiple edge type) networks, for example by taking into account relationships among local clusters and considering the full topology of weighted gene association networks. Bipartite networks contain two node types and therefore work well for predicting relationships between entities of two different types (such as disease-gene associations or drug-disease indications) following a 'guilt-by-association' paradigm. Other approaches incorporate greater-dimension heterogeneous networks as input but conflate types and, while improving predictions compared to simpler approaches, cannot effectively identify influential network components. Heterogeneous networks of arbitrary complexity have also been applied for edge prediction without a formalized feature extraction methodology, which requires manual descriptor determination for each new network design. Recently, new types of edge prediction methods were reported that naturally accommodate any size heterogeneous network. These include data fusion by matrix-factorization and metapath-based techniques. This type of intermediate data fusion can treat all data sources directly (i.e. without transforming data into \"disease space\") and has been successfully used to infer disease similarities and predict gene function in slime mold and baker's yeast. A metapath-based approach was recently developed by researchers studying social sciences to predict future coauthorship and provides an intuitive framework and interpretable models and results. An advantage of metapath-based approaches is that they preserve the network structure and provide the flexibility to explore a diverse set of descriptors. In this work, we extended this methodology to predict the probability that an association between a gene and disease exists.\n",
      "Results. Constructing a heterogeneous network to integrate diverse information domains. Using publicly-available databases and standardized vocabularies, we constructed a heterogeneous network with 40,343 nodes and 1,608,168 edges (Fig 1, S2 Data). Databases were selected based on quality, reusability, throughput, and their aggregate ability to provide a diversified, multiscale portrayal of biology (see Resource selection in Methods). The network was designed to encode entities and relationships relevant to pathogenesis. The network contained 18 node types (metanodes) and 19 edge types (metaedges), displayed in Fig 2A. Entities represented by metanodes consisted of diseases, genes, tissues, pathophysiologies, and gene sets for 14 MSigDB collections including pathways, perturbation signatures, motifs, and Gene Ontology (GO) domains (Table 1). Relationships represented by metaedges consisted of gene-disease association, disease pathophysiology, disease localization, tissue-specific gene expression, protein interaction, and gene-set membership for each MSigDB collection (Table 2). Gene-disease associations were extracted from the GWAS Catalog by overlapping associations into disease-specific loci. Loci were classified as low or high-confidence based on p-value and sample size of the corresponding GWAS (see Associations in Methods and S1 Fig). When possible, for each loci, the most-commonly reported gene across studies was designated as primary and subsequently considered responsible for the association. Additional genes reported for the loci were considered secondary. Only high-confidence primary associations were included in the network yielding 938 associations between 99 diseases and 711 genes (S2 Fig visualizes a subset of these associations). Features quantify the network topology between a gene and disease. To describe the network topology connecting a specific gene and disease, we computed 24 features, each describing a different aspect of connectivity. Each feature corresponds to a type of path (metapath) originating in a given source gene and terminating in a given target disease. The biological interpretation of a feature derives from its metapath (S1 Table), and features simply quantify the prevalence of a specific metapath between any gene-disease pair. To quantify metapath prevalence, we adapted an existing method originally developed for social network analysis (PathPredict), and developed a new metric called degree-weighted path count (DWPC, Fig 2D), which we employed in all but two features. The DWPC downweights paths through high-degree nodes when computing metapath prevalence. The strength of downweighting depends on a single parameter (w), which we optimized to w = 0.4 and that outperformed the top metric resulting from PathPredict (S3 Fig). We calculated DWPC features for the 22 metapaths of length 3 or less that originated with a gene and terminated with disease. Two non-DWPC features were included to assess the pleiotropy of the source gene and the polygenicity of the target disease. Referred to as 'path count' features, they respectively equal the number of diseases associated with the source gene and the number of genes associated with the target disease. For all features, paths with duplicate nodes were excluded, and, if present, the association edge between the source gene and target disease was masked. Machine learning approach to predict the probability of association of gene-disease pairs. Further analysis focused on the 29 diseases with at least ten associated genes (Table 3). The 698 high-confidence primary associations of these 29 diseases were considered positives:gene-disease pairs with positive experimental relationships (as defined in the Associations section of Methods, S2 Fig). The remaining 551,823 (i.e. unassociated) gene-disease pairs were considered negatives. Low-confidence or secondary associations were excluded from either set. We partitioned gene-disease pairs into training (75%) and testing (25%) sets and created a training network with the testing associations removed. To learn the importance of each feature and model the probability of association of a given gene-disease pair, we used regularized logistic regression which is designed to prevent overfitting and accurately estimate regression coefficients when models include many features. Elastic net regression is a regression method that balances two regularization techniques: ridge (which performs coefficient shrinkage) and lasso (which performs coefficient shrinkage and variable selection). On the training set, we optimized the elastic net mixing parameter, a single parameter behind the DWPC metric, and two edge-inclusion thresholds (S3 Fig). While cross-validated performance was similar across elastic net mixing parameters, ridge demonstrated the greatest consistency (S3 Fig), and thus we proceeded with logistic ridge regression as the primary model for predictions. Method prioritizes associations withheld for testing. We extracted network-based features for gene-disease pairs from the training network and modeled the training set. We next evaluated performance on the 25% of gene-disease pairs (175 positives, 137,956 negatives) withheld for testing. Our predictions achieved an area under the ROC curve (AUROC) of 0.83 (Fig 3A) demonstrating an excellent performance in retrieving hidden associations. Importantly, we did not observe any significant degradation of performance from training to testing (Fig 3A), indicating that our disciplined regularization approach avoided overfitting and that predictions for associations included in the network were not biased by their presence in the network. Furthermore, we observed that at 10% recall (the classification threshold where 10% of true positives were predicted as positives), our predictions achieved 16.7% precision (the proportion of predicted positives that were correct). Since the prevalence of positives in our dataset was 0.13%, the observed precision represents a 132-fold enrichment over the expected probability under a uniform distribution of priors (as in GWAS). Predicting associations on the complete network. As a next step in our analysis, we recomputed features on the complete network, which now included the previously withheld testing associations. On all positives and negatives, we fit a ridge model (the primary model for predictions) and a lasso model (for comparison). Standardized coefficients (Fig 4) indicate the effect attributed to each feature by the models. The lasso highlighted features that captured pleiotropy (4 features), pathways (2), transcriptional signatures of perturbations (1) and protein interactions (1). Despite the parsimony of the lasso, performance was similar between models with training AUROCs of 0.83 (ridge) and 0.82 (lasso). However, since multiple features from a correlated group may be causal, the lasso model risks oversimplifying. Ridge regression disperses an effect across a correlated group of features, providing users greater flexibility when interpreting predictions. From the ridge model, we predicted the probability that each protein-coding gene was associated with each analyzed disease (S1 Data) and built a webapp to display the predictions (http://het.io/disease-genes/browse). Degree-preserving network permutations highlight the importance of edge-specificity for top predictions and ten features. Using Markov chain randomized edge-swaps, we created 5 permuted networks. Since metaedge-specific node degree is preserved, features extracted from the permuted network retain unspecific effects. These effects include general measures a disease's polygenicity and a gene's pleiotropy, multifunctionality, and tissue-specificity. On the first permuted network, we partitioned associations into training and testing sets. Testing associations were masked from the network, features were computed, and a ridge model was fit on the training gene-disease pairs. Compared to the unpermuted-network model, testing performance was noticeably inferior: the AUROC declined from 0.83 (Fig 3A) to 0.79 and the AUPRC (area under the precision-recall curve) declined from 0.06 (Fig 3B) to 0.02 (S4 Fig). We interpret the modest decline in AUROC but marked reduction in AUPRC as a direct consequence of the permutation's particularly detrimental effect on top predictions (S4 Fig). In other words, edge-specificity was crucial for top predictions, while general effects gleaned from node degree performed reasonably well when ranking the entire spectrum of protein-coding genes for association. A commonly-overlooked finding is that the discriminatory ability of gene networks largely relies on node-degree rather than the edge-specificity. However, we found that for top predictions:which are the only predictions considered by many applications:edge-specificity was critical. Interestingly, predictions from the permuted-network model displayed a reduced dynamic range with none exceeding 4%, while predictions from the unpermuted-network model exceeded 75% (S4 Fig). Therefore, even though they achieve reasonable AUROC, the permuted-network predictions would have little utility as prior probabilities in a bayesian analysis where dynamic range is crucial. Furthermore, the signal present in permuted-network features was greatly diminished: few features survived the lasso's selection resulting in an average lasso AUROC of 0.70 versus 0.80 for ridge (S5 Fig). Permuting the network significantly reduced the predictiveness of features based on pleiotropy (2 features), protein interactions (2), transcriptional signatures of perturbations (1), tissue-specificity (1), pathways (3), and immunologic signatures (1) (S2 Table). Six of the eight features selected by the lasso and eight of the top ten ridge features (ranked by standardized coefficients) were negatively affected by the permutation. Since our modeling technique preferentially selected/weighted features affected by permutation, we can infer that network components where edge-specificity matters underlie a large portion of predictions. Feature importance identifies the mechanisms underlying associations. We assessed the informativeness of each feature by calculating feature-specific AUROCs. Feature-specific AUROCs universally exceeded 0.5, indicating that network connectivity, regardless of type, positively discriminates associations. However, performance varied widely by feature and within feature from disease to disease (Fig 5). Top performing domains consisted of transcriptional signatures of perturbations (AUROC = 0.74), immunologic signatures (0.70), and pleiotropy (0.68, 0.67, 0.64, 0.63). Notably, the models greatly outperformed any individual feature, highlighting the importance of an integrative approach. Features whose metapaths originate with an association (GaD) metaedge measure pleiotropy (S1 Table). The four pleiotropic features were among the top performing features that did not rely on set-based gene categorization (Fig 5). Of the four features, GaD (any disease) had the highest AUROC despite its lack of disease-specificity, reflecting both the sparsity of disease-specific features and the existence of genetic overlap between seemingly disparate diseases. GaDmPmD and GaDaGaD performed best for immunologic diseases and were affected by permutation, indicating that genetic overlap was greatest between immunologic diseases. On the other hand, the performance of GaDlTlD did not decrease after permutation indicating disease colocalization was not a primary driver of genetic overlap (S5 Fig and S2 Table). We also observed that the lasso regression model discarded the majority of features with a minimal performance deficit, suggesting redundancy among features. Indeed, pairwise feature correlations showed moderate collinearity among features (S6 Fig). Collinearity was especially pervasive with respect to the Perturbations feature, explaining its threefold increase in standardized coefficient in the lasso versus ridge model. The disappearance of all but one other MSigDB-based feature in the lasso model indicated that Perturbations:the feature traversing chemical and genetic transcriptional signatures of perturbations:exhausted meaningful gene-set characterization. In other words, the faulty molecular processes behind pathogenesis align with and are encapsulated by the processes perturbed by chemical and genetic modifications. The Immunologic signatures feature:traversing gene-sets characterizing \"cell types, states, and perturbations within the immune system\":was highly predictive and correlated with Perturbations. As expected this feature performed best for diseases with an immune pathophysiology. The one well-performing neoplastic disease (Fig 5) was chronic lymphocytic leukemia, a hematologic cancer with a strong immune component. Additionally, the performance of both the Perturbation and Immunologic features was affected by permutation indicating information beyond the extent of a gene's multifunctionality was encoded. Existing network-based gene-prioritization methods, frequently rely solely on protein-protein interactions. Our results supported incorporating protein interactions as the two interactome-based features were discriminatory (AUROCs = 0.65, 0.56) and affected by permutation. However, when compared to the integrative models or other top-performing features, performance of features that relied solely on the interactome was severely limited. Pathways, another founding resource for many approaches, proved important with KEGG selected by the lasso and all three pathway resources (AUROCs = 0.61 for KEGG, 0.60 for Reactome, 0.55 for BioCarta) affected by permutation. The GeTlD feature:measuring to what extent a gene is expressed in tissues affected by the disease in question:peaked in performance around AUROC = 0.58 (S3 Fig), was affected by permutation, and required no preexisting knowledge of associated genes. In other words, while approaches based on tissue-specificity may have limited predictive ability on their own, they are broadly applicable (i.e. less susceptible to knowledge bias) and provide orthogonal information that could enhance the overall performance of a model. Gene set robustness. For each type of gene set, we evaluated the effect of increased sparsity on performance by randomly subsampling gene set nodes or edges and measuring the resulting AUROC of the affected feature (S7 Fig). Robustness refers to a gene set collection's ability to withstand a high extent of masking with little performance deficit. Several of the top gene sets had this property, especially GO processes (where supersets are common), which may indicate nodal redundancy. Contrastingly, the MSigDB gene set with the fewest nodes, KEGG, experienced a more immediate and linear decline in performance. Since KEGG avoids duplication and is stringently and manually curated, this finding is expected. To investigate whether the high predictivity of certain gene set collections was due only to size, we compared performance when subsampling nodes to the KEGG level (crosses in S7 Fig). The two top performing collections, perturbations and immunologic signatures, which also happen to be large, continued to perform better than the majority of complete collections. While performance benefited from increasing densities, a resource's sparsity often reflects an intrinsic property of the underlying information type. Therefore, when identifying influential mechanisms of pathogenesis, we prefered unadjusted comparisons using the complete network. Case study: Prioritizing multiple sclerosis associations. The WTCCC2 multiple sclerosis (MS) GWAS tested 465,434 SNPs for 9,772 cases and 17,376 controls and identified over 50 independently associated loci. Since the GWAS Catalog excludes targeted arrays (such as ImmunoChip), this study remains the largest MS GWAS in the Catalog. To evaluate our method's ability to prioritize associations identified in a future study, we masked the WTCCC2 MS study from the GWAS Catalog and created a pre-WTCCC2 network. The number of high-confidence primary MS associations was thus reduced from 50 to 13, with the 37 novel genes identified by WTCCC2 available to evaluate performance. On the pre-WTCCC2 network, we extracted features, fit a ridge model, and predicted each gene's probability of association with MS. Amongst all 18,993 potentially novel genes, the 37 WTCCC2 genes were ranked highly (AUROC = 0.79, Fig 6). Prioritizing statistical candidates with network-based predictions identifies novel multiple sclerosis genes. Finally, we designed a framework for discovering and validating novel MS genes that incorporates our network-based predictions. Meta2.5 is a meta-analysis of all MS GWAS prior to the WTCCC2 study. We calculated genewise p-values for Meta2.5 using VEGAS and observed a large enrichment in nominally significant (p < 0.05) genes, suggesting multiple potential associations (S9 Fig). We combined this set of experimental candidates with the top predictions from the pre-WTCCC2 network to discover genes with both strong statistical and biological evidence of association (S12 Data). To ensure novelty, we excluded genes from GWAS-established MS loci and the extended MHC region. We chose a threshold (S3 Table) for network-based predictions that performed well in prioritizing the genes identified by WTCCC2 (Fig 6). This strategy discovered four genes, three of which:JAK2, REL, RUNX3:achieved Bonferroni validation on VEGAS-converted WTCCC2 p-values (Table 4). The probability of the observed validation rate occurring under random prioritization is 0.01 (S3 Table), demonstrating that incorporating our network-based predictions as a prior increased study power. JAK2 displays overexpression in MS-affected Th17 cells and was implicated in an interactome-based prioritization of GWAS. RUNX3, a transcription factor influencing T lymphocyte development, has been associated with celiac disease and ankylosing spondylitis and was hypermethylated in systemic lupus erythematosus patients. The region containing REL was uncovered in a recent MS ImmunoChip-based study with 14,498 cases. For the gene-dense region containing REL, the ImmunoChip study reported a long non-coding RNA, LINC01185, overlapping the lead-SNP, rs842639. However, since greater than 80% of the genome shows evidence of transcription, the probability of incidental overlap with long non-coding RNA is high. REL, however, is an essential transcription factor for lymphocyte development and plays a critical role in autoimmune inflammation. Hence, gene prioritization through integrative analyses offers not only to streamline loci discovery but also subsequent causal gene identification.\n",
      "Discussion. In this work, we developed a framework to predict the probability that each protein-coding gene is associated with each of 29 complex diseases. Our predictions draw on a diverse set of pathogenically-relevant relationships encoded in a heterogeneous network. The predictions successfully prioritized associations hidden from the network. Using MS as a representative example, we were able to combine our predictions with statistical evidence of association to increase study power and identify three novel susceptibility genes in this disease. The disease-specific performance (measured by the AUROC) for MS was exceeded by twelve other diseases suggesting that our predictions have broad applicability for prioritizing genetic association analyses. Prioritization can range from a genome-wide scale to a single loci where this approach can highlight the causal gene from several candidates within the same association block. For researchers focused on a specific disease, these predictions can be used to propose genes for experimental investigation. Inversely, researchers focused on a specific gene can use this resource to find suggestions for relevant complex disease phenotypes. Most previous explorations of the factors underlying pathogenicity have focused on a single domain such as tissue-specificity, protein interactions, pathways, or disease similarity. The method presented here integrates disparate data sources, learns their importance, and unifies them under a common framework enabling comparison. Therefore, we can conclude that perturbation gene sets:the core of our top-performing feature:are an underutilized resource for disease-associated gene prioritization. Not only did perturbations encompass other set-based gene categorizations, but they greatly outperformed features based on protein interactions, pathways, and tissue-specificity, which form the basis of several prominent prioritization techniques. In addition to characterizing the overall importance of each feature, our online prediction browser visually decomposes an individual prediction into its components. We observed a prominent influence of pleiotropy, consistent with previous studies that identified pervasive overlap of susceptibility loci across complex diseases, especially those of autoimmune nature. Since many existing prioritization techniques are agnostic to the compendia of GWAS associations, they fail to adequately leverage pleiotropy. Unlike approaches initiated from a user-provided gene list, our study only provides predictions for 29 diseases. By not relying on user-provided input, our predictions can serve as independent priors for future analyses. By predicting probabilities, we provide an extensible and interpretable assessment of association that circumvents the limitations inherent to frequentist analyses. Many approaches return no assessment for the majority of genes which fall outside of their set of predicted positives. Here, we overcome this issue and provide a comprehensive and genome-wide output by returning a probability of association for each protein-coding gene. High-throughput biological data is frequently noisy and incomplete. Combining orthogonal resources can help overcome these issues. Accordingly, we found that our integrative model outperformed any individual domain. While this method has shown encouraging performance, some limitations are worth noticing. For example, many biological networks preferentially cover well-studied vicinities. Knowledge biases that span multiple, presumably-orthogonal resources could diminish the benefits of integration. Here, several of the literature-derived domains were removed by the lasso, suggesting redundancy. In addition, biases in network completeness can lead to high-quality predictions for well-studied vicinities and low-quality predictions for poorly-studied vicinities. The permutation analysis provided evidence of this disparity: edge-specificity was critical for top predictions yet only moderately beneficial for the remainder. Subsequently, we caution users to avoid overinterpreting predictions for poorly-characterized genes. To help place predictions in context, the online browser provides a gene's mean prediction across all diseases and a disease's mean prediction across all genes. However, we recognize that false negatives will continue to persist in our predictions, and users should be mindful of this limitation when interpreting results. As more systematic and unbiased resources become available, high-quality predictions will emerge for more network vicinities. We reason that the desirable qualities of our predictions are the consequence of the heterogenous network edge prediction methodology. The approach is versatile (most biological phenomena are decomposable into entities connected by relationships), scalable (no theoretical limit to metagraph complexity or graph size), and efficient (low marginal cost to including an additional network component). We have extended the previous metapath-based framework set forth by PathPredict, by: 1) incorporating regularization allowing coefficient estimation for more features without overfitting; 2) designing a framework for predicting a metaedge that is included in the network; 3) developing an improved metric for assessing path specificity; and 4) implementing a degree-preserving permutation. Metapath-based heterogeneous network edge prediction provides a powerful new platform for bioinformatic discovery.\n",
      "Methods. Ethics statement. This study was approved by the UCSF institutional review board on human subjects under protocol #10-00104. Heterogeneous networks. We created a general framework and open source software package for representing heterogeneous networks. Like traditional graphs, heterogeneous networks consist of nodes connected by edges, except that an additional meta layer defines type. Node type signifies the kind of entity encoded, whereas edge type signifies the kind of relationship encoded. Edge types are comprised of a source node type, target node type, kind (to differentiate between multiple edge types connecting the same node types), and direction (allowing for both directed and undirected edge types). The user defines these types and annotates each node and edge, upon creation, with its corresponding type. The meta layer itself can be represented as a graph consisting of node types connected by edge types. When referring to this graph of types, we use the prefix 'meta'. Metagraphs:called schemas in previous work :consist of metanodes connected by metaedges. In a heterogeneous network, each path, a series of edges with common intermediary nodes, corresponds to a metapath representing the type of path. A path's metapath is the series of metaedges corresponding to that path's edges. The possible metapaths within a heterogeneous network can be enumerated by traversing the metagraph. We implemented this framework as an object-oriented data structure in python and named the resulting package hetio. Users are free to browse, use, or contribute to the software, through the online repository (https://github.com/dhimmel/hetio). Network construction. The included resources, and hence the metaedges and metanodes composing our network, were selected empirically based on a balance among the following properties: 1) quality:relevance to human pathogenesis; high accuracy and an optimal trade-off between false positives and false negatives. In some cases, quality concerns prevented the inclusion of a desired metaedge. For example, we omitted ontology-based disease similarly due to an inaccurate Disease Ontology hierarchy, and we omitted disease comorbidity due to high measurement error for uncommon diseases. For included metaedges, we attempted to select the highest quality resource in that domain. 2) reusability:easily retrievable and parsable; mapped to controlled vocabularies; well documented; amenable to reproducible (scripted) analysis; free of prohibitive reuse stipulations. 3) throughput:broad domain-specific coverage generated using systematic platforms that minimize bias. While genetic interactions have previously proven informative, their sparse characterization in humans was deemed unfavorable for our approach. 4) diversified, multiscale portrayal of biology:capturing, in aggregate, many aspects of pathophysiology across multiple levels of biological complexity. Levels of the hierarchical architecture of biological complexity include the genome, transcriptome, proteome, interactome, metabolome, cell and tissue organization, and phenome. Balancing these considerations, we integrated as many resources as possible within our computational runtime constraints. Nodes. Protein-coding genes (S3 Data) were extracted from the HGNC database. Resources were mapped to HGNC terms via gene symbol (ambiguous symbols were resolved in the order: approved, previous, synonyms) or Entrez identifiers. Disease nodes (S6 and S7 Data) were taken from the Disease Ontology (DO). Due to the limited number of diseases with GWAS, relevant disease references were manually mapped to the DO (S11 Data). Tissues were taken from the BRENDA Tissue Ontology (BTO). Only tissues with profiled expression were included enabling manual mapping. Nodes for the 14 MSigDB metanodes were directly imported from the Molecular Signature Database version 4.0. All MSigDB collections were included except those that were supersets of other collections. For example, 'C3: motif gene sets' was the union of two disjoint collections ('C3: microRNA targets' and 'C3: transcription factor targets') and was therefore excluded. Diseases were classified manually into 10 categories according to pathophysiology. The 'idiopathic' and 'unspecific' categories were not included as pathophysiology nodes, since they do not signify meaningful similarities between member diseases. Associations. Disease-gene associations were extracted from the GWAS Catalog, a compilation of GWAS associations where p < 10-5. First, associations were segregated by disease. GWAS Catalog phenotypes were converted to Experimental Factor Ontology (EFO) terms using mappings produced by the European Bioinformatics Institute. Associations mapping to multiple EFO terms were excluded to eliminate cross-phenotype studies. We manually mapped EFO to DO terms (now included in the DO as cross-references) and annotated each DO term with its associations. Associations were classified as either high or low-confidence, where exceeding two thresholds granted high-confidence status. First, p <= 5*10-8 corresponding to p <= 0.05 after Bonferroni adjustment for one million comparisons (an approximate upper bound for the number of independent SNPs evaluated by most GWAS). Second, a minimum sample size (counting both cases and controls) of 1,000 was required, since studies below this size are underpowered :i.e. any discovered associations are more likely than not to be false:for the majority of true effect size distributions commonly assumed to underlie complex disease etiology. Lead-SNPs were assigned windows:regions wherein the causal SNPs are assumed to lie:retrieved from the DAPPLE server. Windows were calculated for each lead-SNP by finding the furthest upstream and downstream SNPs where r 2 > 0.5 and extending outwards to the next recombination hotspot. Associations were ordered by confidence, sorting on following criteria: high/low confidence, p-value (low to high), and recency. In order of confidence, associations were overlapped by their windows into disease-specific loci (S4 Data). By organizing associations into loci, associations from multiple studies tagging the same underlying signal were condensed (S1 Fig). A locus was classified as high-confidence if any of its composite associations were high-confidence and low-confidence otherwise. For each disease-specific loci, we attempted to identify a primary gene. The primary gene was resolved in the following order: 1) the mode author-reported gene; 2) the containing gene for an intragenic lead-SNP; 3) the mode author-reported gene for an intragenic lead-SNP (in the case of overlapping genes); 4) the mode author-reported gene of the most proximal up and downstream genes. Steps 2-4 were repeated on each association composing the loci, in order of confidence, until a single gene resolved as primary. Loci where ambiguity was unresolvable or where no genes were returned did not receive a primary gene. All non-primary genes:genes that were author-reported, overlapping the lead-SNP, or immediately up or downstream from the lead-SNP:were considered secondary. Accordingly, four categories of processed associations were created: high-confidence primary, high-confidence secondary, low-confidence primary, and low-confidence secondary (S5 Data). We assume that our primary gene annotation for each loci represents the single causal gene responsible for the association. To investigate the validity of this assumption, we evaluated the performance of our predictions separately using each category of association as positives (S8 Fig). For both confidence levels, primary associations outperformed secondary associations suggesting our method succeeded at categorizing causal genes as primary. However, for high-confidence secondary associations, the AUROC equaled 0.74, which could result from multiple causal genes per loci or categorizing sole causal genes as secondary. The performance decline from high to low confidence associations was severe, pointing to a preponderance of falsely identified loci in the GWAS Catalog when p > 5x10-8 or sample size drops below 1000. Protein interactions. Physical protein-protein interactions (S8 Data) were extracted from iRefIndex 12.0, a compilation of 15 primary interaction databases. The iRefIndex was processed with ppiTrim to convert proteins to genes, remove protein complexes, and condense duplicated entries. Tissue-specific gene expression. Tissue-specific gene expression levels (S9 Data) were extracted from the GNF Gene Expression Atlas. Starting with the GCRMA-normalized and multisample-averaged expression values, 44,775 probes were converted to 16,466 HGNC genes and 84 tissues were manually mapped and converted to 77 BTO terms. For both conversions, the geometric mean was used to average expression values. The log base 10 of expression value was used as the threshold criteria for GeT edge inclusion. Disease localization. Disease localization was calculated for the 77 tissues with expression profiles (S10 Data). Literature co-occurrence was used to assess whether a tissue is affected by a disease. We used CoPub 5.0 to extract R-scaled scores between tissues and diseases measuring whether two terms occurred together in Medline abstracts more than would be expected by chance. DO terms for diseases with GWAS and BTO tissues with expression profiles were manually mapped to the 'biological identifier' terminology used by CoPub. The R-scaled score was used as the threshold criteria for TlD edge inclusion. Feature computation metrics. The simplest metapath-based metric is path count (PC): the number of paths, of a specified metapath, between a source and target node. However, PC does not adjust for the extent of graph connectivity along the path. Paths traversing high-degree nodes will account for a large portion of the PC, despite high-degree nodes frequently representing a biologically broad or vague entity with little informativeness. The previous work evaluated several metrics that include a PC denominator to adjust for connectivity and reported that normalized path count (NPC) performed best. The denominator for NPC equals the number of paths from the source to any target plus the number of paths from any target to the source.  where m is the metapath, s is the source node, t is the target node, S m is the set of nodes corresponding to the source metanode of m, and T m is the set of nodes corresponding to the target metanode of m. We adopt the any source/target concept to compute the two GaD features. However, dividing the PC by a denominator is flawed because each path composing the PC deserves a distinct degree adjustment. If two paths:one traversing only high-degree nodes and one traversing only low-degree nodes:compose the PC, the network surrounding the high-degree path will monopolize the NPC denominator and overwhelm the contribution of the low-degree path despite its specificity. Therefore, we developed the degree-weighted path count (DWPC) which individually downweights each path between a source and target node. Each path receives a path-degree product (PDP) calculated by: 1) extracting all metaedge-specific degrees along the path (D path), where each edge composing the path contributes two degrees; 2) raising each degree to the -w power, where w >= 0 and is called the damping exponent; 3) multiplying all exponentiated degrees to yield the PDP. The DWPC equals the sum of PDPs. See Fig 2C and 2D for a visual description of the DWPC. Machine learning approach. PathPredict relied on basic logistic regression to predict coauthorship status from features corresponding to nine distinct metapaths. However, faced with fewer positives to train our model and a large number of features, we adopted a regularized approach, which aims to contain the overfitting tendencies inherent to regression. Regularization penalizes complexity, a trademark of overfitting. We chose the elastic net technique of regularization, which is efficiently implemented for logistic regression by the R glmnet package. Regularized logistic regression requires a parameter, lambda, setting the strength of regularization. We optimized lambda separately for each model fit. Using 10-fold cross-validation and the \"one-standard-error\" rule to choose the optimal lambda from deviance, we adopted a conservative approach designed to prevent overfitting. On the training set of gene-disease pairs, we optimized the elastic net mixing parameter (alpha), the DWPC damping exponent (w), and two edge inclusion thresholds. First, we optimized alpha and w on the 20 features whose metapaths did not include threshold-dependent metaedges. For each combination of alpha and w, we calculated average testing AUROC using 20-fold cross-validation repeated for 10 randomized partitionings. After setting alpha and w, we jointly optimized the two edge-inclusion thresholds using the AUROC for the GeTlD feature, whose metapath is composed from the two edges requiring thresholds (S3 Fig). We adopt standardized coefficients as a measure of feature effect size. Standardized coefficients refer to the coefficients from logistic regression when all features have been transformed to z-scores. Standardization provides a common scale to assess feature effect, both within and across models. Degree-preserving permutation. Starting from the complete network, a permuted network was created by swapping edges separately for each metaedge. Edge swaps were performed by switching the target nodes for two randomly selecting edges. For each metaedge, the number of attempted swaps was ten times the corresponding edge count. We adopted a Markov Chain strategy where additional rounds of permutation were initiated from the most-recently permuted network. A training network was generated from the first permuted network by masking 25% of the associations for testing. Testing performance for the permuted training network model is shown in S4 Fig. When contrasting this performance with the unpermuted-network model, we employed the Condensed-ROC curve to magnify the importance of top predictions. Using the exponential transformation with a magnification factor of 460:the value which maps a FPR of 0.01 to 0.99:we concentrated on the top 1% of predictions (S4 Fig). A one-sided unpaired DeLong test was used to assess whether feature-specific testing AUROCs from the unpermuted network exceeded those from the first permuted network (S2 Table). Gene set subsampling. We performed a subsampling analysis for 15 gene sets:the 14 MSigDB gene sets and tissues:to assess the effect of sparsity on feature-specific performance (S7 Fig). Two without-replacement subsampling schemes were investigated: node masking and edge masking. For a specific gene set and scheme, we masked a percentage of the gene set and calculated the corresponding feature's AUROC. We evaluated a range of percentages and performed ten subsampling repetitions for each percentage. Multiple sclerosis gene discovery. We excluded 588 genes from the discovery phase of the multiple sclerosis analysis. First we excluded genes in the extended MHC region (spanning from SCGN to SYNGAP1 on chromosome 6) due to the complex pattern of linkage characterizing this region containing several highly-penetrant MS-risk alleles. Second, we excluded putative MS genes: high-confidence primary genes from the GWAS Catalog and reported genes for the WTCCC2-replicated loci. We omitted genes in linkage disequilibrium with the putative genes by excluding: 1) consecutive sequences of nominally significant genes (using the WTCCC2-VEGAS p-values) that included a putative gene; and 2) high-confidence secondary genes from the GWAS catalog. Post exclusion, 1211 genes were nominally significant in Meta2.5, four of which exceeded the network-based discovery threshold. Using a hypergeometric test for overrepresentation, we calculated the probability of randomly selecting 4 of the 1211 genes and Bonferroni validating at least 3 of the 4 on WTCCC2 (S3 Table). Data availability. See S1-S12 Datasets for the supporting data and S13 Data for vector figures. The website provides additional resources (http://het.io/disease-genes/downloads/) as well as an interface for browsing results (http://het.io/disease-genes/browse/). Project related code is available from the github repository (https://github.com/dhimmel/hetio).\n"
     ]
    }
   ],
   "source": [
    "# if collapsing to embed by section?\n",
    "\n",
    "for each in df_filtered.groupby('section'):\n",
    "\n",
    "    section_rows = df_filtered[df_filtered['section']==each]\n",
    "\n",
    "    section_rows = section_rows.sort_values(by='index', ascending=True)\n",
    "\n",
    "    section_rows['text'] = (' '.join(section_rows['sentence']))\n",
    "\n",
    "    section_rows = section_rows.drop_duplicates(subset=['text', 'section', ], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "section_rows.drop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('abstract',\n",
       "     index   section                                           sentence  \\\n",
       "  1      1  abstract  The first decade of Genome Wide Association St...   \n",
       "  2      2  abstract                                    Author Summary.   \n",
       "  3      3  abstract  For complex human diseases, identifying the ge...   \n",
       "  \n",
       "         pmid       pmcid                      date  \\\n",
       "  1  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  2  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  3  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  \n",
       "                             authors           journal  \n",
       "  1  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  2  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  3  Himmelstein DS and Baranzini SE  PLoS Comput Biol  ),\n",
       " ('discuss',\n",
       "      index  section                                           sentence  \\\n",
       "  60     60  discuss                                        Discussion.   \n",
       "  61     61  discuss  In this work, we developed a framework to pred...   \n",
       "  62     62  discuss  Most previous explorations of the factors unde...   \n",
       "  63     63  discuss  We observed a prominent influence of pleiotrop...   \n",
       "  64     64  discuss  High-throughput biological data is frequently ...   \n",
       "  65     65  discuss  We reason that the desirable qualities of our ...   \n",
       "  \n",
       "          pmid       pmcid                      date  \\\n",
       "  60  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  61  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  62  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  63  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  64  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  65  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  \n",
       "                              authors           journal  \n",
       "  60  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  61  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  62  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  63  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  64  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  65  Himmelstein DS and Baranzini SE  PLoS Comput Biol  ),\n",
       " ('intro',\n",
       "      index section                                           sentence  \\\n",
       "  4       4   intro                                      Introduction.   \n",
       "  5       5   intro  In the last decade, genome-wide association st...   \n",
       "  6       6   intro  A central assumption in GWAS is that every reg...   \n",
       "  7       7   intro  Several strategies have been attempted to iden...   \n",
       "  8       8   intro  Prioritization strategies generally rely on us...   \n",
       "  9       9   intro  To approach this problem, we resorted to a met...   \n",
       "  10     10   intro  While still a developing field, network-based ...   \n",
       "  \n",
       "          pmid       pmcid                      date  \\\n",
       "  4   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  5   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  6   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  7   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  8   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  9   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  10  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  \n",
       "                              authors           journal  \n",
       "  4   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  5   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  6   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  7   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  8   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  9   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  10  Himmelstein DS and Baranzini SE  PLoS Comput Biol  ),\n",
       " ('methods',\n",
       "       index  section                                           sentence  \\\n",
       "  66      66  methods                                           Methods.   \n",
       "  67      67  methods                                  Ethics statement.   \n",
       "  68      68  methods  This study was approved by the UCSF institutio...   \n",
       "  69      69  methods                            Heterogeneous networks.   \n",
       "  70      70  methods  We created a general framework and open source...   \n",
       "  71      71  methods                              Network construction.   \n",
       "  72      72  methods  The included resources, and hence the metaedge...   \n",
       "  73      73  methods                                             Nodes.   \n",
       "  74      74  methods  Protein-coding genes (S3 Data) were extracted ...   \n",
       "  75      75  methods                                      Associations.   \n",
       "  76      76  methods  Disease-gene associations were extracted from ...   \n",
       "  77      77  methods  Associations were classified as either high or...   \n",
       "  78      78  methods  Lead-SNPs were assigned windows:regions wherei...   \n",
       "  79      79  methods  For each disease-specific loci, we attempted t...   \n",
       "  80      80  methods  Accordingly, four categories of processed asso...   \n",
       "  81      81  methods                              Protein interactions.   \n",
       "  82      82  methods  Physical protein-protein interactions (S8 Data...   \n",
       "  83      83  methods                   Tissue-specific gene expression.   \n",
       "  84      84  methods  Tissue-specific gene expression levels (S9 Dat...   \n",
       "  85      85  methods                              Disease localization.   \n",
       "  86      86  methods  Disease localization was calculated for the 77...   \n",
       "  87      87  methods                       Feature computation metrics.   \n",
       "  88      88  methods  The simplest metapath-based metric is path cou...   \n",
       "  89      89  methods                   The DWPC equals the sum of PDPs.   \n",
       "  90      90  methods  See Fig 2C and 2D for a visual description of ...   \n",
       "  91      91  methods                         Machine learning approach.   \n",
       "  92      92  methods  PathPredict relied on basic logistic regressio...   \n",
       "  93      93  methods  Regularized logistic regression requires a par...   \n",
       "  94      94  methods  On the training set of gene-disease pairs, we ...   \n",
       "  95      95  methods  We adopt standardized coefficients as a measur...   \n",
       "  96      96  methods                     Degree-preserving permutation.   \n",
       "  97      97  methods  Starting from the complete network, a permuted...   \n",
       "  98      98  methods                              Gene set subsampling.   \n",
       "  99      99  methods  We performed a subsampling analysis for 15 gen...   \n",
       "  100    100  methods                 Multiple sclerosis gene discovery.   \n",
       "  101    101  methods  We excluded 588 genes from the discovery phase...   \n",
       "  102    102  methods                                 Data availability.   \n",
       "  103    103  methods  See S1-S12 Datasets for the supporting data an...   \n",
       "  \n",
       "           pmid       pmcid                      date  \\\n",
       "  66   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  67   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  68   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  69   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  70   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  71   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  72   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  73   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  74   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  75   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  76   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  77   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  78   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  79   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  80   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  81   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  82   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  83   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  84   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  85   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  86   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  87   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  88   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  89   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  90   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  91   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  92   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  93   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  94   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  95   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  96   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  97   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  98   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  99   26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  100  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  101  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  102  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  103  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  \n",
       "                               authors           journal  \n",
       "  66   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  67   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  68   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  69   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  70   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  71   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  72   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  73   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  74   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  75   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  76   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  77   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  78   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  79   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  80   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  81   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  82   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  83   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  84   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  85   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  86   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  87   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  88   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  89   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  90   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  91   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  92   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  93   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  94   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  95   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  96   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  97   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  98   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  99   Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  100  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  101  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  102  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  103  Himmelstein DS and Baranzini SE  PLoS Comput Biol  ),\n",
       " ('results',\n",
       "      index  section                                           sentence  \\\n",
       "  11     11  results                                           Results.   \n",
       "  12     12  results  Constructing a heterogeneous network to integr...   \n",
       "  13     13  results  Using publicly-available databases and standar...   \n",
       "  22     22  results  Gene-disease associations were extracted from ...   \n",
       "  23     23  results  Features quantify the network topology between...   \n",
       "  24     24  results  To describe the network topology connecting a ...   \n",
       "  25     25  results  Machine learning approach to predict the proba...   \n",
       "  26     26  results  Further analysis focused on the 29 diseases wi...   \n",
       "  29     29  results  To learn the importance of each feature and mo...   \n",
       "  30     30  results  Method prioritizes associations withheld for t...   \n",
       "  31     31  results  We extracted network-based features for gene-d...   \n",
       "  34     34  results   Predicting associations on the complete network.   \n",
       "  35     35  results  As a next step in our analysis, we recomputed ...   \n",
       "  38     38  results  Degree-preserving network permutations highlig...   \n",
       "  39     39  results  Using Markov chain randomized edge-swaps, we c...   \n",
       "  40     40  results  Compared to the unpermuted-network model, test...   \n",
       "  41     41  results  Interestingly, predictions from the permuted-n...   \n",
       "  42     42  results  Feature importance identifies the mechanisms u...   \n",
       "  43     43  results  We assessed the informativeness of each featur...   \n",
       "  46     46  results  Features whose metapaths originate with an ass...   \n",
       "  47     47  results  We also observed that the lasso regression mod...   \n",
       "  48     48  results  Existing network-based gene-prioritization met...   \n",
       "  49     49  results                               Gene set robustness.   \n",
       "  50     50  results  For each type of gene set, we evaluated the ef...   \n",
       "  51     51  results  Case study: Prioritizing multiple sclerosis as...   \n",
       "  52     52  results  The WTCCC2 multiple sclerosis (MS) GWAS tested...   \n",
       "  55     55  results  Prioritizing statistical candidates with netwo...   \n",
       "  56     56  results  Finally, we designed a framework for discoveri...   \n",
       "  57     57  results  This strategy discovered four genes, three of ...   \n",
       "  \n",
       "          pmid       pmcid                      date  \\\n",
       "  11  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  12  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  13  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  22  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  23  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  24  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  25  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  26  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  29  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  30  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  31  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  34  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  35  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  38  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  39  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  40  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  41  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  42  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  43  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  46  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  47  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  48  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  49  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  50  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  51  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  52  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  55  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  56  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  57  26158728  PMC4497619 2015-07-09 00:00:00+00:00   \n",
       "  \n",
       "                              authors           journal  \n",
       "  11  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  12  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  13  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  22  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  23  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  24  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  25  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  26  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  29  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  30  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  31  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  34  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  35  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  38  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  39  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  40  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  41  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  42  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  43  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  46  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  47  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  48  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  49  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  50  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  51  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  52  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  55  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  56  Himmelstein DS and Baranzini SE  PLoS Comput Biol  \n",
       "  57  Himmelstein DS and Baranzini SE  PLoS Comput Biol  ),\n",
       " ('title',\n",
       "     index section                                           sentence      pmid  \\\n",
       "  0      0   title  Heterogeneous Network Edge Prediction: A Data ...  26158728   \n",
       "  \n",
       "          pmcid                      date                          authors  \\\n",
       "  0  PMC4497619 2015-07-09 00:00:00+00:00  Himmelstein DS and Baranzini SE   \n",
       "  \n",
       "              journal  \n",
       "  0  PLoS Comput Biol  )]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x for x in df_filtered.groupby('section')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloaded full text XML as 4448285.xml\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "def annotate_pmids(pmids):\n",
    "    # Define the PubTator API URL\n",
    "    pubtator_url = \"https://www.ncbi.nlm.nih.gov/research/pubtator-api/publications/export/biocxml\"\n",
    "\n",
    "    # Join the PMIDs into a comma-separated string\n",
    "    pmid_str = ','.join(pmids)\n",
    "\n",
    "    # Send a GET request to the API with the list of PMIDs\n",
    "    response = requests.get(pubtator_url, params={\"pmids\": pmid_str, \"full\": True})\n",
    "\n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        with open(f\"PMC4448285.xml\", \"w\", encoding=\"utf-8\") as f:\n",
    "            f.write(response.text)\n",
    "        print(f\"Downloaded full text XML as 4448285.xml\")\n",
    "    # if response.status_code == 200:\n",
    "        #return response.json()  # Return the response in JSON format\n",
    "    else:\n",
    "        raise Exception(f\"Error {response.status_code}: {response.text}\")\n",
    "\n",
    "# Example usage\n",
    "pmids = [\"25133803\", \"25971816\"]\n",
    "annotations = annotate_pmids(pmids)\n",
    "\n",
    "# Display the annotated results\n",
    "import pprint\n",
    "pprint.pprint(annotations)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import xml.etree.ElementTree as ET\n",
    "\n",
    "def extract_full_text_from_bioc(xml_file):\n",
    "    # Parse the XML file\n",
    "    tree = ET.parse(xml_file)\n",
    "    root = tree.getroot()\n",
    "\n",
    "    # Initialize a list to store text passages\n",
    "    full_text = []\n",
    "\n",
    "    # Iterate through passages in the document\n",
    "    for document in root.findall(\".//document\"):\n",
    "        for passage in document.findall(\"passage\"):\n",
    "            # Extract the text content\n",
    "            text = passage.find(\"text\").text\n",
    "            if text:\n",
    "                full_text.append(text.strip())\n",
    "\n",
    "    # Join all passages into a single string with new lines\n",
    "    #return \"\\n\\n\".join(full_text)\n",
    "    return full_text\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bkg-review-FcnO91kN-py3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
