{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Purpose of this notebook is to get full text articles from PMC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import numpy as np \n",
    "import pandas as pd \n",
    "import matplotlib.pyplot as plt \n",
    "import re\n",
    "import requests\n",
    "\n",
    "# import nltk\n",
    "# from nltk.stem import WordNetLemmatizer\n",
    "# nltk.download('wordnet')\n",
    "\n",
    "#from spellchecker import SpellChecker"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>doi</th>\n",
       "      <th>arxiv</th>\n",
       "      <th>pmid</th>\n",
       "      <th>publicationDate</th>\n",
       "      <th>title</th>\n",
       "      <th>authors</th>\n",
       "      <th>abstract</th>\n",
       "      <th>keywords</th>\n",
       "      <th>meshTerms</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Sun2014</td>\n",
       "      <td>10.1039/C4IB00122B</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25133803.0</td>\n",
       "      <td>2014-08-18</td>\n",
       "      <td>The integrated disease network</td>\n",
       "      <td>Kai Sun and Natalie Buchan and Chris Larminie ...</td>\n",
       "      <td>The growing body of transcriptomic, proteomic,...</td>\n",
       "      <td>\"phenotype, crohn's disease, heterogeneity, ge...</td>\n",
       "      <td>['Computational Biology / methods', 'Databases...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ernst2015</td>\n",
       "      <td>10.1186/s12859-015-0549-5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25971816.0</td>\n",
       "      <td>2015-05-14</td>\n",
       "      <td>KnowLife: a versatile approach for constructin...</td>\n",
       "      <td>Patrick Ernst and Amy Siu and Gerhard Weikum</td>\n",
       "      <td>Background: Biomedical knowledge bases (KB's) ...</td>\n",
       "      <td>['Biomedical text mining','knowledge base','re...</td>\n",
       "      <td>['Biomedical Research', 'Humans', 'Information...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Himmelstein2015</td>\n",
       "      <td>10.1371/journal.pcbi.1004259</td>\n",
       "      <td>NaN</td>\n",
       "      <td>26158728.0</td>\n",
       "      <td>2015-07-09</td>\n",
       "      <td>Heterogeneous Network Edge Prediction: A Data ...</td>\n",
       "      <td>Daniel Himmelstein and Serio Baranzini</td>\n",
       "      <td>The first decade of Genome Wide Association St...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Algorithms\\nAnimals\\nChromosome Mapping / meth...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Himmelstein2017</td>\n",
       "      <td>10.7554/eLife.26726</td>\n",
       "      <td>10.1101/087619v3</td>\n",
       "      <td>28936969.0</td>\n",
       "      <td>2017-09-22</td>\n",
       "      <td>Systematic integration of biomedical knowledge...</td>\n",
       "      <td>Daniel Scott Himmelstein and Antoine Lizee and...</td>\n",
       "      <td>The ability to computationally predict whether...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Computational Biology / methods*\\nDrug Discove...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Martinez2015</td>\n",
       "      <td>10.1016/j.artmed.2014.11.003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25704113.0</td>\n",
       "      <td>2015-01-13</td>\n",
       "      <td>DrugNet: Network-based drug–disease prioritiza...</td>\n",
       "      <td>Víctor Martínez and Carmen Navarro and Carlos ...</td>\n",
       "      <td>Objective: Computational drug repositioning ca...</td>\n",
       "      <td>Data integration; Disease networks; Drug repos...</td>\n",
       "      <td>Area Under Curve\\nComputational Biology*\\nComp...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              name                            doi             arxiv  \\\n",
       "0          Sun2014             10.1039/C4IB00122B               NaN   \n",
       "1        Ernst2015      10.1186/s12859-015-0549-5               NaN   \n",
       "2  Himmelstein2015   10.1371/journal.pcbi.1004259               NaN   \n",
       "3  Himmelstein2017            10.7554/eLife.26726  10.1101/087619v3   \n",
       "4     Martinez2015  10.1016/j.artmed.2014.11.003                NaN   \n",
       "\n",
       "         pmid publicationDate  \\\n",
       "0  25133803.0      2014-08-18   \n",
       "1  25971816.0      2015-05-14   \n",
       "2  26158728.0      2015-07-09   \n",
       "3  28936969.0      2017-09-22   \n",
       "4  25704113.0      2015-01-13   \n",
       "\n",
       "                                               title  \\\n",
       "0                     The integrated disease network   \n",
       "1  KnowLife: a versatile approach for constructin...   \n",
       "2  Heterogeneous Network Edge Prediction: A Data ...   \n",
       "3  Systematic integration of biomedical knowledge...   \n",
       "4  DrugNet: Network-based drug–disease prioritiza...   \n",
       "\n",
       "                                             authors  \\\n",
       "0  Kai Sun and Natalie Buchan and Chris Larminie ...   \n",
       "1       Patrick Ernst and Amy Siu and Gerhard Weikum   \n",
       "2             Daniel Himmelstein and Serio Baranzini   \n",
       "3  Daniel Scott Himmelstein and Antoine Lizee and...   \n",
       "4  Víctor Martínez and Carmen Navarro and Carlos ...   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  The growing body of transcriptomic, proteomic,...   \n",
       "1  Background: Biomedical knowledge bases (KB's) ...   \n",
       "2  The first decade of Genome Wide Association St...   \n",
       "3  The ability to computationally predict whether...   \n",
       "4  Objective: Computational drug repositioning ca...   \n",
       "\n",
       "                                            keywords  \\\n",
       "0  \"phenotype, crohn's disease, heterogeneity, ge...   \n",
       "1  ['Biomedical text mining','knowledge base','re...   \n",
       "2                                                NaN   \n",
       "3                                                NaN   \n",
       "4  Data integration; Disease networks; Drug repos...   \n",
       "\n",
       "                                           meshTerms  \n",
       "0  ['Computational Biology / methods', 'Databases...  \n",
       "1  ['Biomedical Research', 'Humans', 'Information...  \n",
       "2  Algorithms\\nAnimals\\nChromosome Mapping / meth...  \n",
       "3  Computational Biology / methods*\\nDrug Discove...  \n",
       "4  Area Under Curve\\nComputational Biology*\\nComp...  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reading in the publication list\n",
    "excel_file = 'data/bkg registry nelists.xlsx'\n",
    "pubs = pd.read_excel(excel_file, sheet_name='Publication')\n",
    "# check\n",
    "pubs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 87 entries, 0 to 86\n",
      "Data columns (total 10 columns):\n",
      " #   Column           Non-Null Count  Dtype         \n",
      "---  ------           --------------  -----         \n",
      " 0   name             87 non-null     object        \n",
      " 1   doi              86 non-null     object        \n",
      " 2   arxiv            11 non-null     object        \n",
      " 3   pmid             76 non-null     float64       \n",
      " 4   publicationDate  87 non-null     datetime64[ns]\n",
      " 5   title            87 non-null     object        \n",
      " 6   authors          86 non-null     object        \n",
      " 7   abstract         87 non-null     object        \n",
      " 8   keywords         45 non-null     object        \n",
      " 9   meshTerms        65 non-null     object        \n",
      "dtypes: datetime64[ns](1), float64(1), object(8)\n",
      "memory usage: 6.9+ KB\n"
     ]
    }
   ],
   "source": [
    "pubs.info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "testing out pubtator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'PubTator3': [{'_id': '25971816|PMC4448285',\n",
      "                'authors': ['Ernst P', 'Siu A', 'Weikum G'],\n",
      "                'date': '2015-05-14T00:00:00Z',\n",
      "                'id': '4448285',\n",
      "                'infons': {},\n",
      "                'journal': 'BMC Bioinformatics',\n",
      "                'meta': {},\n",
      "                'passages': [{'annotations': [],\n",
      "                              'infons': {'article-id_doi': '10.1186/s12859-015-0549-5',\n",
      "                                         'article-id_pmc': '4448285',\n",
      "                                         'article-id_pmid': '25971816',\n",
      "                                         'article-id_publisher-id': '549',\n",
      "                                         'elocation-id': '157',\n",
      "                                         'kwd': 'Biomedical text mining '\n",
      "                                                'Knowledge base Relation '\n",
      "                                                'extraction',\n",
      "                                         'license': 'This is an Open Access '\n",
      "                                                    'article distributed under '\n",
      "                                                    'the terms of the Creative '\n",
      "                                                    'Commons Attribution '\n",
      "                                                    'License '\n",
      "                                                    '(http://creativecommons.org/licenses/by/4.0), '\n",
      "                                                    'which permits '\n",
      "                                                    'unrestricted use, '\n",
      "                                                    'distribution, and '\n",
      "                                                    'reproduction in any '\n",
      "                                                    'medium, provided the '\n",
      "                                                    'original work is properly '\n",
      "                                                    'credited. The Creative '\n",
      "                                                    'Commons Public Domain '\n",
      "                                                    'Dedication waiver '\n",
      "                                                    '(http://creativecommons.org/publicdomain/zero/1.0/) '\n",
      "                                                    'applies to the data made '\n",
      "                                                    'available in this '\n",
      "                                                    'article, unless otherwise '\n",
      "                                                    'stated.',\n",
      "                                         'name_0': 'surname:Ernst;given-names:Patrick',\n",
      "                                         'name_1': 'surname:Siu;given-names:Amy',\n",
      "                                         'name_2': 'surname:Weikum;given-names:Gerhard',\n",
      "                                         'section_type': 'TITLE',\n",
      "                                         'title': 'Keywords',\n",
      "                                         'type': 'front',\n",
      "                                         'volume': '16',\n",
      "                                         'year': '2015'},\n",
      "                              'offset': 0,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'KnowLife: a versatile approach for '\n",
      "                                      'constructing a large knowledge graph '\n",
      "                                      'for biomedical sciences'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract_title_1'},\n",
      "                              'offset': 96,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Background'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract'},\n",
      "                              'offset': 107,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': \"Biomedical knowledge bases (KB's) have \"\n",
      "                                      'become important assets in life '\n",
      "                                      'sciences. Prior work on KB construction '\n",
      "                                      'has three major limitations. First, '\n",
      "                                      'most biomedical KBs are manually built '\n",
      "                                      'and curated, and cannot keep up with '\n",
      "                                      'the rate at which new findings are '\n",
      "                                      'published. Second, for automatic '\n",
      "                                      'information extraction (IE), the text '\n",
      "                                      'genre of choice has been scientific '\n",
      "                                      'publications, neglecting sources like '\n",
      "                                      'health portals and online communities. '\n",
      "                                      'Third, most prior work on IE has '\n",
      "                                      'focused on the molecular level or '\n",
      "                                      'chemogenomics only, like '\n",
      "                                      'protein-protein interactions or '\n",
      "                                      'gene-drug relationships, or solely '\n",
      "                                      'address highly specific topics such as '\n",
      "                                      'drug effects.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract_title_1'},\n",
      "                              'offset': 763,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Results'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract'},\n",
      "                              'offset': 771,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We address these three limitations by a '\n",
      "                                      'versatile and scalable approach to '\n",
      "                                      'automatic KB construction. Using a '\n",
      "                                      'small number of seed facts for distant '\n",
      "                                      'supervision of pattern-based '\n",
      "                                      'extraction, we harvest a huge number of '\n",
      "                                      'facts in an automated manner without '\n",
      "                                      'requiring any explicit training.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract'},\n",
      "                              'offset': 1059,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We extend previous techniques for '\n",
      "                                      'pattern-based IE with confidence '\n",
      "                                      'statistics, and we combine this '\n",
      "                                      'recall-oriented stage with logical '\n",
      "                                      'reasoning for consistency constraint '\n",
      "                                      'checking to achieve high precision. To '\n",
      "                                      'our knowledge, this is the first method '\n",
      "                                      'that uses consistency checking for '\n",
      "                                      'biomedical relations. Our approach can '\n",
      "                                      'be easily extended to incorporate '\n",
      "                                      'additional relations and constraints.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract'},\n",
      "                              'offset': 1455,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We ran extensive experiments not only '\n",
      "                                      'for scientific publications, but also '\n",
      "                                      'for encyclopedic health portals and '\n",
      "                                      'online communities, creating different '\n",
      "                                      \"KB's based on different configurations. \"\n",
      "                                      'We assess the size and quality of each '\n",
      "                                      'KB, in terms of number of facts and '\n",
      "                                      'precision. The best configured KB, '\n",
      "                                      'KnowLife, contains more than 500,000 '\n",
      "                                      'facts at a precision of 93% for 13 '\n",
      "                                      'relations covering genes, organs, '\n",
      "                                      'diseases, symptoms, treatments, as well '\n",
      "                                      'as environmental and lifestyle risk '\n",
      "                                      'factors.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract_title_1'},\n",
      "                              'offset': 1949,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Conclusion'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract'},\n",
      "                              'offset': 1960,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'KnowLife is a large knowledge base for '\n",
      "                                      'health and life sciences, automatically '\n",
      "                                      'constructed from different Web sources. '\n",
      "                                      'As a unique feature, KnowLife is '\n",
      "                                      'harvested from different text genres '\n",
      "                                      'such as scientific publications, health '\n",
      "                                      'portals, and online communities. Thus, '\n",
      "                                      'it has the potential to serve as '\n",
      "                                      'one-stop portal for a wide range of '\n",
      "                                      'relations and use cases. To showcase '\n",
      "                                      'the breadth and usefulness, we make the '\n",
      "                                      'KnowLife KB accessible through the '\n",
      "                                      'health portal '\n",
      "                                      '(http://knowlife.mpi-inf.mpg.de).'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract_title_1'},\n",
      "                              'offset': 2457,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Electronic supplementary material'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'ABSTRACT',\n",
      "                                         'type': 'abstract'},\n",
      "                              'offset': 2491,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The online version of this article '\n",
      "                                      '(doi:10.1186/s12859-015-0549-5) '\n",
      "                                      'contains supplementary material, which '\n",
      "                                      'is available to authorized users.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'title_1'},\n",
      "                              'offset': 2631,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Introduction'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 2644,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': \"Large knowledge bases (KB's) about \"\n",
      "                                      'entities, their properties, and the '\n",
      "                                      'relationships between entities, have '\n",
      "                                      'become an important asset for semantic '\n",
      "                                      'search, analytics, and smart '\n",
      "                                      'recommendations over Web contents and '\n",
      "                                      'other kinds of Big Data. Notable '\n",
      "                                      'projects are DBpedia, Yago, and the '\n",
      "                                      'Google Knowledge Graph with its public '\n",
      "                                      'core Freebase (freebase.com).'},\n",
      "                             {'annotations': [{'id': '16702',\n",
      "                                               'infons': {'accession': '@DISEASE_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004194',\n",
      "                                                          'name': 'Disease',\n",
      "                                                          'normalized': ['D004194'],\n",
      "                                                          'normalized_id': 'D004194',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 3060}],\n",
      "                                               'text': 'Disease'}],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 2998,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': \"In the biomedical domain, KB's such as \"\n",
      "                                      'the Gene Ontology, the Disease '\n",
      "                                      'Ontology, the National Drug File - '\n",
      "                                      'Reference Terminology, and the '\n",
      "                                      'Foundational Model of Anatomy are '\n",
      "                                      'prominent examples of the rich '\n",
      "                                      'knowledge that is digitally available. '\n",
      "                                      \"However, each of these KB's is highly \"\n",
      "                                      'specialized and covers only a relative '\n",
      "                                      'narrow topic within the life sciences, '\n",
      "                                      'and there is very little interlinkage '\n",
      "                                      \"between the KB's. Thus, in contrast to \"\n",
      "                                      \"the general-domain KB's that power Web \"\n",
      "                                      'search and analytics, there is no way '\n",
      "                                      'of obtaining an integrated view on all '\n",
      "                                      'aspects of biomedical knowledge. The '\n",
      "                                      'lack of a \"one-stop\" KB that spans '\n",
      "                                      'biological, medical, and health '\n",
      "                                      'knowledge, hinders the development of '\n",
      "                                      'advanced search and analytic '\n",
      "                                      'applications in this field.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 3758,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'In order to build a comprehensive '\n",
      "                                      'biomedical KB, the following three '\n",
      "                                      'bottlenecks must be addressed.'},\n",
      "                             {'annotations': [{'id': '16703',\n",
      "                                               'infons': {'accession': None,\n",
      "                                                          'biotype': 'species',\n",
      "                                                          'database': 'ncbi_taxonomy',\n",
      "                                                          'identifier': '9606',\n",
      "                                                          'name': '9606',\n",
      "                                                          'normalized': [9606],\n",
      "                                                          'normalized_id': 9606,\n",
      "                                                          'type': 'Species',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 5,\n",
      "                                                              'offset': 3954}],\n",
      "                                               'text': 'human'}],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 3858,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Beyond manual curation. Biomedical '\n",
      "                                      'knowledge is advancing at rates far '\n",
      "                                      'greater than any single human can '\n",
      "                                      'absorb. Therefore, relying on manual '\n",
      "                                      \"curation of KB's is bound to be a \"\n",
      "                                      'bottleneck. To fully leverage all '\n",
      "                                      'published knowledge, automated '\n",
      "                                      'information extraction (IE) from input '\n",
      "                                      'texts is mandatory.'},\n",
      "                             {'annotations': [{'id': '16704',\n",
      "                                               'infons': {'accession': None,\n",
      "                                                          'biotype': 'species',\n",
      "                                                          'database': 'ncbi_taxonomy',\n",
      "                                                          'identifier': '9606',\n",
      "                                                          'name': '9606',\n",
      "                                                          'normalized': [9606],\n",
      "                                                          'normalized_id': 9606,\n",
      "                                                          'type': 'Species',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 4485}],\n",
      "                                               'text': 'patient'},\n",
      "                                              {'id': '16705',\n",
      "                                               'infons': {'accession': None,\n",
      "                                                          'biotype': 'species',\n",
      "                                                          'database': 'ncbi_taxonomy',\n",
      "                                                          'identifier': '9606',\n",
      "                                                          'name': '9606',\n",
      "                                                          'normalized': [9606],\n",
      "                                                          'normalized_id': 9606,\n",
      "                                                          'type': 'Species',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 4301}],\n",
      "                                               'text': 'patient'}],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 4161,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Beyond scientific literature. Besides '\n",
      "                                      'scientific publications found in PubMed '\n",
      "                                      'Medline and PubMed Central, there are '\n",
      "                                      'substantial efforts on patient-oriented '\n",
      "                                      'health portals such as Mayo Clinic, '\n",
      "                                      \"Medline Plus, UpToDate, Wikipedia's \"\n",
      "                                      'Health Portal, and there are also '\n",
      "                                      'popular online discussion forums such '\n",
      "                                      'as healthboards.com or patient.co.uk. '\n",
      "                                      'All this constitutes a rich universe of '\n",
      "                                      'information, but the information is '\n",
      "                                      'spread across many sources, mostly in '\n",
      "                                      'textual, unstructured and sometimes '\n",
      "                                      'noisy form. Prior work on biomedical IE '\n",
      "                                      'has focused on scientific literature '\n",
      "                                      'only, and completely disregards the '\n",
      "                                      'opportunities that lie in tapping into '\n",
      "                                      'health portals and communities for '\n",
      "                                      'automated IE.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 4853,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Beyond molecular entities. IE from '\n",
      "                                      'biomedical texts has strongly focused '\n",
      "                                      'on entities and relations at the '\n",
      "                                      'molecular level; a typical IE task is '\n",
      "                                      'to extract protein-protein '\n",
      "                                      'interactions. There is very little work '\n",
      "                                      'on comprehensive approaches that link '\n",
      "                                      'diverse entity types, spanning genes, '\n",
      "                                      'diseases, symptoms, anatomic parts, '\n",
      "                                      'drugs, drug effects, etc. In '\n",
      "                                      'particular, no prior work on KB '\n",
      "                                      'construction has addressed the aspects '\n",
      "                                      'of environmental and lifestyle risk '\n",
      "                                      'factors in the development of diseases '\n",
      "                                      'and the effects of drugs and '\n",
      "                                      'therapies.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'title_1'},\n",
      "                              'offset': 5392,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Background'},\n",
      "                             {'annotations': [{'id': '16706',\n",
      "                                               'infons': {'accession': None,\n",
      "                                                          'biotype': 'species',\n",
      "                                                          'database': 'ncbi_taxonomy',\n",
      "                                                          'identifier': '9606',\n",
      "                                                          'name': '9606',\n",
      "                                                          'normalized': [9606],\n",
      "                                                          'normalized_id': 9606,\n",
      "                                                          'type': 'Species',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 5,\n",
      "                                                              'offset': 6000}],\n",
      "                                               'text': 'human'}],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 5403,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The main body of IE research in '\n",
      "                                      'biomedical informatics has focused on '\n",
      "                                      'molecular entities and chemogenomics, '\n",
      "                                      'like Protein-Protein Interactions (PPI) '\n",
      "                                      'or gene-drug relations. These efforts '\n",
      "                                      'have been driven by competitions such '\n",
      "                                      'as BioNLP Shared Task (BioNLP-ST) and '\n",
      "                                      'BioCreative. These shared tasks come '\n",
      "                                      'with pre-annotated corpora as gold '\n",
      "                                      'standard, such as the GENIA corpus, the '\n",
      "                                      'multi-level event extraction (MLEE) '\n",
      "                                      'corpus, and various BioCreative '\n",
      "                                      'corpora. Efforts such as the '\n",
      "                                      'Pharmacogenetics Research Network and '\n",
      "                                      'Knowledge Base (PharmGKB), which '\n",
      "                                      'curates and disseminates knowledge '\n",
      "                                      'about the impact of human genetic '\n",
      "                                      'variations on drug responses, or the '\n",
      "                                      'Open PHACTS project, a pharmacological '\n",
      "                                      'information platform for drug '\n",
      "                                      'discovery, offer knowledge bases with '\n",
      "                                      'annotated text corpora to facilitate '\n",
      "                                      'approaches for these use cases.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 6227,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Most IE work in this line of research '\n",
      "                                      'relies on supervised learning, like '\n",
      "                                      'Support Vector Machines or '\n",
      "                                      'Probabilistic Graphical Models. The '\n",
      "                                      '2012 i2b2 challenge aimed at extracting '\n",
      "                                      'temporal relations from clinical '\n",
      "                                      'narratives. Unsupervised approaches '\n",
      "                                      'have been pursued by, to discover '\n",
      "                                      'associations between genes and diseases '\n",
      "                                      'based on the co-occurrence of entities '\n",
      "                                      'as cues for relations. To further '\n",
      "                                      'improve the quality of discovered '\n",
      "                                      'associations, crowdsourcing has also '\n",
      "                                      'been applied. Burger et al. uses Amazon '\n",
      "                                      'Mechanical Turk to validate '\n",
      "                                      'gene-mutation relations which are '\n",
      "                                      'extracted from PubMed abstracts. Aroyo '\n",
      "                                      'et al. describes a crowdsourcing '\n",
      "                                      'approach to generate gold standard '\n",
      "                                      'annotations for medical relations, '\n",
      "                                      'taking into account the disagreement '\n",
      "                                      'between crowd workers.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 6995,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Pattern-based approaches exploit text '\n",
      "                                      'patterns that connect entities. Many of '\n",
      "                                      'them manually define extraction '\n",
      "                                      'patterns. Kolarik et al. uses Hearst '\n",
      "                                      'patterns to identify terms that '\n",
      "                                      'describe various properties of drugs. '\n",
      "                                      'SemRep manually specifies extraction '\n",
      "                                      'rules obtained from dependency parse '\n",
      "                                      'trees. Outside the biomedical domain, '\n",
      "                                      'sentic patterns leverage commonsense '\n",
      "                                      'and syntactic dependencies to extract '\n",
      "                                      'sentiments from movie reviews. However, '\n",
      "                                      'while manually defined patterns yield '\n",
      "                                      'high precision, they rely on expert '\n",
      "                                      'guidance and do not scale to large and '\n",
      "                                      'potentially noisy inputs and a broader '\n",
      "                                      'scope of relations. Bootstrapping '\n",
      "                                      'approaches such as use a limited number '\n",
      "                                      'of seeds to learn extraction patterns; '\n",
      "                                      'these techniques go back to. Our method '\n",
      "                                      'follows this paradigm, but extends '\n",
      "                                      'prior work with additional statistics '\n",
      "                                      'to quantify the confidence of patterns '\n",
      "                                      'and extracted facts.'},\n",
      "                             {'annotations': [{'id': '16707',\n",
      "                                               'infons': {'accession': None,\n",
      "                                                          'biotype': 'species',\n",
      "                                                          'database': 'ncbi_taxonomy',\n",
      "                                                          'identifier': '9606',\n",
      "                                                          'name': '9606',\n",
      "                                                          'normalized': [9606],\n",
      "                                                          'normalized_id': 9606,\n",
      "                                                          'type': 'Species',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 8523}],\n",
      "                                               'text': 'people'}],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 7879,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'A small number of projects like '\n",
      "                                      'Sofie/Prospera and NELL have combined '\n",
      "                                      'pattern-based extraction with logical '\n",
      "                                      'consistency rules that constrain the '\n",
      "                                      'space of fact candidates. Nebot et al. '\n",
      "                                      'harness the IE methods of for '\n",
      "                                      'populating disease-centric relations. '\n",
      "                                      'This approach uses logical consistency '\n",
      "                                      'reasoning for high precision, but the '\n",
      "                                      'small scale of this work leads to a '\n",
      "                                      'very restricted KB. Movshovitz-Attias '\n",
      "                                      'et al. used NELL to learn instances of '\n",
      "                                      'biological classes, but did not extract '\n",
      "                                      'binary relations and did not make use '\n",
      "                                      'of constraints either. The other works '\n",
      "                                      'on constrained extraction tackle '\n",
      "                                      'non-biological relations only (e.g., '\n",
      "                                      'birthplaces of people or headquarters '\n",
      "                                      'of companies). Our method builds on '\n",
      "                                      'Sofie/Prospera, but additionally '\n",
      "                                      'develops customized constraints for the '\n",
      "                                      'biomedical relations targeted here.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 8691,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Most prior work in biomedical Named '\n",
      "                                      'Entity Recognition (NER) specializes in '\n",
      "                                      'recognizing specific types of entities '\n",
      "                                      'such as proteins and genes, chemicals, '\n",
      "                                      'diseases, and organisms. MetaMap is the '\n",
      "                                      'most notable tool capable of '\n",
      "                                      'recognizing a wide range of entities. '\n",
      "                                      'As for biomedical Named Entity '\n",
      "                                      'Disambiguation (NED), there is '\n",
      "                                      'relatively little prior work. MetaMap '\n",
      "                                      'offers limited NED functionality, while '\n",
      "                                      'others focus on disambiguating between '\n",
      "                                      'genes or small sets of word senses.'},\n",
      "                             {'annotations': [{'id': '16708',\n",
      "                                               'infons': {'biotype': 'disease',\n",
      "                                                          'normalized_id': None,\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': False},\n",
      "                                               'locations': [{'length': 2,\n",
      "                                                              'offset': 9178}],\n",
      "                                               'text': 'IE'}],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 9167,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Most prior IE work processes only '\n",
      "                                      'abstracts of Pubmed articles; few '\n",
      "                                      'projects have considered full-length '\n",
      "                                      'articles from Pubmed Central, let alone '\n",
      "                                      'Web portals and online communities. '\n",
      "                                      'Vydiswaran et al. addressed the issue '\n",
      "                                      'of assessing the credibility of medical '\n",
      "                                      'claims about diseases and their '\n",
      "                                      'treatments in health portals. Mukherjee '\n",
      "                                      'et al. tapped discussion forums to '\n",
      "                                      'assess statements about side effects of '\n",
      "                                      'drugs. White et al. demonstrated how to '\n",
      "                                      'derive insight on drug effects from '\n",
      "                                      'query logs of search engines. Building '\n",
      "                                      'a comprehensive KB from such raw assets '\n",
      "                                      'has been beyond the scope of these '\n",
      "                                      'prior works.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'title_1'},\n",
      "                              'offset': 9776,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Contributions'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 9790,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We present KnowLife, a large KB that '\n",
      "                                      'captures a wide variety of biomedical '\n",
      "                                      'knowledge, automatically extracted from '\n",
      "                                      'different genres of input sources. '\n",
      "                                      \"KnowLife's novel approach to KB \"\n",
      "                                      'construction overcomes the following '\n",
      "                                      'three limitations of prior work.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 10044,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Beyond manual curation. Using distant '\n",
      "                                      'supervision in the form of seed facts '\n",
      "                                      'from existing expert-level knowledge '\n",
      "                                      'collections, the KnowLife processing '\n",
      "                                      'pipeline is able to automatically learn '\n",
      "                                      'textual patterns and harvest a large '\n",
      "                                      'number of relational facts from such '\n",
      "                                      'patterns. In contrast to prior work on '\n",
      "                                      'IE for biomedical data which relies on '\n",
      "                                      'extraction patterns only, our method '\n",
      "                                      'achieves high precision by specifying '\n",
      "                                      'and checking logical consistency '\n",
      "                                      'constraints that fact candidates have '\n",
      "                                      'to satisfy. These constraints are '\n",
      "                                      'customized for the relations of '\n",
      "                                      'interest in KnowLife, and include '\n",
      "                                      'constraints that couple different '\n",
      "                                      'relations. The consistency constraints '\n",
      "                                      'are available as supplementary material '\n",
      "                                      '(see Additional file 1). KnowLife is '\n",
      "                                      'easily extensible, since new relations '\n",
      "                                      'can be added with little manual effort '\n",
      "                                      'and without requiring explicit '\n",
      "                                      'training; only a small number of seed '\n",
      "                                      'facts for each new relation is needed.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 10969,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Beyond scientific literature. KnowLife '\n",
      "                                      'copes with input text at large scale - '\n",
      "                                      'considering not only knowledge from '\n",
      "                                      'scientific publications, but also '\n",
      "                                      'tapping into previously neglected '\n",
      "                                      'textual sources like Web portals on '\n",
      "                                      'health issues and online communities '\n",
      "                                      'with discussion boards. We present an '\n",
      "                                      'extensive evaluation of 22,000 facts on '\n",
      "                                      'how these different genres of input '\n",
      "                                      'texts affect the resulting precision '\n",
      "                                      'and recall of the KB. We also present '\n",
      "                                      'an error analysis that provides further '\n",
      "                                      'insight on the quality and contribution '\n",
      "                                      'of different text genres.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 11522,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Beyond molecular entities. The '\n",
      "                                      'entities and facts in KnowLife go way '\n",
      "                                      'beyond the traditionally covered level '\n",
      "                                      'of proteins and genes. Besides genetic '\n",
      "                                      'factors of diseases, the KB also '\n",
      "                                      'captures diseases, therapies, drugs, '\n",
      "                                      'and risk factors like nutritional '\n",
      "                                      'habits, life-style properties, and side '\n",
      "                                      'effects of treatments.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 11837,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'In summary, the novelty of KnowLife is '\n",
      "                                      'its versatile, largely automated, and '\n",
      "                                      'scalable approach for the comprehensive '\n",
      "                                      'construction of a KB - covering a '\n",
      "                                      'spectrum of different text genres as '\n",
      "                                      'input and distilling a wide variety '\n",
      "                                      'facts from different biomedical areas '\n",
      "                                      'as output. Coupled with an entity '\n",
      "                                      'recognition module that covers the '\n",
      "                                      'entire range of biomedical entities, '\n",
      "                                      'the resulting KB features a much wider '\n",
      "                                      'spectrum of knowledge and use-cases '\n",
      "                                      'than previously built, highly '\n",
      "                                      \"specialized KB's. In terms of \"\n",
      "                                      'methodology, our extraction pipeline '\n",
      "                                      'builds on existing techniques but '\n",
      "                                      'extends them, and is specifically '\n",
      "                                      'customized to the life-science domain. '\n",
      "                                      'Most notably, unlike prior work on '\n",
      "                                      'biomedical IE, KnowLife employs logical '\n",
      "                                      'reasoning for checking consistency '\n",
      "                                      'constraints, tailored to the different '\n",
      "                                      'relations that connect diseases, '\n",
      "                                      'symptoms, drugs, genes, risk factors, '\n",
      "                                      'etc. This constraint checking '\n",
      "                                      'eliminates many false positives that '\n",
      "                                      'are produced by methods that solely '\n",
      "                                      'rely on pattern-based extraction.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'INTRO',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 12845,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'In its best configuration, the KnowLife '\n",
      "                                      'KB contains a total of 542,689 facts '\n",
      "                                      'for 13 different relations, with an '\n",
      "                                      'average precision of 93% (i.e., '\n",
      "                                      'validity of the acquired facts) as '\n",
      "                                      'determined by extensive sampling with '\n",
      "                                      'manual assessment. The precision for '\n",
      "                                      'the different relations ranges from 71% '\n",
      "                                      '(createsRisk: ecofactor x disease) to '\n",
      "                                      '97% (sideEffect:(symptom   disease) x '\n",
      "                                      'drug). All facts in KnowLife carry '\n",
      "                                      'provenance information, so that one can '\n",
      "                                      'explore the evidence for a fact and '\n",
      "                                      'filter by source. We developed a web '\n",
      "                                      'portal that showcases use-cases from '\n",
      "                                      'speed-reading to semantic search along '\n",
      "                                      'with richly annotated literature, the '\n",
      "                                      'details of which are described in the '\n",
      "                                      'demo paper.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'title_1'},\n",
      "                              'offset': 13532,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Methods'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': '12859_2015_549_Fig1_HTML.jpg',\n",
      "                                         'id': 'Fig1',\n",
      "                                         'section_type': 'FIG',\n",
      "                                         'type': 'fig_caption'},\n",
      "                              'offset': 13540,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Overview of the KnowLife KB and '\n",
      "                                      'processing pipeline.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 13593,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Our method for harvesting relational '\n",
      "                                      'facts from text sources is designed as '\n",
      "                                      'a pipeline of processing stages; Figure '\n",
      "                                      '1 gives a pictorial overview. A fact is '\n",
      "                                      'a triple consisting of two entities e '\n",
      "                                      '1,e 2 and a relation R between them; we '\n",
      "                                      'denote a fact by R(e 1,e 2). In the '\n",
      "                                      'following, we describe the input data '\n",
      "                                      'and each stage of the pipeline.  '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 13935,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Input sources'},\n",
      "                             {'annotations': [{'id': '16709',\n",
      "                                               'infons': {'accession': '@GENE_IL4R',\n",
      "                                                          'biotype': 'gene',\n",
      "                                                          'database': 'ncbi_gene',\n",
      "                                                          'identifier': '3566',\n",
      "                                                          'name': 'IL4R',\n",
      "                                                          'ncbi_homologene': '7784',\n",
      "                                                          'normalized': [3566],\n",
      "                                                          'normalized_id': 3566,\n",
      "                                                          'type': 'Gene',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 4,\n",
      "                                                              'offset': 14280}],\n",
      "                                               'text': 'IL4R'},\n",
      "                                              {'id': '16710',\n",
      "                                               'infons': {'accession': '@DISEASE_Asthma',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D001249',\n",
      "                                                          'name': 'Asthma',\n",
      "                                                          'normalized': ['D001249'],\n",
      "                                                          'normalized_id': 'D001249',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 14289}],\n",
      "                                               'text': 'asthma'}],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 13949,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Dictionary We use UMLS (Unified '\n",
      "                                      'Medical Language System) as the '\n",
      "                                      'dictionary of biomedical entities. UMLS '\n",
      "                                      'is a metathesaurus, the largest '\n",
      "                                      'collection of biomedical dictionaries '\n",
      "                                      'containing 2.9 million entities and '\n",
      "                                      '11.4 million entity names and synonyms. '\n",
      "                                      'Each entity has a semantic type '\n",
      "                                      'assigned by experts. For instance, the '\n",
      "                                      'entities IL4R and asthma are of '\n",
      "                                      'semantic types Gene or Genome and '\n",
      "                                      'Disease or Syndrome, respectively. The '\n",
      "                                      'UMLS dictionary enables KnowLife to '\n",
      "                                      'detect entities in text, going beyond '\n",
      "                                      'genes and proteins and covering '\n",
      "                                      'entities about anatomy, physiology, and '\n",
      "                                      'therapy.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab1.xml',\n",
      "                                         'id': 'Tab1',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_caption'},\n",
      "                              'offset': 14531,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' KnowLife relations, their type '\n",
      "                                      'signatures, and number of seeds '},\n",
      "                             {'annotations': [{'id': '16711',\n",
      "                                               'infons': {'accession': '@DISEASE_Drug_Fever',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D000092582',\n",
      "                                                          'name': 'Drug Fever',\n",
      "                                                          'normalized': ['D000092582'],\n",
      "                                                          'normalized_id': 'D000092582',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 15039}],\n",
      "                                               'text': 'Drug Disease'},\n",
      "                                              {'id': '16712',\n",
      "                                               'infons': {'accession': '@DISEASE_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004194',\n",
      "                                                          'name': 'Disease',\n",
      "                                                          'normalized': ['D004194'],\n",
      "                                                          'normalized_id': 'D004194',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 14979}],\n",
      "                                               'text': 'Disease'},\n",
      "                                              {'id': '16713',\n",
      "                                               'infons': {'accession': '@DISEASE_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004194',\n",
      "                                                          'name': 'Disease',\n",
      "                                                          'normalized': ['D004194'],\n",
      "                                                          'normalized_id': 'D004194',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 14639}],\n",
      "                                               'text': 'Disease'},\n",
      "                                              {'id': '16714',\n",
      "                                               'infons': {'accession': '@DISEASE_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004194',\n",
      "                                                          'name': 'Disease',\n",
      "                                                          'normalized': ['D004194'],\n",
      "                                                          'normalized_id': 'D004194',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 15,\n",
      "                                                              'offset': 14764}],\n",
      "                                               'text': 'Disease Disease'},\n",
      "                                              {'id': '16715',\n",
      "                                               'infons': {'accession': '@DISEASE_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004194',\n",
      "                                                          'name': 'Disease',\n",
      "                                                          'normalized': ['D004194'],\n",
      "                                                          'normalized_id': 'D004194',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 15014}],\n",
      "                                               'text': 'Disease'},\n",
      "                                              {'id': '16716',\n",
      "                                               'infons': {'accession': '@DISEASE_Neuromyelitis_Optica',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D009471',\n",
      "                                                          'name': 'Neuromyelitis '\n",
      "                                                                  'Optica',\n",
      "                                                          'normalized': ['D009471'],\n",
      "                                                          'normalized_id': 'D009471',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 14,\n",
      "                                                              'offset': 14864}],\n",
      "                                               'text': 'Device Disease'},\n",
      "                                              {'id': '16717',\n",
      "                                               'infons': {'accession': '@DISEASE_Desmoid_disease_hereditary',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:C535944',\n",
      "                                                          'name': 'Desmoid '\n",
      "                                                                  'disease '\n",
      "                                                                  'hereditary',\n",
      "                                                          'normalized': ['C535944'],\n",
      "                                                          'normalized_id': 'C535944',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 14721}],\n",
      "                                               'text': 'Causes'},\n",
      "                                              {'id': '16718',\n",
      "                                               'infons': {'accession': '@DISEASE_Drug_Fever',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D000092582',\n",
      "                                                          'name': 'Drug Fever',\n",
      "                                                          'normalized': ['D000092582'],\n",
      "                                                          'normalized_id': 'D000092582',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 14703}],\n",
      "                                               'text': 'Drug Disease'},\n",
      "                                              {'id': '16719',\n",
      "                                               'infons': {'accession': '@DISEASE_Signs_and_Symptoms',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D012816',\n",
      "                                                          'name': 'Signs and '\n",
      "                                                                  'Symptoms',\n",
      "                                                          'normalized': ['D012816'],\n",
      "                                                          'normalized_id': 'D012816',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 15003}],\n",
      "                                               'text': 'Symptom'},\n",
      "                                              {'id': '16720',\n",
      "                                               'infons': {'accession': '@DISEASE_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004194',\n",
      "                                                          'name': 'Disease',\n",
      "                                                          'normalized': ['D004194'],\n",
      "                                                          'normalized_id': 'D004194',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 17,\n",
      "                                                              'offset': 14669}],\n",
      "                                               'text': 'Ecofactor Disease'},\n",
      "                                              {'id': '16721',\n",
      "                                               'infons': {'accession': '@DISEASE_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004194',\n",
      "                                                          'name': 'Disease',\n",
      "                                                          'normalized': ['D004194'],\n",
      "                                                          'normalized_id': 'D004194',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 15,\n",
      "                                                              'offset': 14728}],\n",
      "                                               'text': 'Disease Disease'},\n",
      "                                              {'id': '16722',\n",
      "                                               'infons': {'accession': '@DISEASE_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004194',\n",
      "                                                          'name': 'Disease',\n",
      "                                                          'normalized': ['D004194'],\n",
      "                                                          'normalized_id': 'D004194',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 17,\n",
      "                                                              'offset': 14830}],\n",
      "                                               'text': 'Ecofactor Disease'},\n",
      "                                              {'id': '16723',\n",
      "                                               'infons': {'accession': '@DISEASE_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004194',\n",
      "                                                          'name': 'Disease',\n",
      "                                                          'normalized': ['D004194'],\n",
      "                                                          'normalized_id': 'D004194',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 15,\n",
      "                                                              'offset': 14929}],\n",
      "                                               'text': 'Disease Disease'},\n",
      "                                              {'id': '16724',\n",
      "                                               'infons': {'accession': '@DISEASE_Signs_and_Symptoms',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D012816',\n",
      "                                                          'name': 'Signs and '\n",
      "                                                                  'Symptoms',\n",
      "                                                          'normalized': ['D012816'],\n",
      "                                                          'normalized_id': 'D012816',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 17,\n",
      "                                                              'offset': 14908}],\n",
      "                                               'text': 'IsSymptom Symptom'},\n",
      "                                              {'id': '16725',\n",
      "                                               'infons': {'accession': '@DISEASE_Drug_Fever',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D000092582',\n",
      "                                                          'name': 'Drug Fever',\n",
      "                                                          'normalized': ['D000092582'],\n",
      "                                                          'normalized_id': 'D000092582',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 14800}],\n",
      "                                               'text': 'Drug Disease'}],\n",
      "                              'infons': {'file': 'Tab1.xml',\n",
      "                                         'id': 'Tab1',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table',\n",
      "                                         'xml': '<?xml version=\"1.0\" '\n",
      "                                                'encoding=\"UTF-8\"?> <table '\n",
      "                                                'frame=\"hsides\" '\n",
      "                                                'rules=\"groups\"><thead><tr><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Relation</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Domain</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Range</bold> </th><th '\n",
      "                                                'align=\"left\"> <bold>Seed '\n",
      "                                                'facts</bold> '\n",
      "                                                '</th></tr></thead><tbody><tr><td '\n",
      "                                                'align=\"left\">Affects</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">Organ</td><td '\n",
      "                                                'align=\"left\">23</td></tr><tr><td '\n",
      "                                                'align=\"left\">Aggravates</td><td '\n",
      "                                                'align=\"left\">Ecofactor</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">21</td></tr><tr><td '\n",
      "                                                'align=\"left\">Alleviates</td><td '\n",
      "                                                'align=\"left\">Drug</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">18</td></tr><tr><td '\n",
      "                                                'align=\"left\">Causes</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">70</td></tr><tr><td '\n",
      "                                                'align=\"left\">ComplicationOf</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">5</td></tr><tr><td '\n",
      "                                                'align=\"left\">Contraindicates</td><td '\n",
      "                                                'align=\"left\">Drug</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">26</td></tr><tr><td '\n",
      "                                                'align=\"left\">CreatesRisk</td><td '\n",
      "                                                'align=\"left\">Ecofactor</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">103</td></tr><tr><td '\n",
      "                                                'align=\"left\">Diagnoses</td><td '\n",
      "                                                'align=\"left\">Device</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">29</td></tr><tr><td '\n",
      "                                                'align=\"left\">Interacts</td><td '\n",
      "                                                'align=\"left\">Drug</td><td '\n",
      "                                                'align=\"left\">Drug</td><td '\n",
      "                                                'align=\"left\">9</td></tr><tr><td '\n",
      "                                                'align=\"left\">IsSymptom</td><td '\n",
      "                                                'align=\"left\">Symptom or '\n",
      "                                                'Disease</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">69</td></tr><tr><td '\n",
      "                                                'align=\"left\">ReducesRisk</td><td '\n",
      "                                                'align=\"left\">Drug or '\n",
      "                                                'Behavior</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">24</td></tr><tr><td '\n",
      "                                                'align=\"left\">SideEffect</td><td '\n",
      "                                                'align=\"left\">Symptom or '\n",
      "                                                'Disease</td><td '\n",
      "                                                'align=\"left\">Drug</td><td '\n",
      "                                                'align=\"left\">12</td></tr><tr><td '\n",
      "                                                'align=\"left\">Treats</td><td '\n",
      "                                                'align=\"left\">Drug</td><td '\n",
      "                                                'align=\"left\">Disease</td><td '\n",
      "                                                'align=\"left\">58</td></tr></tbody></table> '},\n",
      "                              'offset': 14596,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Relation Domain Range Seed facts   '\n",
      "                                      'Affects Disease Organ 23   Aggravates '\n",
      "                                      'Ecofactor Disease 21   Alleviates Drug '\n",
      "                                      'Disease 18   Causes Disease Disease '\n",
      "                                      '70   ComplicationOf Disease Disease 5   '\n",
      "                                      'Contraindicates Drug Disease 26   '\n",
      "                                      'CreatesRisk Ecofactor Disease 103   '\n",
      "                                      'Diagnoses Device Disease 29   Interacts '\n",
      "                                      'Drug Drug 9   IsSymptom Symptom or '\n",
      "                                      'Disease Disease 69   ReducesRisk Drug '\n",
      "                                      'or Behavior Disease 24   SideEffect '\n",
      "                                      'Symptom or Disease Drug 12   Treats '\n",
      "                                      'Drug Disease 58   '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 15058,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Relations KnowLife currently supports '\n",
      "                                      '13 binary relations between entities, '\n",
      "                                      'each with a type signature constraining '\n",
      "                                      'its domain and range (i.e., its left '\n",
      "                                      'and right argument types). Table 1 '\n",
      "                                      'shows that, for instance, the relation '\n",
      "                                      'affects only holds between diseases and '\n",
      "                                      'organs, but not between diseases and '\n",
      "                                      'drugs. Each type signature consists of '\n",
      "                                      'multiple fine-grained semantic types '\n",
      "                                      'defined by UMLS; specifics for all '\n",
      "                                      'relations are provided as supplementary '\n",
      "                                      'material (see Additional file 2).  '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 15550,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Seed facts. A seed fact R(e 1,e 2) for '\n",
      "                                      'relation R is a triple presumed to be '\n",
      "                                      'true based on expert statements. We '\n",
      "                                      'collected 467 seed facts (see Table 1) '\n",
      "                                      'from the medical online portal '\n",
      "                                      'uptodate.com, a highly regarded '\n",
      "                                      'clinical resource written by physician '\n",
      "                                      'authors. These seed facts are further '\n",
      "                                      'cross-checked in other sources to '\n",
      "                                      'assert their veracity. Example seed '\n",
      "                                      'facts include i s S y m p t o m(C h e s '\n",
      "                                      't P a i n,M y o c a r d i a l I n f a r '\n",
      "                                      'c t i o n) and c r e a t e s R i s k(O '\n",
      "                                      'b e s i t y,D i a b e t e s).'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab2.xml',\n",
      "                                         'id': 'Tab2',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_caption'},\n",
      "                              'offset': 16062,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': \" Overview of KnowLife's input corpus \"},\n",
      "                             {'annotations': [{'id': '16726',\n",
      "                                               'infons': {'biotype': 'species',\n",
      "                                                          'normalized_id': None,\n",
      "                                                          'type': 'Species',\n",
      "                                                          'valid': False},\n",
      "                                               'locations': [{'length': 13,\n",
      "                                                              'offset': 16456}],\n",
      "                                               'text': 'Patient.co.uk'}],\n",
      "                              'infons': {'file': 'Tab2.xml',\n",
      "                                         'id': 'Tab2',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table',\n",
      "                                         'xml': '<?xml version=\"1.0\" '\n",
      "                                                'encoding=\"UTF-8\"?> <table '\n",
      "                                                'frame=\"hsides\" '\n",
      "                                                'rules=\"groups\"><thead><tr><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Genre</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Source</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Documents</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Sentences</bold> '\n",
      "                                                '</th></tr></thead><tbody><tr><td '\n",
      "                                                'align=\"left\">Scientific '\n",
      "                                                'Publications</td><td '\n",
      "                                                'align=\"right\">PubMed '\n",
      "                                                'Medline</td><td '\n",
      "                                                'align=\"right\">580,892</td><td '\n",
      "                                                'align=\"right\">5,875,006</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"right\">PubMed '\n",
      "                                                'Central</td><td '\n",
      "                                                'align=\"right\">12,532</td><td '\n",
      "                                                'align=\"right\">2,765,580</td></tr><tr><td '\n",
      "                                                'align=\"left\">Encyclopedic '\n",
      "                                                'Articles</td><td '\n",
      "                                                'align=\"right\">Drugs.com</td><td '\n",
      "                                                'align=\"right\">31,837</td><td '\n",
      "                                                'align=\"right\">7,586,236</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"right\">Mayo '\n",
      "                                                'Clinic</td><td '\n",
      "                                                'align=\"right\">2,166</td><td '\n",
      "                                                'align=\"right\">570,325</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"right\">Medline '\n",
      "                                                'Plus</td><td '\n",
      "                                                'align=\"right\">3,076</td><td '\n",
      "                                                'align=\"right\">197,055</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"right\">RxList</td><td '\n",
      "                                                'align=\"right\">2,515</td><td '\n",
      "                                                'align=\"right\">1,102,791</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"right\">Wikipedia '\n",
      "                                                'Health</td><td '\n",
      "                                                'align=\"right\">20,893</td><td '\n",
      "                                                'align=\"right\">787,148</td></tr><tr><td '\n",
      "                                                'align=\"left\">Social '\n",
      "                                                'Sources</td><td '\n",
      "                                                'align=\"right\">Healthboards.com</td><td '\n",
      "                                                'align=\"right\">752,778</td><td '\n",
      "                                                'align=\"right\">37,270,371</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"right\">Patient.co.uk</td><td '\n",
      "                                                'align=\"right\">44,610</td><td '\n",
      "                                                'align=\"right\">1,081,420</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"right\"> '\n",
      "                                                '<bold>Total</bold> </td><td '\n",
      "                                                'align=\"right\"> '\n",
      "                                                '<bold>1,451,299</bold> '\n",
      "                                                '</td><td align=\"right\"> '\n",
      "                                                '<bold>57,235,932</bold> '\n",
      "                                                '</td></tr></tbody></table> '},\n",
      "                              'offset': 16102,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Genre Source Documents Sentences   '\n",
      "                                      'Scientific Publications PubMed Medline '\n",
      "                                      '580,892 5,875,006    PubMed Central '\n",
      "                                      '12,532 2,765,580   Encyclopedic '\n",
      "                                      'Articles Drugs.com 31,837 7,586,236    '\n",
      "                                      'Mayo Clinic 2,166 570,325    Medline '\n",
      "                                      'Plus 3,076 197,055    RxList 2,515 '\n",
      "                                      '1,102,791    Wikipedia Health 20,893 '\n",
      "                                      '787,148   Social Sources '\n",
      "                                      'Healthboards.com 752,778 37,270,371    '\n",
      "                                      'Patient.co.uk 44,610 1,081,420    Total '\n",
      "                                      '1,451,299 57,235,932   '},\n",
      "                             {'annotations': [{'id': '16727',\n",
      "                                               'infons': {'accession': None,\n",
      "                                                          'biotype': 'species',\n",
      "                                                          'database': 'ncbi_taxonomy',\n",
      "                                                          'identifier': '9606',\n",
      "                                                          'name': '9606',\n",
      "                                                          'normalized': [9606],\n",
      "                                                          'normalized_id': 9606,\n",
      "                                                          'type': 'Species',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 8,\n",
      "                                                              'offset': 17338}],\n",
      "                                               'text': 'patients'}],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 16520,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Text Corpus. A key asset of this work '\n",
      "                                      'is that we tap into different genres of '\n",
      "                                      'text; Table 2 gives an overview. PubMed '\n",
      "                                      'documents are scientific texts with '\n",
      "                                      'specialized jargon; they have been the '\n",
      "                                      'de-facto standard corpus for biomedical '\n",
      "                                      'text mining. We took all PubMed '\n",
      "                                      'documents published in 2011 that are '\n",
      "                                      'indexed with disease-, drug-, and '\n",
      "                                      'therapy-related MeSH (Medical Subject '\n",
      "                                      'Heading) terms. We further prune out '\n",
      "                                      'documents from inapplicable journals '\n",
      "                                      'such as those not in the English '\n",
      "                                      'language, or those about medical '\n",
      "                                      'ethics. Web portals and encyclopedic '\n",
      "                                      'articles are collaboratively or '\n",
      "                                      'professionally edited, providing '\n",
      "                                      'credible information in layman-oriented '\n",
      "                                      'language. Examples include '\n",
      "                                      'uptodate.com, mayoclinic.com, and the '\n",
      "                                      'relevant parts of en.wikipedia.org. In '\n",
      "                                      'contrast, discussion forums of online '\n",
      "                                      'communities, where patients and '\n",
      "                                      'physicians engage in discussions (often '\n",
      "                                      'anonymously), have a colloquial '\n",
      "                                      'language style, sometimes even slang. '\n",
      "                                      'We tap into all three genres of text to '\n",
      "                                      'demonstrate not only the applicability '\n",
      "                                      'of our system, but also the amount of '\n",
      "                                      'information buried in all of them. We '\n",
      "                                      'use the Stanford CoreNLP software to '\n",
      "                                      'preprocess all texts, such that they '\n",
      "                                      'are tokenized, split into sentences, '\n",
      "                                      'tagged with parts-of-speech, '\n",
      "                                      'lemmatized, and parsed into syntactic '\n",
      "                                      'dependency graphs.  '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 17815,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Entity recognition'},\n",
      "                             {'annotations': [{'id': '16728',\n",
      "                                               'infons': {'accession': '@DISEASE_Anemia',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D000740',\n",
      "                                                          'name': 'Anemia',\n",
      "                                                          'normalized': ['D000740'],\n",
      "                                                          'normalized_id': 'D000740',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 17835}],\n",
      "                                               'text': 'Anemia'},\n",
      "                                              {'id': '16729',\n",
      "                                               'infons': {'accession': '@DISEASE_Sarcoidosis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D012507',\n",
      "                                                          'name': 'Sarcoidosis',\n",
      "                                                          'normalized': ['D012507'],\n",
      "                                                          'normalized_id': 'D012507',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 11,\n",
      "                                                              'offset': 17865}],\n",
      "                                               'text': 'sarcoidosis'}],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 17834,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Anemia is a common symptom of '\n",
      "                                      'sarcoidosis.'},\n",
      "                             {'annotations': [{'id': '16730',\n",
      "                                               'infons': {'accession': '@DISEASE_Arrhythmias_Cardiac',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D001145',\n",
      "                                                          'name': 'Arrhythmias '\n",
      "                                                                  'Cardiac',\n",
      "                                                          'normalized': ['D001145'],\n",
      "                                                          'normalized_id': 'D001145',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 10,\n",
      "                                                              'offset': 17914}],\n",
      "                                               'text': 'arrythmias'},\n",
      "                                              {'id': '16731',\n",
      "                                               'infons': {'accession': '@DISEASE_Myocardial_Infarction',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D009203',\n",
      "                                                          'name': 'Myocardial '\n",
      "                                                                  'Infarction',\n",
      "                                                          'normalized': ['D009203'],\n",
      "                                                          'normalized_id': 'D009203',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 17892}],\n",
      "                                               'text': 'heart attack'}],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 17878,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Eventually, a heart attack leads to '\n",
      "                                      'arrythmias.'},\n",
      "                             {'annotations': [{'id': '16732',\n",
      "                                               'infons': {'accession': '@DISEASE_Pericarditis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D010493',\n",
      "                                                          'name': 'Pericarditis',\n",
      "                                                          'normalized': ['D010493'],\n",
      "                                                          'normalized_id': 'D010493',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 17979}],\n",
      "                                               'text': 'pericarditis'},\n",
      "                                              {'id': '16733',\n",
      "                                               'infons': {'accession': '@DISEASE_Myocardial_Infarction',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D009203',\n",
      "                                                          'name': 'Myocardial '\n",
      "                                                                  'Infarction',\n",
      "                                                          'normalized': ['D009203'],\n",
      "                                                          'normalized_id': 'D009203',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 21,\n",
      "                                                              'offset': 17940}],\n",
      "                                               'text': 'myocardial '\n",
      "                                                       'infarction'}],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 17926,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Ironically, a myocardial infarction can '\n",
      "                                      'also lead to pericarditis.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 17993,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The first stage in the KnowLife '\n",
      "                                      'pipeline identifies sentences that may '\n",
      "                                      'express a relational fact. We apply '\n",
      "                                      'entity recognition to every sentence: a '\n",
      "                                      'sentence with one or more entities is '\n",
      "                                      'relevant for further processing. To '\n",
      "                                      'efficiently handle the large dictionary '\n",
      "                                      'and process large input corpora, we '\n",
      "                                      'employ our own method, using '\n",
      "                                      'string-similarity matching against the '\n",
      "                                      'names in the UMLS dictionary. This '\n",
      "                                      'method is two orders of magnitude '\n",
      "                                      'faster than MetaMap, the most popular '\n",
      "                                      'biomedical entity recognition tool, '\n",
      "                                      'while maintaining comparable accuracy. '\n",
      "                                      'Specifically, we use locality sensitive '\n",
      "                                      'hashing (LSH) with min-wise independent '\n",
      "                                      'permutations (MinHash) to quickly find '\n",
      "                                      'matching candidates. LSH '\n",
      "                                      'probabilistically reduces the '\n",
      "                                      'high-dimensional space of all '\n",
      "                                      'character-level 3-grams, while MinHash '\n",
      "                                      'quickly estimates the similarity '\n",
      "                                      'between two sets of 3-grams. A '\n",
      "                                      'successful match provides us also with '\n",
      "                                      \"the entity's semantic type. If multiple \"\n",
      "                                      'entities are matched to the same string '\n",
      "                                      'in the input text, we currently do not '\n",
      "                                      'apply explicit NED to determine the '\n",
      "                                      'correct entity. Instead, using the '\n",
      "                                      'semantic type hierarchy of UMLS, we '\n",
      "                                      'select the most specifically typed '\n",
      "                                      'entities. Later in the consistency '\n",
      "                                      'reasoning stage, we leverage the type '\n",
      "                                      'signatures to futher prune out '\n",
      "                                      'mismatching entities. At the end of '\n",
      "                                      'this processing stage, we have '\n",
      "                                      'marked-up sentences such as  '},\n",
      "                             {'annotations': [{'id': '16734',\n",
      "                                               'infons': {'accession': '@DISEASE_Myocardial_Infarction',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D009203',\n",
      "                                                          'name': 'Myocardial '\n",
      "                                                                  'Infarction',\n",
      "                                                          'normalized': ['D009203'],\n",
      "                                                          'normalized_id': 'D009203',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 21,\n",
      "                                                              'offset': 19356}],\n",
      "                                               'text': 'myocardial infarction'},\n",
      "                                              {'id': '16735',\n",
      "                                               'infons': {'accession': '@DISEASE_Myocardial_Infarction',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D009203',\n",
      "                                                          'name': 'Myocardial '\n",
      "                                                                  'Infarction',\n",
      "                                                          'normalized': ['D009203'],\n",
      "                                                          'normalized_id': 'D009203',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 19382}],\n",
      "                                               'text': 'heart attack'}],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 19350,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'where myocardial infarction and heart '\n",
      "                                      'attack are synonyms representing the '\n",
      "                                      'same canonical entity.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 19448,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Pattern gathering'},\n",
      "                             {'annotations': [{'id': '16736',\n",
      "                                               'infons': {'accession': '@DISEASE_Sarcoidosis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D012507',\n",
      "                                                          'name': 'Sarcoidosis',\n",
      "                                                          'normalized': ['D012507'],\n",
      "                                                          'normalized_id': 'D012507',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 11,\n",
      "                                                              'offset': 19591}],\n",
      "                                               'text': 'sarcoidosis'},\n",
      "                                              {'id': '16737',\n",
      "                                               'infons': {'accession': '@CHEMICAL_Diclofenac',\n",
      "                                                          'biotype': 'chemical',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004008',\n",
      "                                                          'name': 'Diclofenac',\n",
      "                                                          'normalized': ['D004008'],\n",
      "                                                          'normalized_id': 'D004008',\n",
      "                                                          'type': 'Chemical',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 10,\n",
      "                                                              'offset': 19877}],\n",
      "                                               'text': 'Diclofenac'},\n",
      "                                              {'id': '16738',\n",
      "                                               'infons': {'accession': '@CHEMICAL_Diclofenac',\n",
      "                                                          'biotype': 'chemical',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004008',\n",
      "                                                          'name': 'Diclofenac',\n",
      "                                                          'normalized': ['D004008'],\n",
      "                                                          'normalized_id': 'D004008',\n",
      "                                                          'type': 'Chemical',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 10,\n",
      "                                                              'offset': 19965}],\n",
      "                                               'text': 'Diclofenac'},\n",
      "                                              {'id': '16739',\n",
      "                                               'infons': {'accession': '@DISEASE_Anemia',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D000740',\n",
      "                                                          'name': 'Anemia',\n",
      "                                                          'normalized': ['D000740'],\n",
      "                                                          'normalized_id': 'D000740',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 19580}],\n",
      "                                               'text': 'anemia'}],\n",
      "                              'infons': {'file': '12859_2015_549_Fig2_HTML.jpg',\n",
      "                                         'id': 'Fig2',\n",
      "                                         'section_type': 'FIG',\n",
      "                                         'type': 'fig_caption'},\n",
      "                              'offset': 19466,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Pattern gathering in KnowLife.(a) '\n",
      "                                      'Sentence-level pattern: Dependency '\n",
      "                                      'graph of a sentence with recognized '\n",
      "                                      'entities anemia and sarcoidosis. By '\n",
      "                                      'computing the shortest path (bold '\n",
      "                                      'lines) between the two entities, the '\n",
      "                                      'word sequence symptom of is extracted. '\n",
      "                                      'This sequence is extended by an '\n",
      "                                      'adjectival modifier (amod) which '\n",
      "                                      'results in the extracted pattern common '\n",
      "                                      'symptom of. (b) Document-structure '\n",
      "                                      'pattern: The entity Diclofenac is found '\n",
      "                                      'within the document title and Belching '\n",
      "                                      'within an <li> element. Take Diclofenac '\n",
      "                                      'as the left-hand entity. By traversing '\n",
      "                                      'the DOM tree downwards and coming '\n",
      "                                      'across the heading Side Effects, we '\n",
      "                                      \"extract the heading's text as a \"\n",
      "                                      'pattern. Further traversal leads us to '\n",
      "                                      'Belching, which yields the right-hand '\n",
      "                                      'entity for the pattern.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 20220,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Our method extracts textual patterns '\n",
      "                                      'that connect two recognized entities, '\n",
      "                                      'either by the syntactic structure of a '\n",
      "                                      'sentence or by a path in the DOM '\n",
      "                                      '(Document Object Model) tree of a Web '\n",
      "                                      'page. We extract two types of patterns: '\n",
      "                                      'Sentence-level Patterns: For each pair '\n",
      "                                      'of entities in a sentence, we extract a '\n",
      "                                      'sequence of text tokens connecting the '\n",
      "                                      'entities in the syntactic structure of '\n",
      "                                      'the sentence. Specifically, this is the '\n",
      "                                      'shortest path between the entities in '\n",
      "                                      'the dependency graph obtained from '\n",
      "                                      'parsing the sentence. However, this '\n",
      "                                      'path does not necessarily contain the '\n",
      "                                      'full information to deduce a relation; '\n",
      "                                      'for instance, negations are not '\n",
      "                                      'captured or essential adjectives are '\n",
      "                                      'left out. Therefore, for every captured '\n",
      "                                      'word the following grammatical '\n",
      "                                      'dependencies are added: negation, '\n",
      "                                      'adjectival modifiers, and adverbial '\n",
      "                                      'modifiers. The resulting word sequence '\n",
      "                                      'constitutes a sentence-level pattern. '\n",
      "                                      'An example is shown in Figure 2(a). '\n",
      "                                      'Document-structure Patterns: In Web '\n",
      "                                      'portals like Mayo Clinic or Wikipedia, '\n",
      "                                      'it is common that authors state medical '\n",
      "                                      'facts by using specific document '\n",
      "                                      'structures, like titles, sections, and '\n",
      "                                      'listings. Such structures are encoded '\n",
      "                                      'in the DOM tree of the underlying HTML '\n",
      "                                      'markup. First, we detect if the '\n",
      "                                      'document title, that is, the text '\n",
      "                                      'within the <h1> tag in terms of HTML '\n",
      "                                      'markup, is a single entity. Next, we '\n",
      "                                      'detect if an entity appears in an HTML '\n",
      "                                      'listing, that is, within an <li> tag. '\n",
      "                                      'Starting from the <h1> tag, our method '\n",
      "                                      'traverses the DOM tree downwards and '\n",
      "                                      'determines all intermediate headings, '\n",
      "                                      'i.e. <h2> to <h6> tags, until we reach '\n",
      "                                      'the aforementioned <li> tag. The '\n",
      "                                      'document title serves as left-hand '\n",
      "                                      'entity, the intermediate headings as '\n",
      "                                      'patterns, and the <li> text as '\n",
      "                                      'right-hand entity. These are candidates '\n",
      "                                      'for a relation or an entity argument in '\n",
      "                                      'a relational fact. Figure 2(b) shows an '\n",
      "                                      'example.  '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 22052,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Pattern analysis'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 22069,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The goal of the pattern analysis is to '\n",
      "                                      'identify the most useful seed patterns '\n",
      "                                      'out of all the pattern candidates '\n",
      "                                      'gathered thus far. A seed pattern '\n",
      "                                      'should generalize the over-specific '\n",
      "                                      'phrases encountered in the input texts, '\n",
      "                                      'by containing only the crucial words '\n",
      "                                      'that express a relation and masking out '\n",
      "                                      '(by a wildcard or part-of-speech tag) '\n",
      "                                      'inessential words. This way we arrive '\n",
      "                                      'at high-confidence patterns.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 22473,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We harness the techniques developed in '\n",
      "                                      'the Prospera tool. First, an itemset '\n",
      "                                      'mining algorithm is applied to find '\n",
      "                                      'frequent sub-sequences in the patterns. '\n",
      "                                      'The sub-sequences are weighed by '\n",
      "                                      'statistical analysis, in terms of '\n",
      "                                      'confidence and support. We use the seed '\n",
      "                                      'facts and their co-occurrences with '\n",
      "                                      'certain patterns as a basis to compute '\n",
      "                                      'confidence, such that the confidence '\n",
      "                                      'for a pattern q in a set of sentences S '\n",
      "                                      'is defined as  '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 22900,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'where S X(R i) is the set of all entity '\n",
      "                                      'tuples (e 1,e 2) appearing in any seed '\n",
      "                                      'fact with relation R i and C X(R i) is '\n",
      "                                      'the set of all entity tuples (e 1,e 2) '\n",
      "                                      'appearing in any seed fact without '\n",
      "                                      'relation R i. The rationale is that the '\n",
      "                                      'more strongly a pattern correlates with '\n",
      "                                      'the seed-fact entities of a particular '\n",
      "                                      'relation, the more confident we are '\n",
      "                                      'that the pattern expresses the '\n",
      "                                      'relation. The patterns with confidence '\n",
      "                                      'greater than a threshold (set to 0.3 in '\n",
      "                                      'our experiments) are selected as seed '\n",
      "                                      'patterns.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab3.xml',\n",
      "                                         'id': 'Tab3',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_caption'},\n",
      "                              'offset': 23405,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Examples of seed facts and seed '\n",
      "                                      'patterns as well as automatically '\n",
      "                                      'acquired patterns and facts '},\n",
      "                             {'annotations': [{'id': '16740',\n",
      "                                               'infons': {'accession': '@DISEASE_Warts',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D014860',\n",
      "                                                          'name': 'Warts',\n",
      "                                                          'normalized': ['D014860'],\n",
      "                                                          'normalized_id': 'D014860',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 4,\n",
      "                                                              'offset': 23762}],\n",
      "                                               'text': 'Wart'},\n",
      "                                              {'id': '16741',\n",
      "                                               'infons': {'accession': '@DISEASE_Pericarditis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D010493',\n",
      "                                                          'name': 'Pericarditis',\n",
      "                                                          'normalized': ['D010493'],\n",
      "                                                          'normalized_id': 'D010493',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 24142}],\n",
      "                                               'text': 'Pericarditis'},\n",
      "                                              {'id': '16742',\n",
      "                                               'infons': {'accession': '@DISEASE_Hashimoto_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D050031',\n",
      "                                                          'name': 'Hashimoto '\n",
      "                                                                  'Disease',\n",
      "                                                          'normalized': ['D050031'],\n",
      "                                                          'normalized_id': 'D050031',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 11,\n",
      "                                                              'offset': 24106}],\n",
      "                                               'text': \"Hashimoto's\"},\n",
      "                                              {'id': '16743',\n",
      "                                               'infons': {'accession': '@DISEASE_Anemia',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D000740',\n",
      "                                                          'name': 'Anemia',\n",
      "                                                          'normalized': ['D000740'],\n",
      "                                                          'normalized_id': 'D000740',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 24204}],\n",
      "                                               'text': 'Anemia'},\n",
      "                                              {'id': '16744',\n",
      "                                               'infons': {'accession': '@DISEASE_Obesity',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D009765',\n",
      "                                                          'name': 'Obesity',\n",
      "                                                          'normalized': ['D009765'],\n",
      "                                                          'normalized_id': 'D009765',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 23796}],\n",
      "                                               'text': 'Obesity'},\n",
      "                                              {'id': '16745',\n",
      "                                               'infons': {'accession': '@DISEASE_Stillbirth',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D050497',\n",
      "                                                          'name': 'Stillbirth',\n",
      "                                                          'normalized': ['D050497'],\n",
      "                                                          'normalized_id': 'D050497',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 10,\n",
      "                                                              'offset': 23913}],\n",
      "                                               'text': 'Stillbirth'},\n",
      "                                              {'id': '16746',\n",
      "                                               'infons': {'accession': '@DISEASE_Heart_Diseases',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D006331',\n",
      "                                                          'name': 'Heart '\n",
      "                                                                  'Diseases',\n",
      "                                                          'normalized': ['D006331'],\n",
      "                                                          'normalized_id': 'D006331',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 5,\n",
      "                                                              'offset': 24155}],\n",
      "                                               'text': 'Heart'},\n",
      "                                              {'id': '16747',\n",
      "                                               'infons': {'accession': '@DISEASE_Chemical_and_Drug_Induced_Liver_Injury',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D056486',\n",
      "                                                          'name': 'Chemical '\n",
      "                                                                  'and Drug '\n",
      "                                                                  'Induced '\n",
      "                                                                  'Liver '\n",
      "                                                                  'Injury',\n",
      "                                                          'normalized': ['D056486'],\n",
      "                                                          'normalized_id': 'D056486',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 9,\n",
      "                                                              'offset': 24021}],\n",
      "                                               'text': 'Hepatitis'},\n",
      "                                              {'id': '16748',\n",
      "                                               'infons': {'biotype': 'disease',\n",
      "                                                          'normalized_id': None,\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': False},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 24049}],\n",
      "                                               'text': 'CrohnDisease'},\n",
      "                                              {'id': '16749',\n",
      "                                               'infons': {'accession': '@DISEASE_Pericarditis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D010493',\n",
      "                                                          'name': 'Pericarditis',\n",
      "                                                          'normalized': ['D010493'],\n",
      "                                                          'normalized_id': 'D010493',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 23661}],\n",
      "                                               'text': 'Pericarditis'},\n",
      "                                              {'id': '16750',\n",
      "                                               'infons': {'accession': '@DISEASE_Tendinopathy',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D052256',\n",
      "                                                          'name': 'Tendinopathy',\n",
      "                                                          'normalized': ['D052256'],\n",
      "                                                          'normalized_id': 'D052256',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 10,\n",
      "                                                              'offset': 23963}],\n",
      "                                               'text': 'Tendinitis'},\n",
      "                                              {'id': '16751',\n",
      "                                               'infons': {'accession': None,\n",
      "                                                          'biotype': 'species',\n",
      "                                                          'database': 'ncbi_taxonomy',\n",
      "                                                          'identifier': '9606',\n",
      "                                                          'name': '9606',\n",
      "                                                          'normalized': [9606],\n",
      "                                                          'normalized_id': 9606,\n",
      "                                                          'type': 'Species',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 8,\n",
      "                                                              'offset': 24185}],\n",
      "                                               'text': 'patients'},\n",
      "                                              {'id': '16752',\n",
      "                                               'infons': {'accession': '@DISEASE_Diabetes_Mellitus',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D003920',\n",
      "                                                          'name': 'Diabetes '\n",
      "                                                                  'Mellitus',\n",
      "                                                          'normalized': ['D003920'],\n",
      "                                                          'normalized_id': 'D003920',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 8,\n",
      "                                                              'offset': 23707}],\n",
      "                                               'text': 'Diabetes'},\n",
      "                                              {'id': '16753',\n",
      "                                               'infons': {'accession': '@DISEASE_Sarcoidosis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D012507',\n",
      "                                                          'name': 'Sarcoidosis',\n",
      "                                                          'normalized': ['D012507'],\n",
      "                                                          'normalized_id': 'D012507',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 11,\n",
      "                                                              'offset': 24211}],\n",
      "                                               'text': 'Sarcoidosis'},\n",
      "                                              {'id': '16754',\n",
      "                                               'infons': {'accession': '@DISEASE_Pain',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D010146',\n",
      "                                                          'name': 'Pain',\n",
      "                                                          'normalized': ['D010146'],\n",
      "                                                          'normalized_id': 'D010146',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 4,\n",
      "                                                              'offset': 24044}],\n",
      "                                               'text': 'Pain'},\n",
      "                                              {'id': '16755',\n",
      "                                               'infons': {'biotype': 'disease',\n",
      "                                                          'normalized_id': None,\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': False},\n",
      "                                               'locations': [{'length': 13,\n",
      "                                                              'offset': 23767}],\n",
      "                                               'text': 'Skincarcinoma'},\n",
      "                                              {'id': '16756',\n",
      "                                               'infons': {'accession': '@DISEASE_Pericarditis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D010493',\n",
      "                                                          'name': 'Pericarditis',\n",
      "                                                          'normalized': ['D010493'],\n",
      "                                                          'normalized_id': 'D010493',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 23595}],\n",
      "                                               'text': 'Pericarditis'},\n",
      "                                              {'id': '16757',\n",
      "                                               'infons': {'accession': '@DISEASE_Asthma',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D001249',\n",
      "                                                          'name': 'Asthma',\n",
      "                                                          'normalized': ['D001249'],\n",
      "                                                          'normalized_id': 'D001249',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 23883}],\n",
      "                                               'text': 'Asthma'},\n",
      "                                              {'id': '16758',\n",
      "                                               'infons': {'accession': '@DISEASE_Tuberculosis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D014376',\n",
      "                                                          'name': 'Tuberculosis',\n",
      "                                                          'normalized': ['D014376'],\n",
      "                                                          'normalized_id': 'D014376',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 23582}],\n",
      "                                               'text': 'Tuberculosis'},\n",
      "                                              {'id': '16759',\n",
      "                                               'infons': {'accession': '@DISEASE_Malaria',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D008288',\n",
      "                                                          'name': 'Malaria',\n",
      "                                                          'normalized': ['D008288'],\n",
      "                                                          'normalized_id': 'D008288',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 23905}],\n",
      "                                               'text': 'Malaria'},\n",
      "                                              {'id': '16760',\n",
      "                                               'infons': {'accession': '@DISEASE_Asthma',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D001249',\n",
      "                                                          'name': 'Asthma',\n",
      "                                                          'normalized': ['D001249'],\n",
      "                                                          'normalized_id': 'D001249',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 23804}],\n",
      "                                               'text': 'Asthma'},\n",
      "                                              {'id': '16761',\n",
      "                                               'infons': {'accession': '@DISEASE_Obesity',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D009765',\n",
      "                                                          'name': 'Obesity',\n",
      "                                                          'normalized': ['D009765'],\n",
      "                                                          'normalized_id': 'D009765',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 23699}],\n",
      "                                               'text': 'Obesity'}],\n",
      "                              'infons': {'file': 'Tab3.xml',\n",
      "                                         'id': 'Tab3',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table',\n",
      "                                         'xml': '<?xml version=\"1.0\" '\n",
      "                                                'encoding=\"UTF-8\"?> <table '\n",
      "                                                'frame=\"hsides\" '\n",
      "                                                'rules=\"groups\"><thead><tr><th '\n",
      "                                                'align=\"left\"> <bold>Seed '\n",
      "                                                'facts</bold> </th><th '\n",
      "                                                'align=\"left\"> <bold>Seed '\n",
      "                                                'patterns</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Relations</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Confidences</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Patterns</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Harvested facts</bold> '\n",
      "                                                '</th></tr></thead><tbody><tr><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>u</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic>(<italic>T</italic> '\n",
      "                                                '<italic>u</italic> '\n",
      "                                                '<italic>b</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>u</italic> '\n",
      "                                                '<italic>l</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic>,<italic>P</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>d</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic>)</td><td '\n",
      "                                                'align=\"left\">progress</td><td '\n",
      "                                                'align=\"left\">createsRisk</td><td '\n",
      "                                                'align=\"left\">0.5</td><td '\n",
      "                                                'align=\"left\">which progresses '\n",
      "                                                'to</td><td align=\"left\"> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>u</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic>(<italic>P</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>d</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic>,<italic>T</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>m</italic> '\n",
      "                                                '<italic>p</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>n</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>d</italic> '\n",
      "                                                '<italic>e</italic>)</td></tr><tr><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>R</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>k</italic>(<italic>O</italic> '\n",
      "                                                '<italic>b</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>y</italic>,<italic>D</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>b</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic>)</td><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\">causes</td><td '\n",
      "                                                'align=\"left\">0.5</td><td '\n",
      "                                                'align=\"left\">still '\n",
      "                                                'progressing to</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>R</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>k</italic>(<italic>W</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>t</italic>,<italic>S</italic> '\n",
      "                                                '<italic>k</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>n</italic> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>n</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>m</italic> '\n",
      "                                                '<italic>a</italic>)</td></tr><tr><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>R</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>k</italic>(<italic>O</italic> '\n",
      "                                                '<italic>b</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>y</italic>,<italic>A</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>h</italic> '\n",
      "                                                '<italic>m</italic> '\n",
      "                                                '<italic>a</italic>)</td><td '\n",
      "                                                'align=\"left\">risk '\n",
      "                                                'factor</td><td '\n",
      "                                                'align=\"left\">createsRisk</td><td '\n",
      "                                                'align=\"left\">1.0</td><td '\n",
      "                                                'align=\"left\">children risk '\n",
      "                                                'factors</td><td align=\"left\"> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>R</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>k</italic>(<italic>W</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>d</italic> '\n",
      "                                                '<italic>D</italic> '\n",
      "                                                '<italic>u</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>t</italic>,<italic>A</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>h</italic> '\n",
      "                                                '<italic>m</italic> '\n",
      "                                                '<italic>a</italic>)</td></tr><tr><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>R</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>k</italic>(<italic>M</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>l</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>a</italic>,<italic>S</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>l</italic> '\n",
      "                                                '<italic>l</italic> '\n",
      "                                                '<italic>b</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>h</italic>)</td><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\">have risk '\n",
      "                                                'factors</td><td align=\"left\"> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>R</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>k</italic>(<italic>G</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>l</italic> '\n",
      "                                                '<italic>f</italic>,<italic>T</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>n</italic> '\n",
      "                                                '<italic>d</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>n</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic>)</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\">known risk '\n",
      "                                                'factors</td><td align=\"left\"> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>R</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>k</italic>(<italic>G</italic> '\n",
      "                                                '<italic>B</italic> '\n",
      "                                                '<italic>v</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>u</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>C</italic>,<italic>H</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>p</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic>)</td></tr><tr><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>S</italic> '\n",
      "                                                '<italic>y</italic> '\n",
      "                                                '<italic>m</italic> '\n",
      "                                                '<italic>p</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>m</italic>(<italic>P</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>n</italic>,<italic>C</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>h</italic> '\n",
      "                                                '<italic>n</italic> '\n",
      "                                                '<italic>D</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>e</italic>)</td><td '\n",
      "                                                'align=\"left\">occur</td><td '\n",
      "                                                'align=\"left\">affects</td><td '\n",
      "                                                'align=\"left\">0.67</td><td '\n",
      "                                                'align=\"left\">occurs '\n",
      "                                                'anywhere</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>f</italic> '\n",
      "                                                '<italic>f</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>s</italic>(<italic>H</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>h</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>m</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                \"<sup>'</sup> \"\n",
      "                                                '<italic>s</italic>,<italic>T</italic> '\n",
      "                                                '<italic>h</italic> '\n",
      "                                                '<italic>y</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>d</italic> '\n",
      "                                                '<italic>G</italic> '\n",
      "                                                '<italic>l</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>n</italic> '\n",
      "                                                '<italic>d</italic>)</td></tr><tr><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>f</italic> '\n",
      "                                                '<italic>f</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>s</italic>(<italic>P</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>d</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic>,<italic>H</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>t</italic>)</td><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\">isSymptom</td><td '\n",
      "                                                'align=\"left\">0.33</td><td '\n",
      "                                                'align=\"left\">occurs '\n",
      "                                                'patients</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>S</italic> '\n",
      "                                                '<italic>y</italic> '\n",
      "                                                '<italic>m</italic> '\n",
      "                                                '<italic>p</italic> '\n",
      "                                                '<italic>t</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>m</italic>(<italic>A</italic> '\n",
      "                                                '<italic>n</italic> '\n",
      "                                                '<italic>e</italic> '\n",
      "                                                '<italic>m</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>a</italic>,<italic>S</italic> '\n",
      "                                                '<italic>a</italic> '\n",
      "                                                '<italic>r</italic> '\n",
      "                                                '<italic>c</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>d</italic> '\n",
      "                                                '<italic>o</italic> '\n",
      "                                                '<italic>s</italic> '\n",
      "                                                '<italic>i</italic> '\n",
      "                                                '<italic>s</italic>)</td></tr></tbody></table> '},\n",
      "                              'offset': 23501,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Seed facts Seed patterns Relations '\n",
      "                                      'Confidences Patterns Harvested facts   '\n",
      "                                      'causes(Tuberculosis,Pericarditis) '\n",
      "                                      'progress createsRisk 0.5 which '\n",
      "                                      'progresses to '\n",
      "                                      'causes(Pericarditis,Tamponade)   '\n",
      "                                      'createsRisk(Obesity,Diabetes)  causes '\n",
      "                                      '0.5 still progressing to '\n",
      "                                      'createsRisk(Wart,Skincarcinoma)   '\n",
      "                                      'createsRisk(Obesity,Asthma) risk factor '\n",
      "                                      'createsRisk 1.0 children risk factors '\n",
      "                                      'createsRisk(WoodDust,Asthma)   '\n",
      "                                      'createsRisk(Malaria,Stillbirth)    have '\n",
      "                                      'risk factors '\n",
      "                                      'createsRisk(Golf,Tendinitis)       '\n",
      "                                      'known risk factors '\n",
      "                                      'createsRisk(GBvirusC,Hepatitis)   '\n",
      "                                      'isSymptom(Pain,CrohnDisease) occur '\n",
      "                                      'affects 0.67 occurs anywhere '\n",
      "                                      \"affects(Hashimoto's,ThyroidGland)   \"\n",
      "                                      'affects(Pericarditis,Heart)  isSymptom '\n",
      "                                      '0.33 occurs patients '\n",
      "                                      'isSymptom(Anemia,Sarcoidosis)   '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 24229,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Each non-seed pattern p is then matched '\n",
      "                                      'against the seed pattern set Q using '\n",
      "                                      'Jaccard similarity to compute a weight '\n",
      "                                      'w associating p with a relation.  The '\n",
      "                                      'pattern occurrences together with their '\n",
      "                                      'weights and relations serve as fact '\n",
      "                                      'candidates. Table 3 shows sample seed '\n",
      "                                      'patterns computed from seed facts. The '\n",
      "                                      'table also gives examples for '\n",
      "                                      'automatically acquired patterns and '\n",
      "                                      'facts.  '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 24611,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Consistency reasoning'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 24633,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The pattern analysis stage provides us '\n",
      "                                      'with a large set of fact candidates and '\n",
      "                                      'their supporting patterns. However, '\n",
      "                                      'these contain many false positives. To '\n",
      "                                      'prune these out and improve precision, '\n",
      "                                      'the last stage of KnowLife applies '\n",
      "                                      'logical consistency constraints to the '\n",
      "                                      'fact candidates and accepts only a '\n",
      "                                      'consistent subset of them.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 24962,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We leverage two kinds of manually '\n",
      "                                      'defined semantic constraints: i) the '\n",
      "                                      'type signatures of relations (see Table '\n",
      "                                      '1) for type checking of fact '\n",
      "                                      'candidates, and ii) mutual exclusion '\n",
      "                                      'constraints between certain pairs of '\n",
      "                                      'relations. For example, if a drug has a '\n",
      "                                      'certain symptom as a side effect, it '\n",
      "                                      'cannot treat this symptom at the same '\n",
      "                                      'time. These rules allow us to handle '\n",
      "                                      'conflicting candidate facts. The '\n",
      "                                      'reasoning uses probabilistic weights '\n",
      "                                      'derived from the statistics of the '\n",
      "                                      'candidate gathering phase.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'METHODS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 25460,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'To reason with consistency constraints, '\n",
      "                                      'we follow the framework of, by encoding '\n",
      "                                      'all facts, patterns, and grounded '\n",
      "                                      '(i.e., instantiated) constraints into '\n",
      "                                      'weighted logical clauses. We extend '\n",
      "                                      'this prior work by computing '\n",
      "                                      'informative weights from the confidence '\n",
      "                                      'statistics obtained in the '\n",
      "                                      'pattern-based stage of our IE pipeline. '\n",
      "                                      'We then use a weighted Max-Sat solver '\n",
      "                                      'to reason on the hypotheses space of '\n",
      "                                      'fact candidates, to compute a '\n",
      "                                      'consistent subset of clauses with the '\n",
      "                                      'largest total weight. Due to the '\n",
      "                                      'NP-hardness of the weighted Max-Sat '\n",
      "                                      'problem, we resort to an approximation '\n",
      "                                      'algorithm that combines the '\n",
      "                                      'dominating-unit-clause technique with '\n",
      "                                      \"Johnson's heuristic algorithm. Suchanek \"\n",
      "                                      'et al. has shown that this combination '\n",
      "                                      'empirically gives very good '\n",
      "                                      'approximation ratios. The complete set '\n",
      "                                      'of consistency constraints is in the '\n",
      "                                      'supplementary material (see Additional '\n",
      "                                      'file 1).'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'title_1'},\n",
      "                              'offset': 26334,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Results and discussion'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab4.xml',\n",
      "                                         'id': 'Tab4',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_caption'},\n",
      "                              'offset': 26357,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Evaluation of different text genres '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab4.xml',\n",
      "                                         'id': 'Tab4',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table',\n",
      "                                         'xml': '<?xml version=\"1.0\" '\n",
      "                                                'encoding=\"UTF-8\"?> <table '\n",
      "                                                'frame=\"hsides\" '\n",
      "                                                'rules=\"groups\"><thead><tr><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Relation</bold> '\n",
      "                                                '</th><th align=\"left\" '\n",
      "                                                'colspan=\"4\"> '\n",
      "                                                '<bold>Precision</bold> '\n",
      "                                                '</th><th align=\"left\" '\n",
      "                                                'colspan=\"4\"> <bold>Harvested '\n",
      "                                                'facts</bold> '\n",
      "                                                '</th></tr><tr><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Encyclopedic</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Scientific</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Encyclopedic +</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Encyclopedic +</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Encyclopedic</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Scientific</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Encyclopedic +</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Encyclopedic +</bold> '\n",
      "                                                '</th></tr><tr><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>sources</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>sources</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>scientific</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>scientific +</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>sources</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>sources</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>scientific</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>scientific +</bold> '\n",
      "                                                '</th></tr><tr><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>sources</bold> </th><th '\n",
      "                                                'align=\"left\"> <bold>social '\n",
      "                                                'sources</bold> </th><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>sources</bold> </th><th '\n",
      "                                                'align=\"left\"> <bold>social '\n",
      "                                                'sources</bold> '\n",
      "                                                '</th></tr></thead><tbody><tr><td '\n",
      "                                                'align=\"left\">Affects</td><td '\n",
      "                                                'align=\"left\">0.855 '\n",
      "                                                '+-0.047</td><td '\n",
      "                                                'align=\"left\">0.762 '\n",
      "                                                '+-0.049</td><td align=\"left\"> '\n",
      "                                                '<bold>0.825 +-0.047</bold> '\n",
      "                                                '</td><td align=\"left\">0.767 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">1,278</td><td '\n",
      "                                                'align=\"left\">450</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>2,388</bold> </td><td '\n",
      "                                                'align=\"left\">5,053</td></tr><tr><td '\n",
      "                                                'align=\"left\">Aggravates</td><td '\n",
      "                                                'align=\"left\">0.810 '\n",
      "                                                '+-0.041</td><td '\n",
      "                                                'align=\"left\">0.459 '\n",
      "                                                '+-0.044</td><td align=\"left\"> '\n",
      "                                                '<bold>0.829 +-0.049</bold> '\n",
      "                                                '</td><td align=\"left\">0.785 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">130</td><td '\n",
      "                                                'align=\"left\">371</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>432</bold> </td><td '\n",
      "                                                'align=\"left\">708</td></tr><tr><td '\n",
      "                                                'align=\"left\">Alleviates</td><td '\n",
      "                                                'align=\"left\">0.953 '\n",
      "                                                '+-0.039</td><td '\n",
      "                                                'align=\"left\">0.735 '\n",
      "                                                '+-0.048</td><td align=\"left\"> '\n",
      "                                                '<bold>0.786 +-0.046</bold> '\n",
      "                                                '</td><td align=\"left\">0.736 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">903</td><td '\n",
      "                                                'align=\"left\">4,433</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>4,530</bold> </td><td '\n",
      "                                                'align=\"left\">6,790</td></tr><tr><td '\n",
      "                                                'align=\"left\">Causes</td><td '\n",
      "                                                'align=\"left\">0.904 '\n",
      "                                                '+-0.039</td><td '\n",
      "                                                'align=\"left\">0.674 '\n",
      "                                                '+-0.049</td><td align=\"left\"> '\n",
      "                                                '<bold>0.801 +-0.049</bold> '\n",
      "                                                '</td><td align=\"left\">0.792 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">28,119</td><td '\n",
      "                                                'align=\"left\">19,203</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>47,463</bold> </td><td '\n",
      "                                                'align=\"left\">62,407</td></tr><tr><td '\n",
      "                                                'align=\"left\">Complication</td><td '\n",
      "                                                'align=\"left\">0.917 '\n",
      "                                                '+-0.039</td><td '\n",
      "                                                'align=\"left\">0.397 '\n",
      "                                                '+-0.049</td><td align=\"left\"> '\n",
      "                                                '<bold>0.897 +-0.041</bold> '\n",
      "                                                '</td><td align=\"left\">0.869 '\n",
      "                                                '+-0.046</td><td '\n",
      "                                                'align=\"left\">1,011</td><td '\n",
      "                                                'align=\"left\">1,475</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>1,524</bold> </td><td '\n",
      "                                                'align=\"left\">1,566</td></tr><tr><td '\n",
      "                                                'align=\"left\">Contraindicates</td><td '\n",
      "                                                'align=\"left\">0.874 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">0.710 '\n",
      "                                                '+-0.000</td><td align=\"left\"> '\n",
      "                                                '<bold>0.961 +-0.030</bold> '\n",
      "                                                '</td><td align=\"left\">0.908 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">512</td><td '\n",
      "                                                'align=\"left\">49</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>1,808</bold> </td><td '\n",
      "                                                'align=\"left\">1,831</td></tr><tr><td '\n",
      "                                                'align=\"left\">CreatesRisk</td><td '\n",
      "                                                'align=\"left\">0.878 '\n",
      "                                                '+-0.047</td><td '\n",
      "                                                'align=\"left\">0.569 '\n",
      "                                                '+-0.049</td><td align=\"left\"> '\n",
      "                                                '<bold>0.720 +-0.040</bold> '\n",
      "                                                '</td><td align=\"left\">0.620 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">4,407</td><td '\n",
      "                                                'align=\"left\">24,695</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>18,508</bold> </td><td '\n",
      "                                                'align=\"left\">32,211</td></tr><tr><td '\n",
      "                                                'align=\"left\">Diagnoses</td><td '\n",
      "                                                'align=\"left\">0.964 '\n",
      "                                                '+-0.035</td><td '\n",
      "                                                'align=\"left\">0.839 '\n",
      "                                                '+-0.049</td><td align=\"left\"> '\n",
      "                                                '<bold>0.860 +-0.048</bold> '\n",
      "                                                '</td><td align=\"left\">0.840 '\n",
      "                                                '+-0.047</td><td '\n",
      "                                                'align=\"left\">813</td><td '\n",
      "                                                'align=\"left\">5,920</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>4,832</bold> </td><td '\n",
      "                                                'align=\"left\">9,743</td></tr><tr><td '\n",
      "                                                'align=\"left\">Interacts</td><td '\n",
      "                                                'align=\"left\">0.964 '\n",
      "                                                '+-0.035</td><td '\n",
      "                                                'align=\"left\">0.709 '\n",
      "                                                '+-0.000</td><td align=\"left\"> '\n",
      "                                                '<bold>0.965 +-0.034</bold> '\n",
      "                                                '</td><td align=\"left\">0.957 '\n",
      "                                                '+-0.034</td><td '\n",
      "                                                'align=\"left\">164,912</td><td '\n",
      "                                                'align=\"left\">103</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>164,912</bold> </td><td '\n",
      "                                                'align=\"left\">164,912</td></tr><tr><td '\n",
      "                                                'align=\"left\">IsSymptom</td><td '\n",
      "                                                'align=\"left\">0.891 '\n",
      "                                                '+-0.042</td><td '\n",
      "                                                'align=\"left\">0.482 '\n",
      "                                                '+-0.050</td><td align=\"left\"> '\n",
      "                                                '<bold>0.858 +-0.048</bold> '\n",
      "                                                '</td><td align=\"left\">0.694 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">4,878</td><td '\n",
      "                                                'align=\"left\">2,320</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>6,395</bold> </td><td '\n",
      "                                                'align=\"left\">11,017</td></tr><tr><td '\n",
      "                                                'align=\"left\">ReducesRisk</td><td '\n",
      "                                                'align=\"left\">0.797 '\n",
      "                                                '+-0.045</td><td '\n",
      "                                                'align=\"left\">0.637 '\n",
      "                                                '+-0.046</td><td align=\"left\"> '\n",
      "                                                '<bold>0.762 +-0.048</bold> '\n",
      "                                                '</td><td align=\"left\">0.751 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">1,712</td><td '\n",
      "                                                'align=\"left\">4,684</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>4,489</bold> </td><td '\n",
      "                                                'align=\"left\">5,865</td></tr><tr><td '\n",
      "                                                'align=\"left\">SideEffect</td><td '\n",
      "                                                'align=\"left\">0.956 '\n",
      "                                                '+-0.038</td><td '\n",
      "                                                'align=\"left\">0.826 '\n",
      "                                                '+-0.000</td><td align=\"left\"> '\n",
      "                                                '<bold>0.964 +-0.035</bold> '\n",
      "                                                '</td><td align=\"left\">0.971 '\n",
      "                                                '+-0.026</td><td '\n",
      "                                                'align=\"left\">270,600</td><td '\n",
      "                                                'align=\"left\">139</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>270,709</bold> </td><td '\n",
      "                                                'align=\"left\">271,416</td></tr><tr><td '\n",
      "                                                'align=\"left\">Treats</td><td '\n",
      "                                                'align=\"left\">0.850 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">0.581 '\n",
      "                                                '+-0.045</td><td align=\"left\"> '\n",
      "                                                '<bold>0.898 +-0.041</bold> '\n",
      "                                                '</td><td align=\"left\">0.566 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">11,915</td><td '\n",
      "                                                'align=\"left\">9,318</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>14,699</bold> </td><td '\n",
      "                                                'align=\"left\">35,803</td></tr><tr><td '\n",
      "                                                'align=\"left\">Aggregated '\n",
      "                                                '<sup>*</sup> </td><td '\n",
      "                                                'align=\"left\">0.951</td><td '\n",
      "                                                'align=\"left\">0.630</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>0.933</bold> </td><td '\n",
      "                                                'align=\"left\">0.892</td><td '\n",
      "                                                'align=\"left\">491,190</td><td '\n",
      "                                                'align=\"left\">73,160</td><td '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>542,689</bold> </td><td '\n",
      "                                                'align=\"left\">609,322</td></tr></tbody></table> '},\n",
      "                              'offset': 26395,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Relation Precision Harvested facts    '\n",
      "                                      'Encyclopedic Scientific Encyclopedic + '\n",
      "                                      'Encyclopedic + Encyclopedic Scientific '\n",
      "                                      'Encyclopedic + Encyclopedic +    '\n",
      "                                      'sources sources scientific scientific + '\n",
      "                                      'sources sources scientific scientific '\n",
      "                                      '+      sources social sources   sources '\n",
      "                                      'social sources   Affects 0.855 +-0.047 '\n",
      "                                      '0.762 +-0.049 0.825 +-0.047 0.767 '\n",
      "                                      '+-0.048 1,278 450 2,388 5,053   '\n",
      "                                      'Aggravates 0.810 +-0.041 0.459 +-0.044 '\n",
      "                                      '0.829 +-0.049 0.785 +-0.049 130 371 432 '\n",
      "                                      '708   Alleviates 0.953 +-0.039 0.735 '\n",
      "                                      '+-0.048 0.786 +-0.046 0.736 +-0.048 903 '\n",
      "                                      '4,433 4,530 6,790   Causes 0.904 '\n",
      "                                      '+-0.039 0.674 +-0.049 0.801 +-0.049 '\n",
      "                                      '0.792 +-0.049 28,119 19,203 47,463 '\n",
      "                                      '62,407   Complication 0.917 +-0.039 '\n",
      "                                      '0.397 +-0.049 0.897 +-0.041 0.869 '\n",
      "                                      '+-0.046 1,011 1,475 1,524 1,566   '\n",
      "                                      'Contraindicates 0.874 +-0.048 0.710 '\n",
      "                                      '+-0.000 0.961 +-0.030 0.908 +-0.048 512 '\n",
      "                                      '49 1,808 1,831   CreatesRisk 0.878 '\n",
      "                                      '+-0.047 0.569 +-0.049 0.720 +-0.040 '\n",
      "                                      '0.620 +-0.049 4,407 24,695 18,508 '\n",
      "                                      '32,211   Diagnoses 0.964 +-0.035 0.839 '\n",
      "                                      '+-0.049 0.860 +-0.048 0.840 +-0.047 813 '\n",
      "                                      '5,920 4,832 9,743   Interacts 0.964 '\n",
      "                                      '+-0.035 0.709 +-0.000 0.965 +-0.034 '\n",
      "                                      '0.957 +-0.034 164,912 103 164,912 '\n",
      "                                      '164,912   IsSymptom 0.891 +-0.042 0.482 '\n",
      "                                      '+-0.050 0.858 +-0.048 0.694 +-0.048 '\n",
      "                                      '4,878 2,320 6,395 11,017   ReducesRisk '\n",
      "                                      '0.797 +-0.045 0.637 +-0.046 0.762 '\n",
      "                                      '+-0.048 0.751 +-0.049 1,712 4,684 4,489 '\n",
      "                                      '5,865   SideEffect 0.956 +-0.038 0.826 '\n",
      "                                      '+-0.000 0.964 +-0.035 0.971 +-0.026 '\n",
      "                                      '270,600 139 270,709 271,416   Treats '\n",
      "                                      '0.850 +-0.048 0.581 +-0.045 0.898 '\n",
      "                                      '+-0.041 0.566 +-0.048 11,915 9,318 '\n",
      "                                      '14,699 35,803   Aggregated * 0.951 '\n",
      "                                      '0.630 0.933 0.892 491,190 73,160 '\n",
      "                                      '542,689 609,322   '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab4.xml',\n",
      "                                         'id': 'Tab4',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_foot'},\n",
      "                              'offset': 27956,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': '*Precision values are averaged and '\n",
      "                                      'numbers of harvested facts are summed.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab5.xml',\n",
      "                                         'id': 'Tab5',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_caption'},\n",
      "                              'offset': 28030,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Evaluation of the impact of different '\n",
      "                                      'components '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab5.xml',\n",
      "                                         'id': 'Tab5',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table',\n",
      "                                         'xml': '<?xml version=\"1.0\" '\n",
      "                                                'encoding=\"UTF-8\"?> <table '\n",
      "                                                'frame=\"hsides\" '\n",
      "                                                'rules=\"groups\"><thead><tr><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Relation</bold> '\n",
      "                                                '</th><th align=\"left\" '\n",
      "                                                'colspan=\"4\"> '\n",
      "                                                '<bold>Precision</bold> '\n",
      "                                                '</th><th align=\"left\" '\n",
      "                                                'colspan=\"4\"> <bold>Harvested '\n",
      "                                                'facts</bold> '\n",
      "                                                '</th></tr><tr><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"> <bold>Full '\n",
      "                                                'pipeline</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Without</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Without</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Without</bold> </th><th '\n",
      "                                                'align=\"left\"> <bold>Full '\n",
      "                                                'pipeline</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Without</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Without</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Without</bold> '\n",
      "                                                '</th></tr><tr><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>encyclopedic +</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>document</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>statistical</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>consistency</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>encyclopedic +</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>document</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>statistical</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>consistency</bold> '\n",
      "                                                '</th></tr><tr><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>scientific '\n",
      "                                                'sources</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>structure</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>analysis</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>reasoning</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>scientific '\n",
      "                                                'sources</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>structure</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>analysis</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>reasoning</bold> '\n",
      "                                                '</th></tr></thead><tbody><tr><td '\n",
      "                                                'align=\"left\">Affects</td><td '\n",
      "                                                'align=\"left\">0.825 '\n",
      "                                                '+-0.047</td><td '\n",
      "                                                'align=\"left\">0.882 '\n",
      "                                                '+-0.044</td><td '\n",
      "                                                'align=\"left\">0.821 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">0.171 '\n",
      "                                                '+-0.051</td><td '\n",
      "                                                'align=\"left\">2,388</td><td '\n",
      "                                                'align=\"left\">2,350</td><td '\n",
      "                                                'align=\"left\">4,088</td><td '\n",
      "                                                'align=\"left\">29,477</td></tr><tr><td '\n",
      "                                                'align=\"left\">Aggravates</td><td '\n",
      "                                                'align=\"left\">0.829 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">0.833 '\n",
      "                                                '+-0.036</td><td '\n",
      "                                                'align=\"left\">0.598 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">0.592 '\n",
      "                                                '+-0.053</td><td '\n",
      "                                                'align=\"left\">432</td><td '\n",
      "                                                'align=\"left\">431</td><td '\n",
      "                                                'align=\"left\">592</td><td '\n",
      "                                                'align=\"left\">1,730</td></tr><tr><td '\n",
      "                                                'align=\"left\">Alleviates</td><td '\n",
      "                                                'align=\"left\">0.786 '\n",
      "                                                '+-0.046</td><td '\n",
      "                                                'align=\"left\">0.778 '\n",
      "                                                '+-0.050</td><td '\n",
      "                                                'align=\"left\">0.320 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">0.289 '\n",
      "                                                '+-0.062</td><td '\n",
      "                                                'align=\"left\">4,530</td><td '\n",
      "                                                'align=\"left\">4,387</td><td '\n",
      "                                                'align=\"left\">18,142</td><td '\n",
      "                                                'align=\"left\">16,943</td></tr><tr><td '\n",
      "                                                'align=\"left\">Causes</td><td '\n",
      "                                                'align=\"left\">0.801 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">0.800 '\n",
      "                                                '+-0.046</td><td '\n",
      "                                                'align=\"left\">0.631 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">0.490 '\n",
      "                                                '+-0.069</td><td '\n",
      "                                                'align=\"left\">47,463</td><td '\n",
      "                                                'align=\"left\">30,563</td><td '\n",
      "                                                'align=\"left\">66,833</td><td '\n",
      "                                                'align=\"left\">91,784</td></tr><tr><td '\n",
      "                                                'align=\"left\">Complication</td><td '\n",
      "                                                'align=\"left\">0.897 '\n",
      "                                                '+-0.041</td><td '\n",
      "                                                'align=\"left\">0.781 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">0.376 '\n",
      "                                                '+-0.050</td><td '\n",
      "                                                'align=\"left\">0.739 '\n",
      "                                                '+-0.050</td><td '\n",
      "                                                'align=\"left\">1,524</td><td '\n",
      "                                                'align=\"left\">700</td><td '\n",
      "                                                'align=\"left\">4,812</td><td '\n",
      "                                                'align=\"left\">2,955</td></tr><tr><td '\n",
      "                                                'align=\"left\">Contraindicates</td><td '\n",
      "                                                'align=\"left\">0.961 '\n",
      "                                                '+-0.030</td><td '\n",
      "                                                'align=\"left\">0.914 '\n",
      "                                                '+-0.043</td><td '\n",
      "                                                'align=\"left\">0.122 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">0.630 '\n",
      "                                                '+-0.059</td><td '\n",
      "                                                'align=\"left\">1,808</td><td '\n",
      "                                                'align=\"left\">365</td><td '\n",
      "                                                'align=\"left\">26,298</td><td '\n",
      "                                                'align=\"left\">15,279</td></tr><tr><td '\n",
      "                                                'align=\"left\">CreatesRisk</td><td '\n",
      "                                                'align=\"left\">0.720 '\n",
      "                                                '+-0.040</td><td '\n",
      "                                                'align=\"left\">0.750 '\n",
      "                                                '+-0.044</td><td '\n",
      "                                                'align=\"left\">0.386 '\n",
      "                                                '+-0.047</td><td '\n",
      "                                                'align=\"left\">0.406 '\n",
      "                                                '+-0.067</td><td '\n",
      "                                                'align=\"left\">18,508</td><td '\n",
      "                                                'align=\"left\">17,282</td><td '\n",
      "                                                'align=\"left\">77,158</td><td '\n",
      "                                                'align=\"left\">48,159</td></tr><tr><td '\n",
      "                                                'align=\"left\">Diagnoses</td><td '\n",
      "                                                'align=\"left\">0.860 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">0.887 '\n",
      "                                                '+-0.044</td><td '\n",
      "                                                'align=\"left\">0.802 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">0.303 '\n",
      "                                                '+-0.063</td><td '\n",
      "                                                'align=\"left\">4,832</td><td '\n",
      "                                                'align=\"left\">4,002</td><td '\n",
      "                                                'align=\"left\">7,467</td><td '\n",
      "                                                'align=\"left\">35,326</td></tr><tr><td '\n",
      "                                                'align=\"left\">Interacts</td><td '\n",
      "                                                'align=\"left\">0.965 '\n",
      "                                                '+-0.034</td><td '\n",
      "                                                'align=\"left\">0.858 '\n",
      "                                                '+-0.046</td><td '\n",
      "                                                'align=\"left\">0.953 '\n",
      "                                                '+-0.047</td><td '\n",
      "                                                'align=\"left\">0.941 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">164,912</td><td '\n",
      "                                                'align=\"left\">392</td><td '\n",
      "                                                'align=\"left\">200,935</td><td '\n",
      "                                                'align=\"left\">187,201</td></tr><tr><td '\n",
      "                                                'align=\"left\">IsSymptom</td><td '\n",
      "                                                'align=\"left\">0.858 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">0.691 '\n",
      "                                                '+-0.050</td><td '\n",
      "                                                'align=\"left\">0.625 '\n",
      "                                                '+-0.049</td><td '\n",
      "                                                'align=\"left\">0.328 '\n",
      "                                                '+-0.064</td><td '\n",
      "                                                'align=\"left\">6,395</td><td '\n",
      "                                                'align=\"left\">2,920</td><td '\n",
      "                                                'align=\"left\">9,543</td><td '\n",
      "                                                'align=\"left\">29,776</td></tr><tr><td '\n",
      "                                                'align=\"left\">ReducesRisk</td><td '\n",
      "                                                'align=\"left\">0.762 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">0.729 '\n",
      "                                                '+-0.050</td><td '\n",
      "                                                'align=\"left\">0.228 '\n",
      "                                                '+-0.046</td><td '\n",
      "                                                'align=\"left\">0.406 '\n",
      "                                                '+-0.067</td><td '\n",
      "                                                'align=\"left\">4,489</td><td '\n",
      "                                                'align=\"left\">4,043</td><td '\n",
      "                                                'align=\"left\">11,023</td><td '\n",
      "                                                'align=\"left\">14,729</td></tr><tr><td '\n",
      "                                                'align=\"left\">SideEffect</td><td '\n",
      "                                                'align=\"left\">0.964 '\n",
      "                                                '+-0.035</td><td '\n",
      "                                                'align=\"left\">0.938 '\n",
      "                                                '+-0.048</td><td '\n",
      "                                                'align=\"left\">0.941 '\n",
      "                                                '+-0.046</td><td '\n",
      "                                                'align=\"left\">0.879 '\n",
      "                                                '+-0.050</td><td '\n",
      "                                                'align=\"left\">270,709</td><td '\n",
      "                                                'align=\"left\">924</td><td '\n",
      "                                                'align=\"left\">270,427</td><td '\n",
      "                                                'align=\"left\">338,645</td></tr><tr><td '\n",
      "                                                'align=\"left\">Treats</td><td '\n",
      "                                                'align=\"left\">0.898 '\n",
      "                                                '+-0.041</td><td '\n",
      "                                                'align=\"left\">0.784 '\n",
      "                                                '+-0.050</td><td '\n",
      "                                                'align=\"left\">0.549 '\n",
      "                                                '+-0.050</td><td '\n",
      "                                                'align=\"left\">0.402 '\n",
      "                                                '+-0.067</td><td '\n",
      "                                                'align=\"left\">14,699</td><td '\n",
      "                                                'align=\"left\">14,057</td><td '\n",
      "                                                'align=\"left\">23,473</td><td '\n",
      "                                                'align=\"left\">45,439</td></tr><tr><td '\n",
      "                                                'align=\"left\">Aggregated '\n",
      "                                                '<sup>*</sup> </td><td '\n",
      "                                                'align=\"left\">0.933</td><td '\n",
      "                                                'align=\"left\">0.784</td><td '\n",
      "                                                'align=\"left\">0.777</td><td '\n",
      "                                                'align=\"left\">0.707</td><td '\n",
      "                                                'align=\"left\">542,689</td><td '\n",
      "                                                'align=\"left\">82,416</td><td '\n",
      "                                                'align=\"left\">720,791</td><td '\n",
      "                                                'align=\"left\">857,443</td></tr></tbody></table> '},\n",
      "                              'offset': 28081,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Relation Precision Harvested facts    '\n",
      "                                      'Full pipeline Without Without Without '\n",
      "                                      'Full pipeline Without Without '\n",
      "                                      'Without    encyclopedic + document '\n",
      "                                      'statistical consistency encyclopedic + '\n",
      "                                      'document statistical consistency    '\n",
      "                                      'scientific sources structure analysis '\n",
      "                                      'reasoning scientific sources structure '\n",
      "                                      'analysis reasoning   Affects 0.825 '\n",
      "                                      '+-0.047 0.882 +-0.044 0.821 +-0.048 '\n",
      "                                      '0.171 +-0.051 2,388 2,350 4,088 '\n",
      "                                      '29,477   Aggravates 0.829 +-0.049 0.833 '\n",
      "                                      '+-0.036 0.598 +-0.049 0.592 +-0.053 432 '\n",
      "                                      '431 592 1,730   Alleviates 0.786 '\n",
      "                                      '+-0.046 0.778 +-0.050 0.320 +-0.049 '\n",
      "                                      '0.289 +-0.062 4,530 4,387 18,142 '\n",
      "                                      '16,943   Causes 0.801 +-0.049 0.800 '\n",
      "                                      '+-0.046 0.631 +-0.048 0.490 +-0.069 '\n",
      "                                      '47,463 30,563 66,833 91,784   '\n",
      "                                      'Complication 0.897 +-0.041 0.781 '\n",
      "                                      '+-0.048 0.376 +-0.050 0.739 +-0.050 '\n",
      "                                      '1,524 700 4,812 2,955   Contraindicates '\n",
      "                                      '0.961 +-0.030 0.914 +-0.043 0.122 '\n",
      "                                      '+-0.049 0.630 +-0.059 1,808 365 26,298 '\n",
      "                                      '15,279   CreatesRisk 0.720 +-0.040 '\n",
      "                                      '0.750 +-0.044 0.386 +-0.047 0.406 '\n",
      "                                      '+-0.067 18,508 17,282 77,158 48,159   '\n",
      "                                      'Diagnoses 0.860 +-0.048 0.887 +-0.044 '\n",
      "                                      '0.802 +-0.049 0.303 +-0.063 4,832 4,002 '\n",
      "                                      '7,467 35,326   Interacts 0.965 +-0.034 '\n",
      "                                      '0.858 +-0.046 0.953 +-0.047 0.941 '\n",
      "                                      '+-0.049 164,912 392 200,935 187,201   '\n",
      "                                      'IsSymptom 0.858 +-0.048 0.691 +-0.050 '\n",
      "                                      '0.625 +-0.049 0.328 +-0.064 6,395 2,920 '\n",
      "                                      '9,543 29,776   ReducesRisk 0.762 '\n",
      "                                      '+-0.048 0.729 +-0.050 0.228 +-0.046 '\n",
      "                                      '0.406 +-0.067 4,489 4,043 11,023 '\n",
      "                                      '14,729   SideEffect 0.964 +-0.035 0.938 '\n",
      "                                      '+-0.048 0.941 +-0.046 0.879 +-0.050 '\n",
      "                                      '270,709 924 270,427 338,645   Treats '\n",
      "                                      '0.898 +-0.041 0.784 +-0.050 0.549 '\n",
      "                                      '+-0.050 0.402 +-0.067 14,699 14,057 '\n",
      "                                      '23,473 45,439   Aggregated * 0.933 '\n",
      "                                      '0.784 0.777 0.707 542,689 82,416 '\n",
      "                                      '720,791 857,443   '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab5.xml',\n",
      "                                         'id': 'Tab5',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_foot'},\n",
      "                              'offset': 29691,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': '*Precision values are averaged and '\n",
      "                                      'numbers of harvested facts are summed.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 29765,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We ran extensive experiments with the '\n",
      "                                      'input corpora listed in Table 2, and '\n",
      "                                      \"created different KB's based on \"\n",
      "                                      'different configurations. We assess the '\n",
      "                                      'size and quality of each KB, in terms '\n",
      "                                      'of their numbers of facts and their '\n",
      "                                      'precision evaluated by random sampling '\n",
      "                                      'of facts. Tables 4 and 5 give the '\n",
      "                                      'results, for different choices of input '\n",
      "                                      'corpora and different configurations of '\n",
      "                                      'the KnowLife pipeline, respectively. '\n",
      "                                      'Recall is not evaluated, as there is no '\n",
      "                                      'gold standard for fully comprehensive '\n",
      "                                      'facts. To ensure that our findings are '\n",
      "                                      'significant, for each relation, we '\n",
      "                                      'computed the Wilson confidence interval '\n",
      "                                      'at alpha = 5%, and kept evaluating '\n",
      "                                      'facts until the interval width fell '\n",
      "                                      'below 5%. An interval width of 0% means '\n",
      "                                      'that all the facts were evaluated. Four '\n",
      "                                      'different annotators evaluated the '\n",
      "                                      'facts, judging them as true or false '\n",
      "                                      'based on provenance information. As for '\n",
      "                                      'inter-annotator agreement, 22,002 facts '\n",
      "                                      \"were evaluated; the value of Fleiss' \"\n",
      "                                      'Kappa was 0.505, which indicates a '\n",
      "                                      'moderate agreement among all four '\n",
      "                                      'annotators. The complete set of '\n",
      "                                      'evaluated facts is in the supplementary '\n",
      "                                      'material (see Additional file 3).   '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 30887,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Impact of different text genres'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 30919,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We first discuss the results obtained '\n",
      "                                      'from the different text genres: i) '\n",
      "                                      'scientific (PubMed publications), ii) '\n",
      "                                      'encyclopedic (Web portals like Mayo '\n",
      "                                      'Clinic or Wikipedia), iii) social '\n",
      "                                      '(discussion forums). Table 4 gives, '\n",
      "                                      'column-wise, the number of facts and '\n",
      "                                      'precision figures for four different '\n",
      "                                      'combinations of genres.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 31234,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Generally, combining genres gave more '\n",
      "                                      'facts at a lower precision, as texts of '\n",
      "                                      'lower quality like social sources '\n",
      "                                      'introduced noise. The combination that '\n",
      "                                      'gave the best balance of precision and '\n",
      "                                      'total yield was scientific with '\n",
      "                                      'encyclopedic sources, with a '\n",
      "                                      'micro-averaged precision of 0.933 for a '\n",
      "                                      'total of 542,689 facts. We consider '\n",
      "                                      \"this the best of the KB's that KnowLife \"\n",
      "                                      'generated.'},\n",
      "                             {'annotations': [{'id': '16762',\n",
      "                                               'infons': {'accession': '@DISEASE_Diabetes_Mellitus',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D003920',\n",
      "                                                          'name': 'Diabetes '\n",
      "                                                                  'Mellitus',\n",
      "                                                          'normalized': ['D003920'],\n",
      "                                                          'normalized_id': 'D003920',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 17,\n",
      "                                                              'offset': 32308}],\n",
      "                                               'text': 'Diabetes Syndrome'},\n",
      "                                              {'id': '16763',\n",
      "                                               'infons': {'accession': '@DISEASE_Microcephaly',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D008831',\n",
      "                                                          'name': 'Microcephaly',\n",
      "                                                          'normalized': ['D008831'],\n",
      "                                                          'normalized_id': 'D008831',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 32280}],\n",
      "                                               'text': 'Microcephaly'},\n",
      "                                              {'id': '16764',\n",
      "                                               'infons': {'accession': '@DISEASE_Epilepsy',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004827',\n",
      "                                                          'name': 'Epilepsy',\n",
      "                                                          'normalized': ['D004827'],\n",
      "                                                          'normalized_id': 'D004827',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 8,\n",
      "                                                              'offset': 32294}],\n",
      "                                               'text': 'Epilepsy'}],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 31614,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The best overall precision was achieved '\n",
      "                                      'when using encyclopedic texts only. '\n",
      "                                      'This confirmed our hypothesis that a '\n",
      "                                      'pattern-based approach works best when '\n",
      "                                      'the language is simple and '\n",
      "                                      'grammatically correct. Contrast this '\n",
      "                                      'with scientific publications which '\n",
      "                                      'often exhibit convoluted language, and '\n",
      "                                      'online discussions with a notable '\n",
      "                                      'fraction of grammatically incorrect '\n",
      "                                      'language. In these cases, the quality '\n",
      "                                      'of patterns degraded and precision '\n",
      "                                      'dropped. Incorrect facts stemming from '\n",
      "                                      'errors in the entity recognition step '\n",
      "                                      'were especially rampant in online '\n",
      "                                      'discussions, where colloquial language '\n",
      "                                      '(for example, meds, or short for '\n",
      "                                      'medicines) led to incorrect entities '\n",
      "                                      '(acronym for Microcephaly, Epilepsy, '\n",
      "                                      'and Diabetes Syndrome).'},\n",
      "                             {'annotations': [{'id': '16765',\n",
      "                                               'infons': {'biotype': 'chemical',\n",
      "                                                          'identifier': '-',\n",
      "                                                          'normalized_id': None,\n",
      "                                                          'type': 'Chemical',\n",
      "                                                          'valid': False},\n",
      "                                               'locations': [{'length': 3,\n",
      "                                                              'offset': 32681}],\n",
      "                                               'text': 'DOM'}],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 32328,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The results vary highly across the 13 '\n",
      "                                      'relations in our experiments. The '\n",
      "                                      'number of facts depends on the extent '\n",
      "                                      'to which the text sources express a '\n",
      "                                      'relation, while precision reflects how '\n",
      "                                      'decisively patterns point to that '\n",
      "                                      'relation. Interacts and SideEffect are '\n",
      "                                      'prime examples: the drugs.com portal '\n",
      "                                      'lists many side effects and drug-drug '\n",
      "                                      'interactions by the DOM structure, '\n",
      "                                      'which boosted the extraction accuracy '\n",
      "                                      'of KnowLife, leading to many facts at '\n",
      "                                      'precisions of 95.6% and 96.4%, '\n",
      "                                      'respectively. Facts for the relations '\n",
      "                                      'Alleviates, CreatesRisk, and '\n",
      "                                      'ReducesRisk, on the other hand, mostly '\n",
      "                                      'came from scientific publications, '\n",
      "                                      'which resulted in fewer facts and lower '\n",
      "                                      'precision.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 32995,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'A few relations, however, defied these '\n",
      "                                      'general trends. Patterns of '\n",
      "                                      'Contraindicates were too sparse and '\n",
      "                                      'ambiguous within encyclopedic texts '\n",
      "                                      'alone and also within scientific '\n",
      "                                      'publications alone. However, when the '\n",
      "                                      'two genres were combined, the good '\n",
      "                                      'patterns reached a critical mass to '\n",
      "                                      'break through the confidence threshold, '\n",
      "                                      'giving rise to a sudden increase in '\n",
      "                                      'harvested facts. For the CreatesRisk '\n",
      "                                      'and ReducesRisk relations, combining '\n",
      "                                      'encyclopedic and scientific sources '\n",
      "                                      'increased the number of facts compared '\n",
      "                                      'to using only encyclopedic texts, and '\n",
      "                                      'increased the precision compared to '\n",
      "                                      'using only scientific publications.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 33611,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'As Table 4 shows, incorporating social '\n",
      "                                      'sources brought a significant gain in '\n",
      "                                      'the number of harvested facts, at a '\n",
      "                                      'trade-off of lowered precision. As '\n",
      "                                      'pointed out, there are facts that come '\n",
      "                                      'only from social sources and, depending '\n",
      "                                      'on the use case, it is still worthwhile '\n",
      "                                      'to incorporate them; for example, to '\n",
      "                                      'facilitate search and discovery '\n",
      "                                      'applications where recall may be more '\n",
      "                                      'important. Morever, the patterns '\n",
      "                                      'extracted from encyclopedic and '\n",
      "                                      'scientific sources could be reused to '\n",
      "                                      'annotate text in social sources, so as '\n",
      "                                      'to identify existing information.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab6.xml',\n",
      "                                         'id': 'Tab6',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_caption'},\n",
      "                              'offset': 34161,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Number of fact occurrences in text '\n",
      "                                      'sources '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab6.xml',\n",
      "                                         'id': 'Tab6',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table',\n",
      "                                         'xml': '<?xml version=\"1.0\" '\n",
      "                                                'encoding=\"UTF-8\"?> <table '\n",
      "                                                'frame=\"hsides\" '\n",
      "                                                'rules=\"groups\"><thead><tr><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Genre</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Source</bold> </th><th '\n",
      "                                                'align=\"left\"> <bold>Fact '\n",
      "                                                'occurrences</bold> '\n",
      "                                                '</th></tr></thead><tbody><tr><td '\n",
      "                                                'align=\"left\">Scientific '\n",
      "                                                'Publications</td><td '\n",
      "                                                'align=\"left\">PubMed '\n",
      "                                                'Medline</td><td '\n",
      "                                                'align=\"left\">39,266</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\">PubMed '\n",
      "                                                'Central</td><td '\n",
      "                                                'align=\"left\">6,979</td></tr><tr><td '\n",
      "                                                'align=\"left\">Encyclopedic '\n",
      "                                                'Articles</td><td '\n",
      "                                                'align=\"left\">Drugs.com</td><td '\n",
      "                                                'align=\"left\">461,130</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\">Mayo '\n",
      "                                                'Clinic</td><td '\n",
      "                                                'align=\"left\">35,300</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\">Medline '\n",
      "                                                'Plus</td><td '\n",
      "                                                'align=\"left\">6,559</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\">RxList</td><td '\n",
      "                                                'align=\"left\">5,818</td></tr><tr><td '\n",
      "                                                'align=\"left\"/><td '\n",
      "                                                'align=\"left\">Wikipedia '\n",
      "                                                'Health</td><td '\n",
      "                                                'align=\"left\">17,588</td></tr></tbody></table> '},\n",
      "                              'offset': 34206,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Genre Source Fact occurrences   '\n",
      "                                      'Scientific Publications PubMed Medline '\n",
      "                                      '39,266    PubMed Central 6,979   '\n",
      "                                      'Encyclopedic Articles Drugs.com '\n",
      "                                      '461,130    Mayo Clinic 35,300    '\n",
      "                                      'Medline Plus 6,559    RxList 5,818    '\n",
      "                                      'Wikipedia Health 17,588   '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 34440,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Taking a closer look at the best '\n",
      "                                      'experimental setting, we see that '\n",
      "                                      'scientific and encyclopedic sources in '\n",
      "                                      'KnowLife contribute to a different '\n",
      "                                      'extent to the number of harvested '\n",
      "                                      'facts. Table 6 shows the number of fact '\n",
      "                                      'occurrences in our input sources. '\n",
      "                                      'Recall that a fact can occur in '\n",
      "                                      'multiple sentences in multiple text '\n",
      "                                      'sources. Our experiments show that '\n",
      "                                      'encyclopedic articles are more amenable '\n",
      "                                      'for harvesting facts than scientific '\n",
      "                                      'publications.  '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 34885,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Impact of different components'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 34916,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'In each setting, only one component was '\n",
      "                                      'disabled, and the processing pipeline '\n",
      "                                      'ran with all other components enabled. '\n",
      "                                      'We used the KnowLife setting with '\n",
      "                                      'scientific and encyclopedic sources, '\n",
      "                                      'which, by and large, performed best, as '\n",
      "                                      'the basis for investigating the impact '\n",
      "                                      'of different components in the KnowLife '\n",
      "                                      'pipeline. To this end, we disabled '\n",
      "                                      'individual components: DOM tree '\n",
      "                                      'patterns, statistical analysis of '\n",
      "                                      'patterns, consistency reasoning - each '\n",
      "                                      'disabled separately while retaining the '\n",
      "                                      'others. This way we obtained insight '\n",
      "                                      'into how strongly KnowLife depends on '\n",
      "                                      'each component. Table 5 shows the '\n",
      "                                      'results of this ablation study.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 35546,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' No DOM tree patterns: When '\n",
      "                                      'disregarding patterns on the document '\n",
      "                                      'structure and solely focusing on '\n",
      "                                      'textual patterns, KnowLife degrades in '\n",
      "                                      'precision (from 93% to 78%) and sharply '\n",
      "                                      'drops in the number of acquired facts '\n",
      "                                      '(from ca. 540,000 to 80,000). The '\n",
      "                                      'extent of these general effects varies '\n",
      "                                      'across the different relations. '\n",
      "                                      'Relations whose patterns are '\n",
      "                                      'predominantly encoded in document '\n",
      "                                      'structures - once again Interacts and '\n",
      "                                      'SideEffect - exhibit the most drastic '\n",
      "                                      'loss. On the other hand, relations like '\n",
      "                                      'Affects, Aggravates, Alleviates, and '\n",
      "                                      'Treats, are affected only to a minor '\n",
      "                                      'extent, as their patterns are mostly '\n",
      "                                      'found in free text.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 36181,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' No statistical pattern analysis: Here '\n",
      "                                      'we disabled the statistical analysis of '\n",
      "                                      'pattern confidence and the frequent '\n",
      "                                      'itemset mining for generalizing '\n",
      "                                      'patterns. This way, without confidence '\n",
      "                                      'values, KnowLife kept all patterns, '\n",
      "                                      'including many noisy ones. Patterns '\n",
      "                                      'that would be pruned in the full '\n",
      "                                      'configuration led to poor seed '\n",
      "                                      'patterns; for example, the single word '\n",
      "                                      'causes was taken as a seed pattern for '\n",
      "                                      'both relations SymptomOf and '\n",
      "                                      'Contraindicates. Without frequent '\n",
      "                                      'itemset mining, long and overly '\n",
      "                                      'specific patterns also contributed to '\n",
      "                                      'poor seed patterns. The combined effect '\n",
      "                                      'greatly increased the number of false '\n",
      "                                      'positives, thus dropping in precision '\n",
      "                                      '(from 93% to 77%). In terms of acquired '\n",
      "                                      'facts, not scrutinizing the patterns '\n",
      "                                      'increased the yield (from ca. 540,000 '\n",
      "                                      'to 720,000 facts).'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 36964,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Relations mainly extracted from DOM '\n",
      "                                      'tree patterns, such as Interacts and '\n",
      "                                      'SideEffect, were not much affected. '\n",
      "                                      'Also, relations like Affects and '\n",
      "                                      'Diagnoses exhibited only small losses '\n",
      "                                      'in precision; for these relations, the '\n",
      "                                      'co-occurrence of two types of entities '\n",
      "                                      'is often already sufficient to express '\n",
      "                                      'a relation. The presence of consistency '\n",
      "                                      'constraints on type signatures also '\n",
      "                                      'helped to keep the output quality '\n",
      "                                      'high.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 37377,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' No consistency reasoning: In this '\n",
      "                                      'setting, neither type signatures nor '\n",
      "                                      'other consistency constraints were '\n",
      "                                      'checked. Thus, conflicting facts could '\n",
      "                                      'be accepted, leading to a large '\n",
      "                                      'fraction of false positives. This '\n",
      "                                      'effect was unequivocally witnessed by '\n",
      "                                      'an increase in the number of facts '\n",
      "                                      '(from ca. 540,000 to 850,000) '\n",
      "                                      'accompanied by a sharp decrease in '\n",
      "                                      'precision (from 93% to 70%).'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 37756,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The relations Interacts and SideEffect '\n",
      "                                      'were least affected by this '\n",
      "                                      'degradation, as they are mostly '\n",
      "                                      'expressed in the via document structure '\n",
      "                                      'of encyclopedic texts where entity '\n",
      "                                      'types are implicitly encoded in the DOM '\n",
      "                                      'tree tags (see Figure 2). Here, '\n",
      "                                      'consistency reasoning was not vital.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 38039,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Lessons learned: Overall, this '\n",
      "                                      'ablation study clearly shows that all '\n",
      "                                      'major components of the KnowLife '\n",
      "                                      'pipeline are essential for high quality '\n",
      "                                      '(precision) and high yield (number of '\n",
      "                                      'facts) of the constructed KB. Each of '\n",
      "                                      'the three configurations where one '\n",
      "                                      'component is disabled suffered '\n",
      "                                      'substantial if not dramatic losses in '\n",
      "                                      'either precision or acquired facts, and '\n",
      "                                      'sometimes both. We conclude that the '\n",
      "                                      'full pipeline is a well-designed '\n",
      "                                      'architecture whose strong performance '\n",
      "                                      'cannot be easily achieved by a simpler '\n",
      "                                      'approach.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 38559,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Error analysis'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab7.xml',\n",
      "                                         'id': 'Tab7',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_caption'},\n",
      "                              'offset': 38574,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Error analysis (number of facts in '\n",
      "                                      'brackets) '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab7.xml',\n",
      "                                         'id': 'Tab7',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table',\n",
      "                                         'xml': '<?xml version=\"1.0\" '\n",
      "                                                'encoding=\"UTF-8\"?> <table '\n",
      "                                                'frame=\"hsides\" '\n",
      "                                                'rules=\"groups\"><thead><tr><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\" colspan=\"3\"> '\n",
      "                                                '<bold>Percentage based on '\n",
      "                                                'text genre</bold> '\n",
      "                                                '</th></tr><tr><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Percentage</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Cause of error</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Encyclopaedic</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Scientific</bold> '\n",
      "                                                '</th><th align=\"left\"> '\n",
      "                                                '<bold>Social</bold> '\n",
      "                                                '</th></tr><tr><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"/><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>sources</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>sources</bold> </th><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>sources</bold> '\n",
      "                                                '</th></tr></thead><tbody><tr><td '\n",
      "                                                'align=\"left\">8.16% '\n",
      "                                                '(62)</td><td '\n",
      "                                                'align=\"left\">Preprocessing</td><td '\n",
      "                                                'align=\"left\">38.71% '\n",
      "                                                '(24)</td><td '\n",
      "                                                'align=\"left\">3.23% '\n",
      "                                                '(2)</td><td '\n",
      "                                                'align=\"left\">58.06% '\n",
      "                                                '(36)</td></tr><tr><td '\n",
      "                                                'align=\"left\">27.24% '\n",
      "                                                '(207)</td><td '\n",
      "                                                'align=\"left\">Entity '\n",
      "                                                'Recognition</td><td '\n",
      "                                                'align=\"left\">13.04% '\n",
      "                                                '(27)</td><td '\n",
      "                                                'align=\"left\">45.41% '\n",
      "                                                '(94)</td><td '\n",
      "                                                'align=\"left\">41.55% '\n",
      "                                                '(86)</td></tr><tr><td '\n",
      "                                                'align=\"left\">32.11% '\n",
      "                                                '(244)</td><td '\n",
      "                                                'align=\"left\">Entity '\n",
      "                                                'Disambiguation</td><td '\n",
      "                                                'align=\"left\">12.30% '\n",
      "                                                '(30)</td><td '\n",
      "                                                'align=\"left\">26.23% '\n",
      "                                                '(64)</td><td '\n",
      "                                                'align=\"left\">61.48% '\n",
      "                                                '(150)</td></tr><tr><td '\n",
      "                                                'align=\"left\">1.97% '\n",
      "                                                '(15)</td><td '\n",
      "                                                'align=\"left\">Coreferencing</td><td '\n",
      "                                                'align=\"left\">13.33% '\n",
      "                                                '(2)</td><td '\n",
      "                                                'align=\"left\">13.33% '\n",
      "                                                '(2)</td><td '\n",
      "                                                'align=\"left\">73.33% '\n",
      "                                                '(11)</td></tr><tr><td '\n",
      "                                                'align=\"left\">13.68% '\n",
      "                                                '(104)</td><td '\n",
      "                                                'align=\"left\">Nonexistent '\n",
      "                                                'Relation</td><td '\n",
      "                                                'align=\"left\">23.08% '\n",
      "                                                '(24)</td><td '\n",
      "                                                'align=\"left\">29.81% '\n",
      "                                                '(31)</td><td '\n",
      "                                                'align=\"left\">47.12% '\n",
      "                                                '(49)</td></tr><tr><td '\n",
      "                                                'align=\"left\">9.21% '\n",
      "                                                '(70)</td><td '\n",
      "                                                'align=\"left\">Pattern Relation '\n",
      "                                                'Duality</td><td '\n",
      "                                                'align=\"left\">24.29% '\n",
      "                                                '(17)</td><td '\n",
      "                                                'align=\"left\">27.14% '\n",
      "                                                '(19)</td><td '\n",
      "                                                'align=\"left\">48.57% '\n",
      "                                                '(34)</td></tr><tr><td '\n",
      "                                                'align=\"left\">3.29% '\n",
      "                                                '(25)</td><td '\n",
      "                                                'align=\"left\">Swapped left and '\n",
      "                                                'right-hand entity</td><td '\n",
      "                                                'align=\"left\">28.00% '\n",
      "                                                '(7)</td><td '\n",
      "                                                'align=\"left\">24.00% '\n",
      "                                                '(6)</td><td '\n",
      "                                                'align=\"left\">48.00% '\n",
      "                                                '(12)</td></tr><tr><td '\n",
      "                                                'align=\"left\">3.03% '\n",
      "                                                '(23)</td><td '\n",
      "                                                'align=\"left\">Negation</td><td '\n",
      "                                                'align=\"left\">17.39% '\n",
      "                                                '(4)</td><td '\n",
      "                                                'align=\"left\">21.74% '\n",
      "                                                '(5)</td><td '\n",
      "                                                'align=\"left\">60.87% '\n",
      "                                                '(14)</td></tr><tr><td '\n",
      "                                                'align=\"left\">1.32% '\n",
      "                                                '(10)</td><td '\n",
      "                                                'align=\"left\">Factually '\n",
      "                                                'Wrong</td><td '\n",
      "                                                'align=\"left\">40.00% '\n",
      "                                                '(4)</td><td '\n",
      "                                                'align=\"left\">10.00% '\n",
      "                                                '(1)</td><td '\n",
      "                                                'align=\"left\">50.00% '\n",
      "                                                '(5)</td></tr></tbody></table> '},\n",
      "                              'offset': 38621,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': '  Percentage based on text genre   '\n",
      "                                      'Percentage Cause of error Encyclopaedic '\n",
      "                                      'Scientific Social     sources sources '\n",
      "                                      'sources   8.16% (62) Preprocessing '\n",
      "                                      '38.71% (24) 3.23% (2) 58.06% (36)   '\n",
      "                                      '27.24% (207) Entity Recognition 13.04% '\n",
      "                                      '(27) 45.41% (94) 41.55% (86)   32.11% '\n",
      "                                      '(244) Entity Disambiguation 12.30% (30) '\n",
      "                                      '26.23% (64) 61.48% (150)   1.97% (15) '\n",
      "                                      'Coreferencing 13.33% (2) 13.33% (2) '\n",
      "                                      '73.33% (11)   13.68% (104) Nonexistent '\n",
      "                                      'Relation 23.08% (24) 29.81% (31) 47.12% '\n",
      "                                      '(49)   9.21% (70) Pattern Relation '\n",
      "                                      'Duality 24.29% (17) 27.14% (19) 48.57% '\n",
      "                                      '(34)   3.29% (25) Swapped left and '\n",
      "                                      'right-hand entity 28.00% (7) 24.00% (6) '\n",
      "                                      '48.00% (12)   3.03% (23) Negation '\n",
      "                                      '17.39% (4) 21.74% (5) 60.87% (14)   '\n",
      "                                      '1.32% (10) Factually Wrong 40.00% (4) '\n",
      "                                      '10.00% (1) 50.00% (5)   '},\n",
      "                             {'annotations': [{'id': '16766',\n",
      "                                               'infons': {'accession': '@DISEASE_Sarcoidosis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D012507',\n",
      "                                                          'name': 'Sarcoidosis',\n",
      "                                                          'normalized': ['D012507'],\n",
      "                                                          'normalized_id': 'D012507',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 11,\n",
      "                                                              'offset': 39387}],\n",
      "                                               'text': 'sarcoidosis'},\n",
      "                                              {'id': '16767',\n",
      "                                               'infons': {'accession': '@DISEASE_Anemia',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D000740',\n",
      "                                                          'name': 'Anemia',\n",
      "                                                          'normalized': ['D000740'],\n",
      "                                                          'normalized_id': 'D000740',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 39357}],\n",
      "                                               'text': 'Anemia'}],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 39357,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Anemia is a common symptom of '\n",
      "                                      'sarcoidosis.  '},\n",
      "                             {'annotations': [{'id': '16768',\n",
      "                                               'infons': {'accession': '@DISEASE_Anemia',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D000740',\n",
      "                                                          'name': 'Anemia',\n",
      "                                                          'normalized': ['D000740'],\n",
      "                                                          'normalized_id': 'D000740',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 6,\n",
      "                                                              'offset': 39437}],\n",
      "                                               'text': 'anemia'},\n",
      "                                              {'id': '16769',\n",
      "                                               'infons': {'accession': '@DISEASE_Sarcoidosis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D012507',\n",
      "                                                          'name': 'Sarcoidosis',\n",
      "                                                          'normalized': ['D012507'],\n",
      "                                                          'normalized_id': 'D012507',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 11,\n",
      "                                                              'offset': 39422}],\n",
      "                                               'text': 'sarcoidosis'}],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 39402,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'A common symptom of sarcoidosis is '\n",
      "                                      'anemia.'},\n",
      "                             {'annotations': [{'id': '16770',\n",
      "                                               'infons': {'accession': '@DISEASE_Microcephaly',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D008831',\n",
      "                                                          'name': 'Microcephaly',\n",
      "                                                          'normalized': ['D008831'],\n",
      "                                                          'normalized_id': 'D008831',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 12,\n",
      "                                                              'offset': 41218}],\n",
      "                                               'text': 'Microcephaly'},\n",
      "                                              {'id': '16771',\n",
      "                                               'infons': {'accession': '@DISEASE_Epilepsy',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D004827',\n",
      "                                                          'name': 'Epilepsy',\n",
      "                                                          'normalized': ['D004827'],\n",
      "                                                          'normalized_id': 'D004827',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 8,\n",
      "                                                              'offset': 41232}],\n",
      "                                               'text': 'Epilepsy'},\n",
      "                                              {'id': '16772',\n",
      "                                               'infons': {'accession': '@DISEASE_Diabetes_Mellitus',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D003920',\n",
      "                                                          'name': 'Diabetes '\n",
      "                                                                  'Mellitus',\n",
      "                                                          'normalized': ['D003920'],\n",
      "                                                          'normalized_id': 'D003920',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 17,\n",
      "                                                              'offset': 41246}],\n",
      "                                               'text': 'Diabetes Syndrome'}],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 39445,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We analyzed the causes of error for all '\n",
      "                                      '760 facts annotated as incorrect from '\n",
      "                                      'the experimental setting using the full '\n",
      "                                      'information extraction pipeline and all '\n",
      "                                      'three text genres. This setting allows '\n",
      "                                      'us to compare the utility of the '\n",
      "                                      'different components as well as the '\n",
      "                                      'different genres. As seen in Table 7, '\n",
      "                                      'we categorize the errors as follows: '\n",
      "                                      'Preprocessing: At the start of the '\n",
      "                                      'pipeline, incorrect sentence '\n",
      "                                      'segmentation divided a text passage '\n",
      "                                      'into incomplete sentences, or left '\n",
      "                                      'multiple sentences undivided. This in '\n",
      "                                      'turn lead to incorrect parsing of '\n",
      "                                      'syntactic dependency graphs. In '\n",
      "                                      'addition, there were incorrectly parsed '\n",
      "                                      'DOM trees in Web portal documents. Not '\n",
      "                                      'surprisingly, almost all preprocessing '\n",
      "                                      'errors came from encyclopedic and '\n",
      "                                      'social sources due to their DOM tree '\n",
      "                                      'structure and poor language style, '\n",
      "                                      'respectively. Entity Recognition: '\n",
      "                                      'Certain entities were not correctly '\n",
      "                                      'recognized. Complex entities are '\n",
      "                                      'composed of multiple simple entities; '\n",
      "                                      'examples include muscle protein '\n",
      "                                      'breakdown recognized as muscle protein '\n",
      "                                      'and breakdown, or arrest of cystic '\n",
      "                                      'growth recognized as arrest and cystic '\n",
      "                                      'growth. Paraphrasing and misspelling '\n",
      "                                      'entities cause their textual '\n",
      "                                      'expressions to deviate from dictionary '\n",
      "                                      'entries. Idiomatic expressions were '\n",
      "                                      'incorrectly picked up as entities. For '\n",
      "                                      'instance, there is no actual physical '\n",
      "                                      'activity in the English idiom in the '\n",
      "                                      'long run. Entity Disambiguation: '\n",
      "                                      'Selecting an incorrect entity out of '\n",
      "                                      'multiple matching candidates caused '\n",
      "                                      'this error, primarily due to two '\n",
      "                                      'reasons. First, the type signatures of '\n",
      "                                      'our relations were not sufficient to '\n",
      "                                      'futher prune out mismatching entities '\n",
      "                                      'during fact extraction. Second, '\n",
      "                                      'colloquial terms not curated in the '\n",
      "                                      'UMLS dictionary were incorrectly '\n",
      "                                      'resolved. For example, meds for '\n",
      "                                      'medicines was disambiguated as the '\n",
      "                                      'entity Microcephaly, Epilepsy, and '\n",
      "                                      'Diabetes Syndrome. Coreferencing: Due '\n",
      "                                      'to the lack of coreference resolution, '\n",
      "                                      'correct entities were obscured by '\n",
      "                                      'phrases such as this protein or the '\n",
      "                                      'tunnel structure. Nonexistent relation: '\n",
      "                                      'Two entities might co-occur within the '\n",
      "                                      'same sentence without sharing a '\n",
      "                                      'relation. When a pattern occurrence '\n",
      "                                      'between such entities was nevertheless '\n",
      "                                      'extracted, it resulted in an '\n",
      "                                      'unsubstantiated relation. Pattern '\n",
      "                                      'Relation Duality: A pattern that can '\n",
      "                                      'express two relations was harvested but '\n",
      "                                      'assigned to an incorrect relation. For '\n",
      "                                      'example, the pattern mimic was '\n",
      "                                      'incorrectly assigned to the relation '\n",
      "                                      'isSymptom. Swapped left and right-hand '\n",
      "                                      'entity: The harvested fact was '\n",
      "                                      'incorrect because the left- and '\n",
      "                                      'right-hand entities were swapped. '\n",
      "                                      'Consider the example fact i s S y m p t '\n",
      "                                      'o m(A n e m i a,S a r c o i d o s i s), '\n",
      "                                      'which can be expressed by either '\n",
      "                                      'sentence:  '},\n",
      "                             {'annotations': [{'id': '16773',\n",
      "                                               'infons': {'accession': '@DISEASE_Osteoarthritis',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D010003',\n",
      "                                                          'name': 'Osteoarthritis',\n",
      "                                                          'normalized': ['D010003'],\n",
      "                                                          'normalized_id': 'D010003',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 14,\n",
      "                                                              'offset': 42678}],\n",
      "                                               'text': 'osteoarthritis'},\n",
      "                                              {'id': '16774',\n",
      "                                               'infons': {'accession': '@DISEASE_Heart_Diseases',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D006331',\n",
      "                                                          'name': 'Heart '\n",
      "                                                                  'Diseases',\n",
      "                                                          'normalized': ['D006331'],\n",
      "                                                          'normalized_id': 'D006331',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 13,\n",
      "                                                              'offset': 42727}],\n",
      "                                               'text': 'heart disease'},\n",
      "                                              {'id': '16775',\n",
      "                                               'infons': {'accession': '@DISEASE_Arthritis_Reactive',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D016918',\n",
      "                                                          'name': 'Arthritis '\n",
      "                                                                  'Reactive',\n",
      "                                                          'normalized': ['D016918'],\n",
      "                                                          'normalized_id': 'D016918',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 18,\n",
      "                                                              'offset': 42501}],\n",
      "                                               'text': 'reactive arthritis'}],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 42087,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'In both cases, the same pattern is a '\n",
      "                                      'common symptom of is extracted. In '\n",
      "                                      'sentence 2, however, an incorrect fact '\n",
      "                                      'would be extracted since the order in '\n",
      "                                      'which the entities occur is reversed. '\n",
      "                                      'Negation: This error was caused by not '\n",
      "                                      'detecting negation expressed in the '\n",
      "                                      'text. The word expressing the negation '\n",
      "                                      'may occur textually far away from the '\n",
      "                                      'entities, as in It is disputed whether '\n",
      "                                      'early antibiotic treatment prevents '\n",
      "                                      'reactive arthritis, and thus escaped '\n",
      "                                      'our pattern gathering method. In other '\n",
      "                                      'cases, the negation phrase will require '\n",
      "                                      'subtle semantic understanding to tease '\n",
      "                                      'out, as in Except for osteoarthritis, I '\n",
      "                                      'think my symptoms are all from heart '\n",
      "                                      'disease. Factually Wrong: Although our '\n",
      "                                      'methods successfully harvested a fact, '\n",
      "                                      'the underlying text evidence made a '\n",
      "                                      'wrong statement. Lessons learned: '\n",
      "                                      'Overall, this error analysis confirms '\n",
      "                                      'that scientific and encyclopedic '\n",
      "                                      'sources contain well-written texts that '\n",
      "                                      'are amenable to a text mining pipeline. '\n",
      "                                      'Social sources, with their poorer '\n",
      "                                      'quality of language style as well as '\n",
      "                                      'information content, were the biggest '\n",
      "                                      'contributor in almost all error '\n",
      "                                      'categories. Errors in entity '\n",
      "                                      'recognition and disambiguation '\n",
      "                                      'accounted for close to 60% of all '\n",
      "                                      'errors; overcoming them will require '\n",
      "                                      'better methods that go beyond a '\n",
      "                                      'dictionary, and incorporate deeper '\n",
      "                                      'linguistic and semantic understanding.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 43410,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Coverage'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab8.xml',\n",
      "                                         'id': 'Tab8',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table_caption'},\n",
      "                              'offset': 43419,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': ' Top-20 pairs of inter-connected '\n",
      "                                      'biomedical areas within KnowLife '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'file': 'Tab8.xml',\n",
      "                                         'id': 'Tab8',\n",
      "                                         'section_type': 'TABLE',\n",
      "                                         'type': 'table',\n",
      "                                         'xml': '<?xml version=\"1.0\" '\n",
      "                                                'encoding=\"UTF-8\"?> <table '\n",
      "                                                'frame=\"hsides\" '\n",
      "                                                'rules=\"groups\"><thead><tr><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Biomedical areas</bold> '\n",
      "                                                '</th><th align=\"left\"/><th '\n",
      "                                                'align=\"left\"> '\n",
      "                                                '<bold>Connections</bold> '\n",
      "                                                '</th></tr></thead><tbody><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Chemicals</td><td '\n",
      "                                                'align=\"left\">310482</td></tr><tr><td '\n",
      "                                                'align=\"left\">Chemicals</td><td '\n",
      "                                                'align=\"left\">Chemicals</td><td '\n",
      "                                                'align=\"left\">190160</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">36677</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Procedures</td><td '\n",
      "                                                'align=\"left\">14169</td></tr><tr><td '\n",
      "                                                'align=\"left\">Chemicals</td><td '\n",
      "                                                'align=\"left\">Physiology</td><td '\n",
      "                                                'align=\"left\">5397</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Genes</td><td '\n",
      "                                                'align=\"left\">3831</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Living '\n",
      "                                                'Beings</td><td '\n",
      "                                                'align=\"left\">2539</td></tr><tr><td '\n",
      "                                                'align=\"left\">Chemicals</td><td '\n",
      "                                                'align=\"left\">Drugs</td><td '\n",
      "                                                'align=\"left\">2455</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Anatomy</td><td '\n",
      "                                                'align=\"left\">2895</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Devices</td><td '\n",
      "                                                'align=\"left\">792</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Activities</td><td '\n",
      "                                                'align=\"left\">592</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Drugs</td><td '\n",
      "                                                'align=\"left\">511</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Objects</td><td '\n",
      "                                                'align=\"left\">505</td></tr><tr><td '\n",
      "                                                'align=\"left\">Chemicals</td><td '\n",
      "                                                'align=\"left\">Procedures</td><td '\n",
      "                                                'align=\"left\">544</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Physiology</td><td '\n",
      "                                                'align=\"left\">370</td></tr><tr><td '\n",
      "                                                'align=\"left\">Procedures</td><td '\n",
      "                                                'align=\"left\">Physiology</td><td '\n",
      "                                                'align=\"left\">123</td></tr><tr><td '\n",
      "                                                'align=\"left\">Procedures</td><td '\n",
      "                                                'align=\"left\">Living '\n",
      "                                                'Beings</td><td '\n",
      "                                                'align=\"left\">99</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Geographical '\n",
      "                                                'Areas</td><td '\n",
      "                                                'align=\"left\">82</td></tr><tr><td '\n",
      "                                                'align=\"left\">Genes</td><td '\n",
      "                                                'align=\"left\">Physiology</td><td '\n",
      "                                                'align=\"left\">51</td></tr><tr><td '\n",
      "                                                'align=\"left\">Disorders</td><td '\n",
      "                                                'align=\"left\">Phenomena</td><td '\n",
      "                                                'align=\"left\">50</td></tr></tbody></table> '},\n",
      "                              'offset': 43486,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Biomedical areas  Connections   '\n",
      "                                      'Disorders Chemicals 310482   Chemicals '\n",
      "                                      'Chemicals 190160   Disorders Disorders '\n",
      "                                      '36677   Disorders Procedures 14169   '\n",
      "                                      'Chemicals Physiology 5397   Disorders '\n",
      "                                      'Genes 3831   Disorders Living Beings '\n",
      "                                      '2539   Chemicals Drugs 2455   Disorders '\n",
      "                                      'Anatomy 2895   Disorders Devices 792   '\n",
      "                                      'Disorders Activities 592   Disorders '\n",
      "                                      'Drugs 511   Disorders Objects 505   '\n",
      "                                      'Chemicals Procedures 544   Disorders '\n",
      "                                      'Physiology 370   Procedures Physiology '\n",
      "                                      '123   Procedures Living Beings 99   '\n",
      "                                      'Disorders Geographical Areas 82   Genes '\n",
      "                                      'Physiology 51   Disorders Phenomena '\n",
      "                                      '50   '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 44054,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The overriding goal of KnowLife has '\n",
      "                                      'been to create a versatile KB that '\n",
      "                                      'spans many areas within the life '\n",
      "                                      'sciences. To illustrate which areas are '\n",
      "                                      'covered by KnowLife, we refer to the '\n",
      "                                      'semantic groups defined by. Table 8 '\n",
      "                                      'shows the number of acquired facts for '\n",
      "                                      'pairs of the thirteen different areas '\n",
      "                                      'inter-connected in our KB. This can be '\n",
      "                                      'seen as an indicator that we achieved '\n",
      "                                      'our goal at least to some extent.  '},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'RESULTS',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 44461,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The predominant number of facts '\n",
      "                                      'involves entities of the semantic group '\n",
      "                                      'Disorders, for two reasons. First, with '\n",
      "                                      'our choice of relations, disorders '\n",
      "                                      'appear in almost all type signatures. '\n",
      "                                      'Second, entities of type clinical '\n",
      "                                      'finding are covered by the group '\n",
      "                                      'Disorders, and these are frequent in '\n",
      "                                      'all text genres. However, this type '\n",
      "                                      'also includes diverse, non-disorder '\n",
      "                                      'entities such as pregnancy, which is '\n",
      "                                      'clearly not a disorder.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'title_1'},\n",
      "                              'offset': 44883,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Conclusions'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 44895,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Application benefit'},\n",
      "                             {'annotations': [{'id': '16776',\n",
      "                                               'infons': {'accession': None,\n",
      "                                                          'biotype': 'species',\n",
      "                                                          'database': 'ncbi_taxonomy',\n",
      "                                                          'identifier': '9606',\n",
      "                                                          'name': '9606',\n",
      "                                                          'normalized': [9606],\n",
      "                                                          'normalized_id': 9606,\n",
      "                                                          'type': 'Species',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 7,\n",
      "                                                              'offset': 45187}],\n",
      "                                               'text': 'patient'}],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 44915,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'To showcase the usefulness of KnowLife, '\n",
      "                                      'we developed a health portal '\n",
      "                                      '(http://knowlife.mpi-inf.mpg.de) that '\n",
      "                                      'allows interactive exploration of the '\n",
      "                                      'harvested facts and their input '\n",
      "                                      'sources. The KnowLife portal supports a '\n",
      "                                      'number of use cases for different '\n",
      "                                      'information needs. A patient may wish '\n",
      "                                      'to find out the side effects of a '\n",
      "                                      'specific drug, by searching for the '\n",
      "                                      'drug name and browsing the SideEffect '\n",
      "                                      'facts and their provenance. A physician '\n",
      "                                      'may want to \"speed read\" publications '\n",
      "                                      'and online discussions on treatment '\n",
      "                                      'options for an unfamiliar disease. '\n",
      "                                      'Provenance information is vital here, '\n",
      "                                      'as the physician would want to consider '\n",
      "                                      'the recency and authority of the '\n",
      "                                      'sources for certain statements. The '\n",
      "                                      'health portal also provides a function '\n",
      "                                      'for on-the-fly annotation of new text '\n",
      "                                      'from publications or social media, '\n",
      "                                      'leveraging known patterns to highlight '\n",
      "                                      'any relations found.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'title_2'},\n",
      "                              'offset': 45784,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Future work'},\n",
      "                             {'annotations': [{'id': '16777',\n",
      "                                               'infons': {'accession': '@DISEASE_Paraplegia',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D010264',\n",
      "                                                          'name': 'Paraplegia',\n",
      "                                                          'normalized': ['D010264'],\n",
      "                                                          'normalized_id': 'D010264',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 3,\n",
      "                                                              'offset': 46014}],\n",
      "                                               'text': 'leg'},\n",
      "                                              {'id': '16778',\n",
      "                                               'infons': {'accession': '@DISEASE_Amputation_Congenital',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:C565682',\n",
      "                                                          'name': 'Amputation '\n",
      "                                                                  'Congenital',\n",
      "                                                          'normalized': ['C565682'],\n",
      "                                                          'normalized_id': 'C565682',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 13,\n",
      "                                                              'offset': 45994}],\n",
      "                                               'text': 'amputation of'}],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 45796,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'In the future, we plan to improve the '\n",
      "                                      'entity recognition to accommodate a '\n",
      "                                      'wider variety of entities beyond those '\n",
      "                                      'in UMLS. For instance, colloquial usage '\n",
      "                                      '(meds for medicines) and composite '\n",
      "                                      'entities (amputation of right leg) are '\n",
      "                                      'not yet addressed. Entities within UMLS '\n",
      "                                      'also require more sophisticated '\n",
      "                                      'disambiguation. For instance, the text '\n",
      "                                      'occurrence stress may be correctly '\n",
      "                                      'distinguished between the brand name of '\n",
      "                                      'a drug and the psychological feeling.'},\n",
      "                             {'annotations': [{'id': '16779',\n",
      "                                               'infons': {'accession': '@DISEASE_Fever',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D005334',\n",
      "                                                          'name': 'Fever',\n",
      "                                                          'normalized': ['D005334'],\n",
      "                                                          'normalized_id': 'D005334',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 5,\n",
      "                                                              'offset': 46453}],\n",
      "                                               'text': 'Fever'},\n",
      "                                              {'id': '16780',\n",
      "                                               'infons': {'accession': '@DISEASE_Lupus_Erythematosus_Systemic',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D008180',\n",
      "                                                          'name': 'Lupus '\n",
      "                                                                  'Erythematosus '\n",
      "                                                                  'Systemic',\n",
      "                                                          'normalized': ['D008180'],\n",
      "                                                          'normalized_id': 'D008180',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 11,\n",
      "                                                              'offset': 46475}],\n",
      "                                               'text': 'Lupus Flare'}],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 46247,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Finally, we would like to address the '\n",
      "                                      'challenge of mining and representing '\n",
      "                                      'the context of harvested facts. Binary '\n",
      "                                      'relations are often not sufficient to '\n",
      "                                      'express medical knowledge. For example, '\n",
      "                                      'the statement Fever is a symptom of '\n",
      "                                      'Lupus Flare during pregnancy cannot be '\n",
      "                                      'suitably represented by a binary fact.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'paragraph'},\n",
      "                              'offset': 46553,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'We plan to cope with such statements by '\n",
      "                                      'extracting ternary and higher-arity '\n",
      "                                      'relations, with appropriate extensions '\n",
      "                                      'of both pattern-based extraction and '\n",
      "                                      'consistency reasoning.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'title_1'},\n",
      "                              'offset': 46728,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Additional files'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'footnote'},\n",
      "                              'offset': 46745,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Competing interests'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'footnote'},\n",
      "                              'offset': 46765,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The authors declare that they have no '\n",
      "                                      'competing interests.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'footnote'},\n",
      "                              'offset': 46824,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': \"Authors' contributions\"},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'CONCL',\n",
      "                                         'type': 'footnote'},\n",
      "                              'offset': 46849,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'AS developed the entity recognition and '\n",
      "                                      'prepared the input sources. PE '\n",
      "                                      'implemented the information extraction '\n",
      "                                      'pipeline, integrated all components and '\n",
      "                                      'designed the evaluation. AS and PE '\n",
      "                                      'conducted the evaluation and drafted '\n",
      "                                      'the manuscript. The work was advised by '\n",
      "                                      'GW. All authors read and approved the '\n",
      "                                      'final manuscript.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'REF',\n",
      "                                         'type': 'title'},\n",
      "                              'offset': 47167,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'References'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Barbosa;given-names:D',\n",
      "                                         'name_1': 'surname:Wang;given-names:H',\n",
      "                                         'name_2': 'surname:Yu;given-names:C',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of '\n",
      "                                                   'International Conference '\n",
      "                                                   'On Data Engineering (ICDE)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2013'},\n",
      "                              'offset': 47178,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Shallow information extraction for the '\n",
      "                                      'knowledge web'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Suchanek;given-names:F',\n",
      "                                         'name_1': 'surname:Weikum;given-names:G',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of '\n",
      "                                                   'International Conference '\n",
      "                                                   'On Data Engineering (ICDE)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2013'},\n",
      "                              'offset': 47231,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Knowledge harvesting from text and web '\n",
      "                                      'sources'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '167',\n",
      "                                         'issue': '2',\n",
      "                                         'lpage': '95',\n",
      "                                         'name_0': 'surname:Lehmann;given-names:J',\n",
      "                                         'name_1': 'surname:Isele;given-names:R',\n",
      "                                         'name_2': 'surname:Jakob;given-names:M',\n",
      "                                         'name_3': 'surname:Jentzsch;given-names:A',\n",
      "                                         'name_4': 'surname:Kontokostas;given-names:D',\n",
      "                                         'name_5': 'surname:Mendes;given-names:PN',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Semantic Web J',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '6',\n",
      "                                         'year': '2013'},\n",
      "                              'offset': 47278,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'DBpedia - a large-scale, multilingual '\n",
      "                                      'knowledge base extracted from '\n",
      "                                      'Wikipedia'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Hoffart;given-names:J',\n",
      "                                         'name_1': 'surname:Suchanek;given-names:F',\n",
      "                                         'name_2': 'surname:Berberich;given-names:K',\n",
      "                                         'name_3': 'surname:Weikum;given-names:G',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Special '\n",
      "                                                   'issue of the Artificial '\n",
      "                                                   'Intelligence Journal',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2013'},\n",
      "                              'offset': 47358,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'YAGO2: A Spatially and Temporally '\n",
      "                                      'Enhanced Knowledge Base from Wikipedia'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '575',\n",
      "                                         'issue': '18',\n",
      "                                         'lpage': '81',\n",
      "                                         'name_0': 'surname:Pyysalo;given-names:S',\n",
      "                                         'name_1': 'surname:Ohta;given-names:T',\n",
      "                                         'name_2': 'surname:Miwa;given-names:M',\n",
      "                                         'name_3': 'surname:Cho;given-names:H-C',\n",
      "                                         'name_4': 'surname:Tsujii;given-names:J',\n",
      "                                         'name_5': 'surname:Ananiadou;given-names:S',\n",
      "                                         'pub-id_doi': '10.1093/bioinformatics/bts407',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '28',\n",
      "                                         'year': '2012'},\n",
      "                              'offset': 47431,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Event extraction across multiple levels '\n",
      "                                      'of biological organization'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '4',\n",
      "                                         'issue': 'Suppl 8',\n",
      "                                         'name_0': 'surname:Arighi;given-names:C',\n",
      "                                         'name_1': 'surname:Roberts;given-names:P',\n",
      "                                         'name_2': 'surname:Agarwal;given-names:S',\n",
      "                                         'name_3': 'surname:Bhattacharya;given-names:S',\n",
      "                                         'name_4': 'surname:Cesareni;given-names:G',\n",
      "                                         'name_5': 'surname:Chatr-aryamontri;given-names:A',\n",
      "                                         'pub-id_doi': '10.1186/1471-2105-12-S8-S4',\n",
      "                                         'pub-id_pmid': '21208430',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'BMC Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '12',\n",
      "                                         'year': '2011'},\n",
      "                              'offset': 47498,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'BioCreative III interactive task: An '\n",
      "                                      'overview'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '10',\n",
      "                                         'issue': '1',\n",
      "                                         'name_0': 'surname:Kim;given-names:JD',\n",
      "                                         'name_1': 'surname:Ohta;given-names:T',\n",
      "                                         'name_2': 'surname:Tsujii;given-names:J',\n",
      "                                         'pub-id_doi': '10.1186/1471-2105-9-10',\n",
      "                                         'pub-id_pmid': '18182099',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'BMC Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '9',\n",
      "                                         'year': '2008'},\n",
      "                              'offset': 47544,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Corpus annotation for mining biomedical '\n",
      "                                      'events from literature'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '414',\n",
      "                                         'issue': '4',\n",
      "                                         'lpage': '7',\n",
      "                                         'name_0': 'surname:Whirl-Carrillo;given-names:M',\n",
      "                                         'name_1': 'surname:McDonagh;given-names:E',\n",
      "                                         'name_2': 'surname:Hebert;given-names:J',\n",
      "                                         'name_3': 'surname:Gong;given-names:L',\n",
      "                                         'name_4': 'surname:Sangkuhl;given-names:K',\n",
      "                                         'pub-id_doi': '10.1038/clpt.2012.96',\n",
      "                                         'pub-id_pmid': '22992668',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Clinical Pharmacol Ther',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '92',\n",
      "                                         'year': '2012'},\n",
      "                              'offset': 47607,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Pharmacogenomics knowledge for '\n",
      "                                      'personalized medicine'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '1188',\n",
      "                                         'issue': '21',\n",
      "                                         'lpage': '98',\n",
      "                                         'name_0': 'surname:Williams;given-names:AJ',\n",
      "                                         'name_1': 'surname:Harland;given-names:L',\n",
      "                                         'name_2': 'surname:Groth;given-names:P',\n",
      "                                         'name_3': 'surname:Pettifer;given-names:S',\n",
      "                                         'name_4': 'surname:Chichester;given-names:C',\n",
      "                                         'name_5': 'surname:Willighagen;given-names:EL',\n",
      "                                         'pub-id_doi': '10.1016/j.drudis.2012.05.016',\n",
      "                                         'pub-id_pmid': '22683805',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Drug Discov Today',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '17',\n",
      "                                         'year': '2012'},\n",
      "                              'offset': 47660,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Open PHACTS: semantic interoperability '\n",
      "                                      'for drug discovery'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Buyko;given-names:E',\n",
      "                                         'name_1': 'surname:Faessler;given-names:E',\n",
      "                                         'name_2': 'surname:Wermter;given-names:J',\n",
      "                                         'name_3': 'surname:Hahn;given-names:U',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Workshop on '\n",
      "                                                   'Current Trends in '\n",
      "                                                   'Biomedical Natural '\n",
      "                                                   'Language Processing '\n",
      "                                                   '(BioNLP): Shared Task',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2009'},\n",
      "                              'offset': 47718,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Event extraction from trimmed '\n",
      "                                      'dependency graphs'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '131',\n",
      "                                         'issue': '1',\n",
      "                                         'lpage': '46',\n",
      "                                         'name_0': 'surname:Miwa;given-names:M',\n",
      "                                         'name_1': 'surname:Saetre;given-names:R',\n",
      "                                         'name_2': 'surname:Kim;given-names:J-D',\n",
      "                                         'name_3': 'surname:Tsujii;given-names:J',\n",
      "                                         'pub-id_doi': '10.1142/S0219720010004586',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'J Bioinformatics Comput '\n",
      "                                                   'Biol',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '8',\n",
      "                                         'year': '2010'},\n",
      "                              'offset': 47766,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Event extraction with complex event '\n",
      "                                      'classification using rich features'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Bjorne;given-names:J',\n",
      "                                         'name_1': 'surname:Salakoski;given-names:T',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Workshop on '\n",
      "                                                   'Current Trends in '\n",
      "                                                   'Biomedical Natural '\n",
      "                                                   'Language Processing '\n",
      "                                                   '(BioNLP): Shared Task',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2011'},\n",
      "                              'offset': 47837,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Generalizing biomedical event '\n",
      "                                      'extraction'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '1',\n",
      "                                         'issue': 'S8',\n",
      "                                         'name_0': 'surname:Krallinger;given-names:M',\n",
      "                                         'name_1': 'surname:Izarzugaza;given-names:JMG',\n",
      "                                         'name_2': 'surname:Penagos;given-names:CR',\n",
      "                                         'name_3': 'surname:Valencia;given-names:A',\n",
      "                                         'pub-id_doi': '10.1186/1471-2105-10-S8-S1',\n",
      "                                         'pub-id_pmid': '19118496',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'BMC Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '10',\n",
      "                                         'year': '2009'},\n",
      "                              'offset': 47878,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Extraction of human kinase mutations '\n",
      "                                      'from literature, databases and '\n",
      "                                      'genotyping studies'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Rosario;given-names:B',\n",
      "                                         'name_1': 'surname:Hearst;given-names:MA',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Annual '\n",
      "                                                   'Meeting on Association for '\n",
      "                                                   'Computational Linguistics '\n",
      "                                                   '(ACL)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2004'},\n",
      "                              'offset': 47965,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Classifying semantic relations in '\n",
      "                                      'bioscience texts'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '207',\n",
      "                                         'issue': '1',\n",
      "                                         'name_0': 'surname:Bundschus;given-names:M',\n",
      "                                         'name_1': 'surname:Dejori;given-names:M',\n",
      "                                         'name_2': 'surname:Stetter;given-names:M',\n",
      "                                         'name_3': 'surname:Tresp;given-names:V',\n",
      "                                         'name_4': 'surname:Kriegel;given-names:HP',\n",
      "                                         'pub-id_doi': '10.1186/1471-2105-9-207',\n",
      "                                         'pub-id_pmid': '18433469',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'BMC Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '9',\n",
      "                                         'year': '2008'},\n",
      "                              'offset': 48016,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Extraction of semantic biomedical '\n",
      "                                      'relations from text using conditional '\n",
      "                                      'random fields'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '806',\n",
      "                                         'issue': '5',\n",
      "                                         'lpage': '13',\n",
      "                                         'name_0': 'surname:Sun;given-names:W',\n",
      "                                         'name_1': 'surname:Rumshisky;given-names:A',\n",
      "                                         'name_2': 'surname:Uzuner;given-names:O',\n",
      "                                         'pub-id_doi': '10.1136/amiajnl-2013-001628',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'J Am Med Informatics Assoc',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '20',\n",
      "                                         'year': '2013'},\n",
      "                              'offset': 48102,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Evaluating temporal relations in '\n",
      "                                      'clinical text: 2012 i2b2 challenge'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'REF', 'type': 'ref'},\n",
      "                              'offset': 48170,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Bravo A, Cases M, Queralt-Rosinach N, '\n",
      "                                      'Sanz F, Furlong L. A knowledge-driven '\n",
      "                                      'approach to extract disease-related '\n",
      "                                      'biomarkers from the literature. BioMed '\n",
      "                                      'Res Int. 2014. article ID: 253128.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'REF', 'type': 'ref'},\n",
      "                              'offset': 48356,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Chun HW, Tsuruoka Y, Kim JD, Shiba R, '\n",
      "                                      'Nagata N, Hishiki T, et al. Extraction '\n",
      "                                      'of gene-disease relations from medline '\n",
      "                                      'using domain dictionaries and machine '\n",
      "                                      'learning. In: Proceedings of Pacific '\n",
      "                                      'Symposium of Biocomputing: 2006. p. '\n",
      "                                      '4-15.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '457',\n",
      "                                         'issue': '5',\n",
      "                                         'lpage': '68',\n",
      "                                         'name_0': 'surname:Leroy;given-names:G',\n",
      "                                         'name_1': 'surname:Chen;given-names:H',\n",
      "                                         'pub-id_doi': '10.1002/asi.20135',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'J Am Soc Inform Sci '\n",
      "                                                   'Technol.',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '56',\n",
      "                                         'year': '2005'},\n",
      "                              'offset': 48591,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Genescene: An ontology-enhanced '\n",
      "                                      'integration of linguistic and '\n",
      "                                      'co-occurrence based relations in '\n",
      "                                      'biomedical texts'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Rindflesch;given-names:TC',\n",
      "                                         'name_1': 'surname:Libbus;given-names:B',\n",
      "                                         'name_2': 'surname:Hristovski;given-names:D',\n",
      "                                         'name_3': 'surname:Aronson;given-names:AR',\n",
      "                                         'name_4': 'surname:Kilicoglu;given-names:H',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of American '\n",
      "                                                   'Medical Informatics '\n",
      "                                                   'Association (AMIA) Annual '\n",
      "                                                   'Symposium',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2003'},\n",
      "                              'offset': 48703,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Semantic relations asserting the '\n",
      "                                      'etiology of genetic diseases'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '1925',\n",
      "                                         'issue': '16',\n",
      "                                         'lpage': '33',\n",
      "                                         'name_0': 'surname:Good;given-names:BM',\n",
      "                                         'name_1': 'surname:Su;given-names:AI',\n",
      "                                         'pub-id_doi': '10.1093/bioinformatics/btt333',\n",
      "                                         'pub-id_pmid': '23782614',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '29',\n",
      "                                         'year': '2013'},\n",
      "                              'offset': 48765,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Crowdsourcing for bioinformatics'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '187',\n",
      "                                         'issue': '1',\n",
      "                                         'lpage': '203',\n",
      "                                         'name_0': 'surname:Ranard;given-names:BL',\n",
      "                                         'name_1': 'surname:Ha;given-names:YP',\n",
      "                                         'name_2': 'surname:Meisel;given-names:ZF',\n",
      "                                         'name_3': 'surname:Asch;given-names:DA',\n",
      "                                         'name_4': 'surname:Hill;given-names:SS',\n",
      "                                         'name_5': 'surname:Becker;given-names:LB',\n",
      "                                         'pub-id_doi': '10.1007/s11606-013-2536-8',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'J General Intern Med',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '29',\n",
      "                                         'year': '2014'},\n",
      "                              'offset': 48798,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Crowdsourcing-harnessing the masses to '\n",
      "                                      'advance health and medicine, a '\n",
      "                                      'systematic review'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'section_type': 'REF', 'type': 'ref'},\n",
      "                              'offset': 48888,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Burger JD, Doughty E, Khare R, Wei C-H, '\n",
      "                                      'Mishra R, Aberdeen J, et al.Hybrid '\n",
      "                                      'curation of gene-mutation relations '\n",
      "                                      'combining automated extraction and '\n",
      "                                      'crowdsourcing. Database. 2014; 2014. '\n",
      "                                      'article ID: bau094.'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Aroyo;given-names:L',\n",
      "                                         'name_1': 'surname:Welty;given-names:C',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'AAAI Fall Symposium Series',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2013'},\n",
      "                              'offset': 49093,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Measuring crowd truth for medical '\n",
      "                                      'relation extraction'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '78',\n",
      "                                         'issue': '1',\n",
      "                                         'name_0': 'surname:Hunter;given-names:L',\n",
      "                                         'name_1': 'surname:Lu;given-names:Z',\n",
      "                                         'name_2': 'surname:Firby;given-names:J',\n",
      "                                         'name_3': 'surname:Baumgartner;given-names:W',\n",
      "                                         'name_4': 'surname:Johnson;given-names:H',\n",
      "                                         'name_5': 'surname:Ogren;given-names:P',\n",
      "                                         'pub-id_doi': '10.1186/1471-2105-9-78',\n",
      "                                         'pub-id_pmid': '18237434',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'BMC Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '9',\n",
      "                                         'year': '2008'},\n",
      "                              'offset': 49147,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'OpenDMAP: An open source, '\n",
      "                                      'ontology-driven concept analysis '\n",
      "                                      'engine, with applications to capturing '\n",
      "                                      'knowledge regarding protein transport, '\n",
      "                                      'protein interactions and '\n",
      "                                      'cell-type-specific gene expression'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Torii;given-names:M',\n",
      "                                         'name_1': 'surname:Arighi;given-names:CN',\n",
      "                                         'name_2': 'surname:Wang;given-names:Q',\n",
      "                                         'name_3': 'surname:Wu;given-names:CH',\n",
      "                                         'name_4': 'surname:Vijay-Shanker;given-names:K',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of '\n",
      "                                                   'International Conference '\n",
      "                                                   'on Bioinformatics, '\n",
      "                                                   'Computational Biology and '\n",
      "                                                   'Biomedical Informatics '\n",
      "                                                   '(BCB)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2013'},\n",
      "                              'offset': 49344,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Text mining of protein phosphorylation '\n",
      "                                      'information using a generalizable '\n",
      "                                      'rule-based approach'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '309',\n",
      "                                         'issue': '11',\n",
      "                                         'name_0': 'surname:Muller;given-names:HM',\n",
      "                                         'name_1': 'surname:Kenny;given-names:EE',\n",
      "                                         'name_2': 'surname:Sternberg;given-names:PW',\n",
      "                                         'pub-id_doi': '10.1371/journal.pbio.0020309',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'PLoS Biology',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '2',\n",
      "                                         'year': '2004'},\n",
      "                              'offset': 49437,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Textpresso: an ontology-based '\n",
      "                                      'information retrieval and extraction '\n",
      "                                      'system for biological literature'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '155',\n",
      "                                         'issue': '1',\n",
      "                                         'name_0': 'surname:Wattarujeekrit;given-names:T',\n",
      "                                         'name_1': 'surname:Shah;given-names:P',\n",
      "                                         'name_2': 'surname:Collier;given-names:N',\n",
      "                                         'pub-id_doi': '10.1186/1471-2105-5-155',\n",
      "                                         'pub-id_pmid': '15494078',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'BMC Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '5',\n",
      "                                         'year': '2004'},\n",
      "                              'offset': 49537,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'PASBio: predicate-argument structures '\n",
      "                                      'for event extraction in molecular '\n",
      "                                      'biology'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': 'i264',\n",
      "                                         'issue': '13',\n",
      "                                         'lpage': '72',\n",
      "                                         'name_0': 'surname:Kolarik;given-names:C',\n",
      "                                         'name_1': 'surname:Hofmann-Apitius;given-names:M',\n",
      "                                         'name_2': 'surname:Zimmermann;given-names:M',\n",
      "                                         'name_3': 'surname:Fluck;given-names:J',\n",
      "                                         'pub-id_doi': '10.1093/bioinformatics/btm196',\n",
      "                                         'pub-id_pmid': '17646305',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '23',\n",
      "                                         'year': '2007'},\n",
      "                              'offset': 49617,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Identification of new drug '\n",
      "                                      'classification terms in textual '\n",
      "                                      'resources'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Hearst;given-names:M',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of the 14th '\n",
      "                                                   'Conference on '\n",
      "                                                   'Computational Linguistics '\n",
      "                                                   '(CoLing)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '1992'},\n",
      "                              'offset': 49686,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Automatic acquisition of hyponyms from '\n",
      "                                      'large text corpora'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '462',\n",
      "                                         'issue': '6',\n",
      "                                         'lpage': '77',\n",
      "                                         'name_0': 'surname:Rindflesch;given-names:TC',\n",
      "                                         'name_1': 'surname:Fiszman;given-names:M',\n",
      "                                         'pub-id_doi': '10.1016/j.jbi.2003.11.003',\n",
      "                                         'pub-id_pmid': '14759819',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'J Biomed Inform',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '36',\n",
      "                                         'year': '2003'},\n",
      "                              'offset': 49744,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The interaction of domain knowledge and '\n",
      "                                      'linguistic structure in natural '\n",
      "                                      'language processing: interpreting '\n",
      "                                      'hypernymic propositions in biomedical '\n",
      "                                      'text'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '45',\n",
      "                                         'issue': '0',\n",
      "                                         'lpage': '63',\n",
      "                                         'name_0': 'surname:Poria;given-names:S',\n",
      "                                         'name_1': 'surname:Cambria;given-names:E',\n",
      "                                         'name_2': 'surname:Winterstein;given-names:G',\n",
      "                                         'name_3': 'surname:Huang;given-names:G-B',\n",
      "                                         'pub-id_doi': '10.1016/j.knosys.2014.05.005',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Knowledge-Based Syst',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '69',\n",
      "                                         'year': '2014'},\n",
      "                              'offset': 49893,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Sentic patterns: dependency-based rules '\n",
      "                                      'for concept-level sentiment analysis'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '585',\n",
      "                                         'issue': 'W1',\n",
      "                                         'lpage': '91',\n",
      "                                         'name_0': 'surname:Thomas;given-names:P',\n",
      "                                         'name_1': 'surname:Starlinger;given-names:J',\n",
      "                                         'name_2': 'surname:Vowinkel;given-names:A',\n",
      "                                         'name_3': 'surname:Arzt;given-names:S',\n",
      "                                         'name_4': 'surname:Leser;given-names:U',\n",
      "                                         'pub-id_doi': '10.1093/nar/gks563',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Nucleic Acids Res',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '40',\n",
      "                                         'year': '2012'},\n",
      "                              'offset': 49970,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'GeneView: a comprehensive semantic '\n",
      "                                      'search engine for PubMed'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '105',\n",
      "                                         'issue': '1',\n",
      "                                         'name_0': 'surname:Xu;given-names:R',\n",
      "                                         'name_1': 'surname:Li;given-names:L',\n",
      "                                         'name_2': 'surname:Wang;given-names:Q',\n",
      "                                         'pub-id_doi': '10.1186/1471-2105-15-105',\n",
      "                                         'pub-id_pmid': '24725842',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'BMC Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '15',\n",
      "                                         'year': '2014'},\n",
      "                              'offset': 50030,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'dRiskKB: a large-scale disease-disease '\n",
      "                                      'risk relationship knowledge base '\n",
      "                                      'constructed from biomedical text'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Brin;given-names:S',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Selected Papers from the '\n",
      "                                                   'International Workshop on '\n",
      "                                                   'The World Wide Web and '\n",
      "                                                   'Databases (WebDB)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '1998'},\n",
      "                              'offset': 50135,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Extracting patterns and relations from '\n",
      "                                      'the World Wide Web'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Agichtein;given-names:E',\n",
      "                                         'name_1': 'surname:Gravano;given-names:L',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of the Fifth '\n",
      "                                                   'ACM Conference on Digital '\n",
      "                                                   'Libraries (DL)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2000'},\n",
      "                              'offset': 50193,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Snowball: Extracting relations from '\n",
      "                                      'large plain-text collections'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Suchanek;given-names:F',\n",
      "                                         'name_1': 'surname:Sozio;given-names:M',\n",
      "                                         'name_2': 'surname:Weikum;given-names:G',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of '\n",
      "                                                   'International World Wide '\n",
      "                                                   'Web Conference (WWW)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2009'},\n",
      "                              'offset': 50258,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'SOFIE: A self-organizing framework for '\n",
      "                                      'information extraction'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Nakashole;given-names:N',\n",
      "                                         'name_1': 'surname:Theobald;given-names:M',\n",
      "                                         'name_2': 'surname:Weikum;given-names:G',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of '\n",
      "                                                   'International Conference '\n",
      "                                                   'on Web Search and Data '\n",
      "                                                   'Mining (WSDM)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2011'},\n",
      "                              'offset': 50320,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Scalable knowledge harvesting with high '\n",
      "                                      'precision and high recall'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Carlson;given-names:A',\n",
      "                                         'name_1': 'surname:Betteridge;given-names:J',\n",
      "                                         'name_2': 'surname:Kisiel;given-names:B',\n",
      "                                         'name_3': 'surname:Settles;given-names:B',\n",
      "                                         'name_4': 'surname:Hruschka '\n",
      "                                                   'Jr;given-names:ER',\n",
      "                                         'name_5': 'surname:Mitchell;given-names:TM',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of the '\n",
      "                                                   'Association for the '\n",
      "                                                   'Advancement of Artificial '\n",
      "                                                   'Intelligence (AAAI) '\n",
      "                                                   'Conference',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2010'},\n",
      "                              'offset': 50386,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Toward an architecture for never-ending '\n",
      "                                      'language learning'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Nebot;given-names:V',\n",
      "                                         'name_1': 'surname:Ye;given-names:M',\n",
      "                                         'name_2': 'surname:Albrecht;given-names:M',\n",
      "                                         'name_3': 'surname:Eom;given-names:J-H',\n",
      "                                         'name_4': 'surname:Weikum;given-names:G',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of '\n",
      "                                                   'International World Wide '\n",
      "                                                   'Web Conference (WWW)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2011'},\n",
      "                              'offset': 50444,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'DIDO: A disease-determinants ontology '\n",
      "                                      'from Web sources'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Movshovitz-Attias;given-names:D',\n",
      "                                         'name_1': 'surname:Cohen;given-names:WW',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Workshop on '\n",
      "                                                   'Biomedical Natural '\n",
      "                                                   'Language Processing '\n",
      "                                                   '(BioNLP)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2012'},\n",
      "                              'offset': 50499,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Bootstrapping biomedical ontologies for '\n",
      "                                      'scientific text using NELL'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '229',\n",
      "                                         'issue': '3',\n",
      "                                         'lpage': '36',\n",
      "                                         'name_0': 'surname:Aronson;given-names:AR',\n",
      "                                         'name_1': 'surname:Lang;given-names:FM',\n",
      "                                         'pub-id_doi': '10.1136/jamia.2009.002733',\n",
      "                                         'pub-id_pmid': '20442139',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'J Am Med Inform Assoc',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '17',\n",
      "                                         'year': '2010'},\n",
      "                              'offset': 50566,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'An overview of MetaMap: historical '\n",
      "                                      'perspective and recent advances'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '254',\n",
      "                                         'issue': '2',\n",
      "                                         'lpage': '60',\n",
      "                                         'name_0': 'surname:Harmston;given-names:N',\n",
      "                                         'name_1': 'surname:Filsell;given-names:W',\n",
      "                                         'name_2': 'surname:Stumpf;given-names:M',\n",
      "                                         'pub-id_doi': '10.1093/bioinformatics/btr640',\n",
      "                                         'pub-id_pmid': '22135416',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Bioinformatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '28',\n",
      "                                         'year': '2012'},\n",
      "                              'offset': 50633,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Which species is it? Species-driven '\n",
      "                                      'gene name disambiguation using random '\n",
      "                                      'walks over a mixture of adjacency '\n",
      "                                      'matrices'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '842',\n",
      "                                         'issue': '5',\n",
      "                                         'lpage': '9',\n",
      "                                         'name_0': 'surname:Chasin;given-names:R',\n",
      "                                         'name_1': 'surname:Rumshisky;given-names:A',\n",
      "                                         'name_2': 'surname:Uzuner;given-names:O',\n",
      "                                         'name_3': 'surname:Szolovits;given-names:P',\n",
      "                                         'pub-id_doi': '10.1136/amiajnl-2013-002133',\n",
      "                                         'pub-id_pmid': '24441986',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'J Am Med Inform Assoc',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '21',\n",
      "                                         'year': '2014'},\n",
      "                              'offset': 50750,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Word sense disambiguation in the '\n",
      "                                      'clinical domain: a comparison of '\n",
      "                                      'knowledge-rich and knowledge-poor '\n",
      "                                      'unsupervised methods'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Vydiswaran;given-names:VGV',\n",
      "                                         'name_1': 'surname:Zhai;given-names:C',\n",
      "                                         'name_2': 'surname:Roth;given-names:D',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Workshop on '\n",
      "                                                   'Data Mining for Medicine '\n",
      "                                                   'and Healthcare (DMMH)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2011'},\n",
      "                              'offset': 50871,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Gauging the Internet doctor: Ranking '\n",
      "                                      'medical claims based on community '\n",
      "                                      'knowledge'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Mukherjee;given-names:S',\n",
      "                                         'name_1': 'surname:Weikum;given-names:G',\n",
      "                                         'name_2': 'surname:Danescu-Niculescu-Mizil;given-names:C',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Conference '\n",
      "                                                   'on Knowledge Discovery and '\n",
      "                                                   'Data Mining (KDD)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2014'},\n",
      "                              'offset': 50952,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'People on drugs: Credibility of user '\n",
      "                                      'statements in health communities'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '239',\n",
      "                                         'issue': '2',\n",
      "                                         'lpage': '46',\n",
      "                                         'name_0': 'surname:White;given-names:RW',\n",
      "                                         'name_1': 'surname:Harpaz;given-names:R',\n",
      "                                         'name_2': 'surname:Shah;given-names:NH',\n",
      "                                         'name_3': 'surname:DuMouchel;given-names:W',\n",
      "                                         'name_4': 'surname:Horvitz;given-names:E',\n",
      "                                         'pub-id_doi': '10.1038/clpt.2014.77',\n",
      "                                         'pub-id_pmid': '24713590',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Clin Pharmacol Ther',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '96',\n",
      "                                         'year': '2014'},\n",
      "                              'offset': 51022,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Toward enhanced pharmacovigilance using '\n",
      "                                      'patient-generated data on the Internet'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Ernst;given-names:P',\n",
      "                                         'name_1': 'surname:Meng;given-names:C',\n",
      "                                         'name_2': 'surname:Siu;given-names:A',\n",
      "                                         'name_3': 'surname:Weikum;given-names:G',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of '\n",
      "                                                   'International Conference '\n",
      "                                                   'on Data Engineering (ICDE)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2014'},\n",
      "                              'offset': 51101,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'KnowLife: a knowledge graph for health '\n",
      "                                      'and life sciences'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Siu;given-names:A',\n",
      "                                         'name_1': 'surname:Nguyen;given-names:DB',\n",
      "                                         'name_2': 'surname:Weikum;given-names:G',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Workshop on '\n",
      "                                                   'Data Mining for Healthcare '\n",
      "                                                   '(DMH) at Conference on '\n",
      "                                                   'Knowledge Discovery and '\n",
      "                                                   'Data Mining (KDD)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2013'},\n",
      "                              'offset': 51158,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Fast entity recognition in biomedical '\n",
      "                                      'text'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Charikar;given-names:MS',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Symposium '\n",
      "                                                   'on Theory of Computing '\n",
      "                                                   '(STOC)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '2002'},\n",
      "                              'offset': 51201,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Similarity estimation techniques from '\n",
      "                                      'rounding algorithms'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'name_0': 'surname:Broder;given-names:AZ',\n",
      "                                         'name_1': 'surname:Charikar;given-names:M',\n",
      "                                         'name_2': 'surname:Frieze;given-names:AM',\n",
      "                                         'name_3': 'surname:Mitzenmacher;given-names:M',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Proceedings of Symposium '\n",
      "                                                   'on Theory of Computing '\n",
      "                                                   '(STOC)',\n",
      "                                         'type': 'ref',\n",
      "                                         'year': '1998'},\n",
      "                              'offset': 51259,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Min-wise independent permutations'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '63',\n",
      "                                         'issue': '1',\n",
      "                                         'lpage': '88',\n",
      "                                         'name_0': 'surname:Niedermeier;given-names:R',\n",
      "                                         'name_1': 'surname:Rossmath;given-names:P',\n",
      "                                         'pub-id_doi': '10.1006/jagm.2000.1075',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'J Algorithms',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '36',\n",
      "                                         'year': '2000'},\n",
      "                              'offset': 51293,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'New upper bounds for maximum '\n",
      "                                      'satisfiability'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '256',\n",
      "                                         'issue': '3',\n",
      "                                         'lpage': '78',\n",
      "                                         'name_0': 'surname:Johnson;given-names:DS',\n",
      "                                         'pub-id_doi': '10.1016/S0022-0000(74)80044-9',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'J Comput Syst Sci',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '9',\n",
      "                                         'year': '1974'},\n",
      "                              'offset': 51337,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Approximation algorithms for '\n",
      "                                      'combinatorial problems'},\n",
      "                             {'annotations': [],\n",
      "                              'infons': {'fpage': '216',\n",
      "                                         'lpage': '20',\n",
      "                                         'name_0': 'surname:McCray;given-names:AT',\n",
      "                                         'name_1': 'surname:Burgun;given-names:A',\n",
      "                                         'name_2': 'surname:Bodenreider;given-names:O',\n",
      "                                         'section_type': 'REF',\n",
      "                                         'source': 'Stud Health Technol '\n",
      "                                                   'Informatics',\n",
      "                                         'type': 'ref',\n",
      "                                         'volume': '1',\n",
      "                                         'year': '2001'},\n",
      "                              'offset': 51389,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'Aggregating UMLS semantic types for '\n",
      "                                      'reducing conceptual complexity'}],\n",
      "                'pmcid': 'PMC4448285',\n",
      "                'pmid': 25971816,\n",
      "                'relations': [],\n",
      "                'relations_display': []},\n",
      "               {'_id': '25133803|None',\n",
      "                'authors': ['Sun K', 'Buchan N', 'Larminie C', 'Pržulj N'],\n",
      "                'date': '2014-11-01T00:00:00Z',\n",
      "                'id': '25133803',\n",
      "                'infons': {},\n",
      "                'journal': 'Integr Biol (Camb)',\n",
      "                'meta': {},\n",
      "                'passages': [{'annotations': [],\n",
      "                              'infons': {'authors': 'Sun K, Buchan N, Larminie '\n",
      "                                                    'C, Przulj N',\n",
      "                                         'journal': 'Integr Biol (Camb). 2014 '\n",
      "                                                    'Nov;6(11):1069-79. doi: '\n",
      "                                                    '10.1039/c4ib00122b.',\n",
      "                                         'type': 'title',\n",
      "                                         'year': '2014'},\n",
      "                              'offset': 0,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The integrated disease network.'},\n",
      "                             {'annotations': [{'id': '2',\n",
      "                                               'infons': {'accession': '@DISEASE_Crohn_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D003424',\n",
      "                                                          'name': 'Crohn '\n",
      "                                                                  'Disease',\n",
      "                                                          'normalized': ['D003424'],\n",
      "                                                          'normalized_id': 'D003424',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 15,\n",
      "                                                              'offset': 1339}],\n",
      "                                               'text': \"Crohn's disease\"},\n",
      "                                              {'id': '3',\n",
      "                                               'infons': {'accession': '@DISEASE_Crohn_Disease',\n",
      "                                                          'biotype': 'disease',\n",
      "                                                          'database': 'ncbi_mesh',\n",
      "                                                          'identifier': 'MESH:D003424',\n",
      "                                                          'name': 'Crohn '\n",
      "                                                                  'Disease',\n",
      "                                                          'normalized': ['D003424'],\n",
      "                                                          'normalized_id': 'D003424',\n",
      "                                                          'type': 'Disease',\n",
      "                                                          'valid': True},\n",
      "                                               'locations': [{'length': 15,\n",
      "                                                              'offset': 1447}],\n",
      "                                               'text': \"Crohn's disease\"}],\n",
      "                              'infons': {'type': 'abstract'},\n",
      "                              'offset': 32,\n",
      "                              'relations': [],\n",
      "                              'sentences': [],\n",
      "                              'text': 'The growing body of transcriptomic, '\n",
      "                                      'proteomic, metabolomic and genomic data '\n",
      "                                      'generated from disease states provides '\n",
      "                                      'a great opportunity to improve our '\n",
      "                                      'current understanding of the molecular '\n",
      "                                      'mechanisms driving diseases and shared '\n",
      "                                      'between diseases. The use of both '\n",
      "                                      'clinical and molecular phenotypes will '\n",
      "                                      'lead to better disease understanding '\n",
      "                                      'and classification. In this study, we '\n",
      "                                      'set out to gain novel insights into '\n",
      "                                      'diseases and their relationships by '\n",
      "                                      'utilising knowledge gained from '\n",
      "                                      'system-level molecular data. We '\n",
      "                                      'integrated different types of '\n",
      "                                      'biological data including genome-wide '\n",
      "                                      'association studies data, '\n",
      "                                      'disease-chemical associations, '\n",
      "                                      'biological pathways and Gene Ontology '\n",
      "                                      'annotations into an Integrated Disease '\n",
      "                                      'Network (IDN), a heterogeneous network '\n",
      "                                      'where nodes are bio-entities and edges '\n",
      "                                      'between nodes represent their '\n",
      "                                      'associations. We also introduced a '\n",
      "                                      'novel disease similarity measure to '\n",
      "                                      'infer disease-disease associations from '\n",
      "                                      'the IDN. Our predicted associations '\n",
      "                                      'were systemically evaluated against the '\n",
      "                                      'Medical Subject Heading classification '\n",
      "                                      'and a statistical measure of disease '\n",
      "                                      'co-occurrence in PubMed. The strong '\n",
      "                                      'correlation between our predictions and '\n",
      "                                      'co-occurrence associations indicated '\n",
      "                                      'the ability of our approach to recover '\n",
      "                                      'known disease associations. '\n",
      "                                      'Furthermore, we presented a case study '\n",
      "                                      \"of Crohn's disease. We demonstrated \"\n",
      "                                      'that our approach not only identified '\n",
      "                                      'well-established connections between '\n",
      "                                      \"Crohn's disease and other diseases, but \"\n",
      "                                      'also revealed new, interesting '\n",
      "                                      'connections consistent with emerging '\n",
      "                                      'literature. Our approach also enabled '\n",
      "                                      'ready access to the knowledge '\n",
      "                                      'supporting these new connections, '\n",
      "                                      'making this a powerful approach for '\n",
      "                                      'exploring connections between '\n",
      "                                      'diseases. '}],\n",
      "                'pmcid': None,\n",
      "                'pmid': 25133803,\n",
      "                'relations': [],\n",
      "                'relations_display': []}]}\n"
     ]
    }
   ],
   "source": [
    "def annotate_pmids(pmids):\n",
    "    # Define the PubTator API URL\n",
    "    pubtator_url = \"https://www.ncbi.nlm.nih.gov/research/pubtator-api/publications/export/biocjson\"\n",
    "\n",
    "    # Join the PMIDs into a comma-separated string\n",
    "    pmid_str = ','.join(pmids)\n",
    "\n",
    "    # Send a GET request to the API with the list of PMIDs\n",
    "    response = requests.get(pubtator_url, params={\"pmids\": pmid_str, \"full\": True})\n",
    "\n",
    "    # Check if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        return response.json()  # Return the response in JSON format\n",
    "    else:\n",
    "        raise Exception(f\"Error {response.status_code}: {response.text}\")\n",
    "\n",
    "# Example usage\n",
    "pmids = [\"25133803\", \"25971816\"]\n",
    "annotations = annotate_pmids(pmids)\n",
    "\n",
    "# Display the annotated results\n",
    "import pprint\n",
    "pprint.pprint(annotations)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bkg-review-FcnO91kN-py3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
