{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package punkt_tab to\n",
      "[nltk_data]     /Users/anglup/nltk_data...\n",
      "[nltk_data]   Package punkt_tab is already up-to-date!\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from ast import literal_eval\n",
    "import os, sys\n",
    "sys.path.append('../pubmed_rag')\n",
    "#from utils import get_chunks\n",
    "from bioc import (\n",
    "    mean_pooling\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "get the data for the vx db first. \n",
    "\n",
    "Note: The primary key and vector fields use their default names (“id” and “vector”).\n",
    "\n",
    "to insert into vx database need each row as a dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in\n",
    "filename = '/Users/anglup/GitHub/pubmed-rag/biocjson/all_embeddings.csv'\n",
    "df = pd.read_csv(filename, )\n",
    "# rename id and vector cols\n",
    "df = df.rename(\n",
    "    columns={\n",
    "        'Unnamed: 0':'id',\n",
    "        'embedding':'vector'\n",
    "    }, \n",
    ")\n",
    "# ensure vector is a list\n",
    "df['vector'] = df['vector'].apply(literal_eval)\n",
    "\n",
    "# sanity check\n",
    "#data.head()\n",
    "\n",
    "# to list of dicts\n",
    "data = df.to_dict(orient='records')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>section</th>\n",
       "      <th>pmid</th>\n",
       "      <th>pmcid</th>\n",
       "      <th>date</th>\n",
       "      <th>authors</th>\n",
       "      <th>journal</th>\n",
       "      <th>text</th>\n",
       "      <th>vector</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>abstract</td>\n",
       "      <td>28924171</td>\n",
       "      <td>PMC5603535</td>\n",
       "      <td>2017-09-18 00:00:00+00:00</td>\n",
       "      <td>Luo Y and Zhao X and Zhou J and Yang J and Zha...</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>The emergence of large-scale genomic, chemical...</td>\n",
       "      <td>[-0.027584053575992584, 0.054618917405605316, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>discuss</td>\n",
       "      <td>28924171</td>\n",
       "      <td>PMC5603535</td>\n",
       "      <td>2017-09-18 00:00:00+00:00</td>\n",
       "      <td>Luo Y and Zhao X and Zhou J and Yang J and Zha...</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>Discussion . Recent advances in large-scale ex...</td>\n",
       "      <td>[0.00014715130964759737, 0.054599590599536896,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>discuss</td>\n",
       "      <td>28924171</td>\n",
       "      <td>PMC5603535</td>\n",
       "      <td>2017-09-18 00:00:00+00:00</td>\n",
       "      <td>Luo Y and Zhao X and Zhou J and Yang J and Zha...</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>On the other hand, based on the systematic int...</td>\n",
       "      <td>[0.00014715130964759737, 0.054599590599536896,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>discuss</td>\n",
       "      <td>28924171</td>\n",
       "      <td>PMC5603535</td>\n",
       "      <td>2017-09-18 00:00:00+00:00</td>\n",
       "      <td>Luo Y and Zhao X and Zhou J and Yang J and Zha...</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>Our findings are particularly helpful for unde...</td>\n",
       "      <td>[0.00014715130964759737, 0.054599590599536896,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>intro</td>\n",
       "      <td>28924171</td>\n",
       "      <td>PMC5603535</td>\n",
       "      <td>2017-09-18 00:00:00+00:00</td>\n",
       "      <td>Luo Y and Zhao X and Zhou J and Yang J and Zha...</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>Introduction . Computational prediction of dru...</td>\n",
       "      <td>[-0.039880603551864624, -0.011334901675581932,...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   id   section      pmid       pmcid                       date  \\\n",
       "0   0  abstract  28924171  PMC5603535  2017-09-18 00:00:00+00:00   \n",
       "1   1   discuss  28924171  PMC5603535  2017-09-18 00:00:00+00:00   \n",
       "2   2   discuss  28924171  PMC5603535  2017-09-18 00:00:00+00:00   \n",
       "3   3   discuss  28924171  PMC5603535  2017-09-18 00:00:00+00:00   \n",
       "4   4     intro  28924171  PMC5603535  2017-09-18 00:00:00+00:00   \n",
       "\n",
       "                                             authors     journal  \\\n",
       "0  Luo Y and Zhao X and Zhou J and Yang J and Zha...  Nat Commun   \n",
       "1  Luo Y and Zhao X and Zhou J and Yang J and Zha...  Nat Commun   \n",
       "2  Luo Y and Zhao X and Zhou J and Yang J and Zha...  Nat Commun   \n",
       "3  Luo Y and Zhao X and Zhou J and Yang J and Zha...  Nat Commun   \n",
       "4  Luo Y and Zhao X and Zhou J and Yang J and Zha...  Nat Commun   \n",
       "\n",
       "                                                text  \\\n",
       "0  The emergence of large-scale genomic, chemical...   \n",
       "1  Discussion . Recent advances in large-scale ex...   \n",
       "2  On the other hand, based on the systematic int...   \n",
       "3  Our findings are particularly helpful for unde...   \n",
       "4  Introduction . Computational prediction of dru...   \n",
       "\n",
       "                                              vector  \n",
       "0  [-0.027584053575992584, 0.054618917405605316, ...  \n",
       "1  [0.00014715130964759737, 0.054599590599536896,...  \n",
       "2  [0.00014715130964759737, 0.054599590599536896,...  \n",
       "3  [0.00014715130964759737, 0.054599590599536896,...  \n",
       "4  [-0.039880603551864624, -0.011334901675581932,...  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pymilvus import MilvusClient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "client = MilvusClient('demo.db')\n",
    "\n",
    "if client.has_collection(collection_name=\"demo_collection\"):\n",
    "    client.drop_collection(collection_name=\"demo_collection\")\n",
    "client.create_collection(\n",
    "    collection_name=\"demo_collection\",\n",
    "    dimension=768,  # The vectors we will use in this demo has 768 dimensions\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'insert_count': 47, 'ids': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46]}\n"
     ]
    }
   ],
   "source": [
    "result = client.insert(collection_name=\"demo_collection\", data=data)\n",
    "print(result)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "yay first vx db created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/anglup/Library/Caches/pypoetry/virtualenvs/pubmed-rag-iB5Z1S4t-py3.11/lib/python3.11/site-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from transformers import AutoTokenizer, AutoModel\n",
    "import torch\n",
    "import torch.nn.functional as F\n",
    "\n",
    "chosen_model = 'sentence-transformers/all-mpnet-base-v2'\n",
    "\n",
    "#logger.info(f\"Loading model {chosen_model} from HuggingFace\")\n",
    "tokenizer = AutoTokenizer.from_pretrained(chosen_model)\n",
    "model = AutoModel.from_pretrained(chosen_model)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sentence embeddings:\n",
      "tensor([[ 3.2914e-02, -3.1554e-02,  7.1610e-03, -9.6834e-03, -2.0804e-02,\n",
      "         -2.9990e-02, -2.1607e-02,  2.8089e-02,  8.4443e-02, -2.1861e-03,\n",
      "          2.7061e-02,  7.4580e-02, -8.7139e-03,  9.1949e-02,  8.1665e-03,\n",
      "          4.3600e-02,  6.0810e-03,  5.2895e-02,  9.5613e-03, -4.6859e-02,\n",
      "         -2.3129e-02,  1.7358e-02,  1.5377e-03, -1.5145e-03, -7.0703e-03,\n",
      "          1.4116e-03,  8.3259e-03,  2.2533e-02,  1.5436e-02, -4.8137e-02,\n",
      "         -1.2683e-02, -1.7956e-02, -1.6711e-02,  8.5764e-03,  1.2389e-06,\n",
      "         -4.5091e-02, -5.2964e-02,  3.6673e-02, -4.8655e-02, -3.8745e-02,\n",
      "          5.4656e-02, -4.1287e-02,  5.5994e-02, -3.5536e-02,  7.0920e-03,\n",
      "         -4.3131e-02,  1.9771e-02,  2.5608e-02,  2.1471e-02,  1.9946e-02,\n",
      "         -3.8807e-03, -9.9989e-03,  7.1122e-02, -2.4125e-02,  2.6173e-02,\n",
      "         -1.0321e-02,  5.3093e-02,  1.8854e-02,  5.6223e-02, -1.3137e-02,\n",
      "         -6.1156e-03,  3.4451e-02, -3.0967e-02, -2.8824e-02,  1.1615e-01,\n",
      "         -2.5856e-02,  8.1065e-03, -4.4640e-02, -1.7135e-02,  3.6702e-02,\n",
      "          5.1191e-02, -9.4401e-03,  3.1080e-02,  2.3571e-02,  1.2413e-02,\n",
      "          5.0558e-02, -2.7077e-02, -7.1922e-02,  1.2000e-03,  2.6856e-02,\n",
      "         -2.3197e-02, -5.3287e-02,  2.1316e-02, -4.2007e-02,  2.5669e-02,\n",
      "          3.7112e-02, -1.2321e-02, -2.7033e-03, -1.3759e-02, -6.5150e-02,\n",
      "          3.2603e-02,  2.3146e-02,  4.2642e-02,  1.2623e-02,  8.3257e-03,\n",
      "         -4.8532e-03,  5.5017e-02,  3.9194e-02,  6.5474e-02,  7.4496e-03,\n",
      "          1.3034e-02, -3.1529e-02,  4.1050e-02,  3.2768e-02, -4.5258e-02,\n",
      "         -5.4481e-02, -2.8537e-02,  4.5836e-02, -4.5385e-02, -8.7518e-03,\n",
      "         -7.3551e-02, -3.5672e-02, -2.3485e-02,  6.1499e-02,  5.7029e-02,\n",
      "         -2.7190e-02, -3.1037e-02,  4.9689e-03, -2.5067e-02,  3.0446e-03,\n",
      "          9.0955e-02,  8.4790e-03,  6.9967e-02,  9.5816e-03, -1.9797e-02,\n",
      "          2.5280e-02,  3.5898e-02, -8.3848e-03,  5.5617e-03, -6.4408e-03,\n",
      "          8.5968e-03, -3.1983e-02,  3.7262e-02, -2.2941e-03, -2.1453e-02,\n",
      "          6.0605e-02,  4.0530e-02,  1.7032e-02, -5.6151e-02, -2.8053e-02,\n",
      "         -1.4454e-02, -1.1667e-02,  2.4280e-02,  2.0447e-02,  5.6120e-02,\n",
      "         -1.3065e-02, -1.2668e-02,  1.6565e-02, -2.6242e-03, -2.4155e-02,\n",
      "         -8.2504e-02,  9.5480e-03, -5.4143e-02, -1.5658e-02, -8.7187e-03,\n",
      "         -1.5428e-02,  6.5507e-04, -5.9103e-02, -2.0456e-02, -1.8136e-02,\n",
      "          1.6955e-02,  2.0869e-02, -4.2371e-02, -7.1672e-02,  2.0594e-02,\n",
      "         -6.0338e-02,  3.2418e-02, -1.5351e-02,  1.6823e-03,  6.3702e-02,\n",
      "         -3.3155e-02, -1.3478e-02, -1.8215e-02, -7.0134e-03,  8.5645e-03,\n",
      "          2.3042e-02,  5.4350e-02, -3.1622e-02, -1.1138e-02,  1.0235e-02,\n",
      "         -1.5895e-03, -7.3709e-02, -3.6586e-02,  5.4886e-03,  1.7014e-02,\n",
      "          3.2697e-02, -3.6272e-02,  1.5613e-02, -2.5340e-03,  1.8036e-02,\n",
      "          2.5521e-03,  5.4626e-03,  1.8885e-02, -1.6830e-02,  6.6463e-03,\n",
      "         -1.6568e-02, -3.0708e-02, -2.8580e-02,  1.1330e-02, -1.8120e-02,\n",
      "          4.3670e-05, -1.1332e-03, -2.7901e-02,  1.1238e-01,  2.3016e-02,\n",
      "          7.4745e-03, -3.7245e-02,  1.3703e-02, -7.8485e-02, -8.5396e-03,\n",
      "          1.1658e-02,  1.2465e-03, -2.6440e-02, -1.1279e-02,  1.0068e-02,\n",
      "          1.1841e-02,  2.7699e-02,  1.0840e-02, -3.2977e-02,  1.9685e-02,\n",
      "         -3.4747e-02, -7.7949e-02, -3.8366e-03, -2.8593e-02,  2.1635e-02,\n",
      "          4.8446e-02, -5.1120e-02, -2.8038e-02, -5.8560e-03,  7.7153e-02,\n",
      "          5.1097e-02,  4.3568e-02,  6.5248e-02, -3.5962e-02, -2.7289e-02,\n",
      "         -7.6688e-03,  4.4159e-02,  1.3352e-02,  3.3902e-02, -3.0963e-02,\n",
      "          2.1652e-02, -3.4026e-02, -1.6230e-03, -3.1015e-02,  1.9205e-02,\n",
      "          1.9004e-02,  4.5462e-02,  3.0760e-02,  1.2406e-02,  6.7472e-03,\n",
      "         -3.7245e-02, -2.3027e-03, -1.4260e-02,  4.1671e-02, -3.9358e-02,\n",
      "          7.7941e-03, -1.0828e-01,  2.8803e-02, -1.1731e-02,  6.2654e-02,\n",
      "         -3.4593e-02, -5.9116e-03, -1.9807e-03, -3.1118e-02, -3.4713e-02,\n",
      "         -3.1691e-02,  8.1935e-02, -4.3849e-03, -9.7061e-03,  4.2721e-02,\n",
      "          4.9385e-02, -7.6013e-03, -5.0407e-02,  8.7976e-03, -6.6492e-02,\n",
      "         -3.6742e-02,  7.1271e-02, -4.6696e-02, -3.4033e-02,  7.5075e-03,\n",
      "          1.1875e-02, -3.2553e-02, -1.2217e-02,  3.8999e-03, -1.8005e-02,\n",
      "          4.6242e-02, -1.4740e-02, -2.5166e-02,  3.8574e-02,  2.9543e-02,\n",
      "          2.2067e-02,  1.5749e-02, -1.3611e-02,  1.4257e-02, -3.5701e-03,\n",
      "          3.7489e-02, -2.0592e-03, -7.5710e-03, -6.2802e-02,  1.9393e-02,\n",
      "          3.7429e-03, -3.5017e-02,  1.1873e-02, -2.3732e-02, -3.0934e-02,\n",
      "          2.9441e-02,  6.6271e-02,  5.4350e-04, -3.7203e-02, -1.1069e-03,\n",
      "          2.0651e-02, -2.4067e-02, -1.2290e-03, -2.7741e-02,  9.2491e-03,\n",
      "         -5.6329e-03, -4.3284e-02, -2.4937e-02,  3.2368e-02, -4.5903e-03,\n",
      "         -1.0269e-02,  3.3314e-03,  4.8657e-02,  3.2238e-04,  6.2418e-02,\n",
      "         -3.7754e-02,  6.4884e-03, -2.3191e-04, -4.8598e-02,  4.3056e-02,\n",
      "          1.7767e-02,  3.1818e-03,  3.6232e-02, -1.0087e-03,  9.4982e-04,\n",
      "          5.7391e-02, -1.7322e-02, -2.5145e-02, -1.5047e-02,  2.6345e-02,\n",
      "          3.1433e-02,  1.9757e-02, -3.9710e-03,  1.2059e-01,  6.8294e-02,\n",
      "         -2.2815e-03,  2.1676e-02,  3.3986e-02,  1.4855e-02,  2.8337e-02,\n",
      "          1.8305e-02,  1.3638e-02,  6.9527e-03,  1.7724e-02,  6.7834e-03,\n",
      "          2.4540e-02,  1.1869e-02, -3.8580e-02, -3.0781e-02,  3.1185e-02,\n",
      "          2.5022e-02,  8.1905e-03, -1.8927e-02, -2.2764e-02, -6.1571e-03,\n",
      "          5.5039e-02, -2.4790e-02, -1.7111e-02, -3.2683e-02, -7.6706e-03,\n",
      "         -1.3776e-02, -1.3562e-02,  5.1277e-02,  2.5873e-03,  2.4158e-02,\n",
      "         -7.4751e-02,  1.6535e-02, -1.2766e-03, -1.5082e-01, -2.9216e-02,\n",
      "          4.7766e-02, -7.6734e-02,  2.3535e-02, -7.4651e-02, -5.4070e-03,\n",
      "         -1.7469e-02, -3.6084e-02, -3.7901e-02,  2.4999e-02,  2.4214e-02,\n",
      "         -6.1907e-03, -9.8019e-02, -5.0038e-02, -7.3743e-02, -4.7588e-02,\n",
      "         -3.7031e-02, -6.2201e-02, -1.7503e-02,  9.8887e-03, -4.1602e-02,\n",
      "         -1.4807e-02, -2.9043e-02, -5.1765e-02, -2.7741e-02,  2.9153e-02,\n",
      "         -8.1355e-03, -3.7748e-02,  4.0907e-02,  3.3292e-02,  8.0087e-03,\n",
      "         -1.2395e-03, -3.8638e-02,  1.9323e-03, -1.9611e-02,  2.8044e-02,\n",
      "          7.6517e-03,  2.1399e-02, -4.9480e-02,  5.0347e-02, -4.2077e-02,\n",
      "         -4.9417e-02, -5.7855e-02,  7.7640e-02, -1.3233e-02, -4.5433e-02,\n",
      "         -4.5518e-02, -2.6653e-02,  5.2212e-03,  3.3207e-02,  3.3133e-02,\n",
      "          3.2139e-02, -1.8733e-02,  9.0410e-03,  8.4568e-02, -6.6727e-02,\n",
      "          4.6531e-02,  4.8901e-02, -6.4851e-02,  5.9511e-02, -2.5920e-02,\n",
      "          4.3920e-03,  2.8276e-02, -3.0105e-02,  2.9685e-02,  5.2043e-03,\n",
      "          1.4474e-02, -7.4932e-02, -5.1318e-02, -2.3778e-02,  4.7660e-02,\n",
      "         -9.5957e-03,  2.1800e-02, -8.1763e-02,  2.8894e-02, -3.3449e-03,\n",
      "         -9.7487e-03, -4.5752e-03,  4.7272e-02, -4.2698e-02, -1.4844e-02,\n",
      "          7.1121e-03,  2.6958e-02,  4.9850e-03,  2.5738e-03,  1.5141e-02,\n",
      "          7.0013e-02,  2.6353e-02, -8.3978e-03, -8.3924e-02, -7.2939e-02,\n",
      "          3.3423e-02, -2.9763e-02, -7.9721e-03, -1.9924e-02,  4.1949e-02,\n",
      "         -5.3766e-03, -1.5474e-02,  3.3781e-02, -6.0196e-02,  2.0463e-02,\n",
      "          6.6121e-02, -7.1241e-03,  3.0068e-02,  6.5260e-02, -4.8099e-02,\n",
      "          2.6549e-02,  1.1352e-02,  2.6002e-02,  1.1430e-02, -3.3481e-02,\n",
      "          1.0549e-02, -2.0048e-02,  2.0546e-02, -4.2930e-02, -4.6154e-03,\n",
      "         -2.5189e-02,  2.9090e-03, -3.1918e-02,  2.1244e-02,  8.4770e-03,\n",
      "          5.3498e-02, -2.2332e-02, -1.8963e-02, -7.8184e-04, -5.2185e-02,\n",
      "         -8.9665e-03, -1.3305e-03,  1.3825e-02, -4.0001e-02, -3.5542e-02,\n",
      "         -3.5721e-02,  3.4751e-02,  1.5152e-02, -2.4955e-02, -2.5464e-02,\n",
      "         -1.5753e-02,  3.0527e-02,  3.0647e-02,  5.2673e-02, -1.9395e-02,\n",
      "          1.0234e-02, -4.3430e-02,  4.4407e-02, -1.0705e-02,  5.4099e-02,\n",
      "         -3.1215e-03,  3.6521e-02,  2.3662e-02, -4.5755e-02,  5.7262e-02,\n",
      "          2.3374e-02,  1.6501e-02, -5.2043e-02, -2.0671e-02, -1.6123e-02,\n",
      "         -1.6329e-02,  1.7800e-02,  3.0145e-02, -1.3670e-02, -1.4481e-02,\n",
      "         -3.7204e-02, -2.5191e-02,  2.8898e-02,  3.0439e-02, -9.4445e-03,\n",
      "          3.8031e-02, -2.6656e-02, -2.1228e-02,  5.1086e-02, -5.3264e-02,\n",
      "         -1.3289e-02,  6.2648e-03, -3.0033e-02, -2.0567e-02,  5.2450e-02,\n",
      "         -4.6553e-33,  6.1565e-02, -2.1010e-02,  1.5284e-02,  2.0026e-02,\n",
      "         -2.3812e-02, -3.0580e-02, -4.6135e-02, -6.7561e-02, -7.8444e-03,\n",
      "          8.0045e-02,  6.3476e-02, -2.6525e-02, -5.3128e-03,  1.6705e-02,\n",
      "         -2.8189e-02,  2.4696e-02, -3.5871e-02,  2.9094e-02,  1.5196e-02,\n",
      "          4.5943e-03,  4.1323e-02,  6.5255e-02,  1.8533e-02,  3.1523e-02,\n",
      "          1.4327e-02, -1.8295e-03,  5.4844e-02,  3.4927e-02, -1.9421e-02,\n",
      "          4.4031e-02, -3.6819e-03,  3.9090e-02, -4.0796e-03,  1.6716e-03,\n",
      "          3.1123e-03, -8.2856e-02, -5.3428e-02, -3.2822e-03,  3.3902e-02,\n",
      "         -2.8977e-02, -4.8912e-02, -6.4138e-02,  5.5074e-02, -6.5768e-03,\n",
      "          2.8058e-02, -9.4609e-03,  3.3101e-02, -1.1668e-02, -5.0792e-02,\n",
      "          1.1005e-02,  4.3651e-02,  5.1264e-03,  1.9639e-03,  5.9136e-02,\n",
      "         -9.0122e-03,  5.5150e-04, -3.3859e-02, -4.3724e-02, -3.1238e-02,\n",
      "          1.6888e-02, -4.7060e-02,  1.6217e-02,  1.7486e-02, -2.8565e-02,\n",
      "          3.6401e-02,  1.7044e-02, -7.4965e-02,  1.5017e-02,  7.5523e-02,\n",
      "          5.5751e-03, -3.4013e-02,  6.4081e-02,  3.0084e-02, -4.6664e-02,\n",
      "         -4.3355e-02,  1.8619e-02, -3.6715e-02, -1.4537e-02, -4.9691e-02,\n",
      "          6.7160e-03, -1.2323e-02,  3.2166e-02,  6.1535e-02, -4.0227e-02,\n",
      "         -1.1158e-02, -2.8462e-02, -4.8057e-02, -1.3822e-02,  1.7044e-03,\n",
      "         -3.2899e-02, -3.5950e-02,  8.8071e-02, -2.6501e-02, -2.9861e-02,\n",
      "          3.0584e-02,  3.2500e-02,  2.9262e-02,  9.9300e-03, -1.8032e-03,\n",
      "         -2.9035e-03, -1.0428e-02,  1.6856e-02,  1.8472e-02, -2.7981e-02,\n",
      "          5.1810e-03, -1.0306e-03, -2.5305e-02, -1.5786e-02,  1.8227e-02,\n",
      "          1.3018e-02, -1.1989e-03, -2.1324e-02,  1.3897e-02, -5.0795e-02,\n",
      "         -5.5830e-03,  3.0711e-02,  9.5792e-03,  1.4083e-02,  3.3121e-02,\n",
      "          3.4771e-02,  1.7077e-02,  3.8848e-02,  4.2792e-02, -9.9609e-03,\n",
      "         -1.8794e-02,  3.9103e-02,  8.5676e-03,  7.0509e-02, -4.1938e-02,\n",
      "         -4.2199e-03, -6.2155e-03,  3.8944e-02,  2.2057e-07,  8.0815e-02,\n",
      "          1.9611e-02,  4.4767e-02, -3.6916e-03,  1.9323e-02, -5.5266e-02,\n",
      "         -1.8240e-03, -8.3036e-04, -6.9532e-02, -3.0995e-02,  5.6786e-02,\n",
      "          9.2810e-04,  1.1724e-02, -1.1070e-02, -5.2490e-02, -4.6792e-02,\n",
      "         -5.7304e-03, -8.3911e-02, -1.2801e-02, -8.9562e-03,  1.9826e-02,\n",
      "          1.4550e-03,  3.8946e-02,  2.1189e-02, -5.8513e-03, -2.3990e-03,\n",
      "         -3.3685e-02, -2.4007e-02,  2.9393e-02,  4.3075e-02,  3.3155e-02,\n",
      "         -6.8213e-02, -1.9329e-02,  7.4223e-02, -2.7674e-03, -4.5462e-02,\n",
      "         -1.9773e-02, -4.4112e-03, -3.6899e-02,  8.1637e-03,  6.3302e-02,\n",
      "         -2.2145e-02,  8.2071e-03, -9.1054e-04, -1.4031e-02, -2.4016e-02,\n",
      "         -3.6397e-02,  4.4609e-02, -8.7922e-03, -4.7595e-02,  6.5054e-02,\n",
      "         -2.2885e-02,  3.1262e-02,  4.7826e-02, -2.3999e-02,  3.7427e-02,\n",
      "          1.5840e-02,  9.7233e-03,  3.7426e-02,  7.1887e-02, -7.8797e-03,\n",
      "         -5.7750e-02,  4.5699e-03, -1.8221e-02,  5.4924e-02, -7.5358e-02,\n",
      "          1.8405e-02,  1.0819e-34,  1.2194e-02, -3.4390e-02,  1.1639e-02,\n",
      "          2.6625e-02,  1.5467e-02, -4.0677e-02, -2.6944e-02, -1.3965e-02,\n",
      "         -4.7218e-02, -8.8289e-02, -1.4457e-02]])\n"
     ]
    }
   ],
   "source": [
    "# Tokenize sentences\n",
    "encoded_input = tokenizer(\n",
    "    \"Can you please tell me what nodes and edges I should include in a biological knowledge graph for drug repurposing?\", \n",
    "    padding=True, \n",
    "    truncation=False, \n",
    "    return_tensors='pt'\n",
    ")\n",
    "\n",
    "# Compute token embeddings\n",
    "with torch.no_grad():\n",
    "    model_output = model(**encoded_input)\n",
    "\n",
    "# Perform pooling\n",
    "sentence_embeddings = mean_pooling(model_output, encoded_input['attention_mask'])\n",
    "\n",
    "# Normalize embeddings\n",
    "sentence_embeddings = F.normalize(sentence_embeddings, p=2, dim=1)\n",
    "\n",
    "print(\"Sentence embeddings:\")\n",
    "print(sentence_embeddings)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...\n",
      "To disable this warning, you can either:\n",
      "\t- Avoid using `tokenizers` before the fork if possible\n",
      "\t- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)\n"
     ]
    }
   ],
   "source": [
    "client = MilvusClient('../demo.db')\n",
    "\n",
    "result = client.search(\n",
    "    collection_name=\"main\",\n",
    "    data=sentence_embeddings.detach().numpy(),\n",
    "    #filter=\"section == 'abstract'\",\n",
    "    limit=10,\n",
    "    output_fields=[\"text\", \"pmid\", 'section'],\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'id': 593,\n",
       "  'distance': 0.7601258754730225,\n",
       "  'entity': {'section': 'title',\n",
       "   'pmid': 35840060,\n",
       "   'text': 'KG-Predict: A knowledge graph computational framework for drug repurposing.'}},\n",
       " {'id': 426,\n",
       "  'distance': 0.724739134311676,\n",
       "  'entity': {'section': 'title',\n",
       "   'pmid': 34297570,\n",
       "   'text': 'Knowledge Graph-Based Approaches to Drug Repurposing for COVID-19.'}},\n",
       " {'id': 534,\n",
       "  'distance': 0.6885125041007996,\n",
       "  'entity': {'section': 'title',\n",
       "   'pmid': 34181736,\n",
       "   'text': 'CROssBAR: comprehensive resource of biomedical relations with knowledge graph representations.'}},\n",
       " {'id': 649,\n",
       "  'distance': 0.6845976114273071,\n",
       "  'entity': {'section': 'methods',\n",
       "   'pmid': 36175836,\n",
       "   'text': 'To the extent that it incorporates multiple graph structural variations, RTX-KG2 can also be used as a test-bed for evaluating the performance of structurally generalizable graph analysis methods such as a subset of us have done for the case of a structurally generalizable node-node similarity measure . Another application of RTX-KG2 is as training data for drug repurposing models; previous work by our team utilized RTX-KG2\\'s predecessor, RTX-KG1, to train a random forest model that predicts novel drug treatments for diseases . More recent extensions to this work have utilized the canonicalized version of RTX-KG2 as training data . Comparison to other knowledge graphs/providers . To objectively evaluate the size and semantic richness of RTX-KG2 in comparison to other biomedical knowledge graphs, we compared it to four other knowledge graphs that are in active use for translational applications: Hetionet, SPOKE, the SRI Reference Knowledge Graph, and ROBOKOP . For the counts of meta-triples (i.e., counts of edges with a given pattern of subject category, object category, and predicate, which provide quantitative information about the richness of the knowledge graph for providing relationships of particular types), we accessed the ROBOKOP and SPOKE graphs via their SmartAPI-registered Translator API (\"TRAPI\") endpoints on March 15, 2022 . For Hetionet and ROBOKOP, we used published node and edge counts . For the SRI Reference Knowledge Graph, we downloaded the graph in KGX TSV format from the Biomedical Data Translator Knowledge Graph Exchange (KGE) and analyzed the graph locally . Node and edge counts for various knowledge graphs are shown in Table 2 . The large number of edges in the ROBOKOP knowledge graph versus RTX-KG2 reflects the latter\\'s practice of joining edges that have identical subject node, object node, and predicate type . The large number of nodes in the SRI Reference Knowledge Graph versus RTX-KG2pre is largely due to more InformationContentEntity nodes (3.7 million vs. 144,396) and SequenceVariant nodes (2.4 million vs. 0) in the former . Notably, RTX-KG2 contains a richer set of meta-triples (distinct combinations of subject node category, edge predicate, and object node category) versus the other knowledge graphs (Table 3); RTX-KG2c contains 4.6-fold more meta-triples than the second-ranked knowledge graph and 18.5-fold more meta-triples than the third-ranked knowledge graph, by meta-triple count . In general we would expect a graph with a greater number of meta-triples to be able to provide answers to a wider variety of queries, which is somewhat corroborated by the findings in the following paragraph.'}},\n",
       " {'id': 648,\n",
       "  'distance': 0.6845976114273071,\n",
       "  'entity': {'section': 'methods',\n",
       "   'pmid': 36175836,\n",
       "   'text': 'The ARAX API (arax.rtx.ai/api/arax/v1.2/openapi.json), which itself queries the RTX-KG2 API, can be used to achieve multi-hop RTX-KG2 queries . Further, RTX-KG2c is archived in Biolink Knowledge Graph eXchange TSV format (KGX TSV format, documented at github: biolink/kgx) through the Translator Knowledge Graph Exchange (KGE; see Fig . 1) archive and registry system (github: NCATSTranslator/Knowledge_Graph_Exchange_Registry) (currently in testing phase). Utility and discussion . Uptake and adoption . Due to its comprehensiveness and/or its speed, RTX-KG2 is already being used as a core knowledge provider (see github: NCATSTranslator/Translator-All/wiki/KG2) or knowledge graph by five diverse reasoning agents within the Translator system: ARAX, which our team developed and which provides sophisticated workflow capabilities and overlay of virtual edges for associations based on literature co-occurrence or network structural equivalence; mediKanren, which provides sophisticated network motif-finding and path-finding using the miniKanren logic programming language; BioThings Explorer, the engine for autonomous querying of distributed biomedical knowledge, described in the \"Background\" section; ARAGORN (github: ranking-agent/aragorn), a reasoning agent that has unique capabilities for coalescing and ranking knowledge subgraphs; and the Explanatory Agent (github: NCATSTranslator/Explanatory-Agent), a reasoning agent that uses natural language-understanding models in order to explain and rank results . Use of the RTX-KG2 API appears to be increasing over time, with an average of 1,084 queries per day over the 9 months prior to this writing (September 2021 - June 2022) vs. an average of 1,417 queries per day over the last 3 months (April - June 2022), a 1.3-fold increase . Programs that query RTX-KG2 may optionally identify themselves; of the 51% of queries from the last 9 months in which the submitter was identified, approximately 63% were by the Explanatory Agent, 23% by ARAX, 7% by BioThings Explorer, and 7% by ARAGORN . Rather than using the RTX-KG2 API, the mediKanren reasoning agent uses a bulk download of RTX-KG2pre in conjunction with their own canonicalization algorithm . In addition to its primary intended use-case for on-demand knowledge exploration and concept-specific reasoning, the RTX-KG2 knowledge graph can be used as a structure prior for data-driven network inference, for example, causal network learning . We have recently described a computational method, Kg2Causal, for using a general-purpose biomedical knowledge graph to extract a network structure prior distribution for data-driven causal network inference from multivariate observations . Using the predecessor graph, RTX-KG1, we found that using a general knowledge graph as a prior significantly improved the accuracy of data-driven causal network inference compared to using any of several uninformative network structure priors.'}},\n",
       " {'id': 647,\n",
       "  'distance': 0.6845976114273071,\n",
       "  'entity': {'section': 'methods',\n",
       "   'pmid': 36175836,\n",
       "   'text': 'By number of edges, the largest contributing knowledge source for RTX-KG2.7.3pre is SemMedDB, which has 19.3M edges (comprising about a third of the edges in the graph), followed by PathWhiz (13.7M edges), NCBI Taxonomy (3.6M edges), and DrugBank (2.8M edges). In terms of their total (i.e., inout) vertex degree distributions, both KG2pre and KG2c appear to be approximately scale-free (Fig . 6) with a power law exponent of 2.43, meaning that the frequency of concepts with connectivity k decreases as . Figure 7 highlights the frequencies of various combinations of subject node category and object node category appearing together in edges in KG2c, indicating (1) high levels of cross-category axioms among \"molecular entity\", \"small molecule\", and \"chemical entity\" and (2) high levels of connections between \"pathway\" and \"molecular entity\", \"small molecule\", \"molecular activity\", \"organism taxon\", \"anatomical entity\", and \"transcript\". Note that the category-category frequency heatmap is not expected to be symmetric for a knowledge graph (such as RTX-KG2) with a high proportion of relationship types that have non-reflexive subject-object semantics . RTX-KG2 access channels . The complete software code for building RTX-KG2 and for hosting an indexed RTX-KG2 graph database in Neo4j is publicly available in an open-source repository (see Availability of data and materials). In addition, the content of the latest RTX-KG2 graphs (version KG2.7.3) that we have built can be easily accessed via three different channels (see Availability of data and materials), depending on the use-case: (i) serialized flat-file download (as described below); (ii) REpresentational State Transfer (REST) API (i.e., a web API); or (iii) web browser user interface, through the ARAX system . Tab-separated value (TSV) serializations of RTX-KG2pre and JSON serializations of RTX-KG2c are available in a public GitHub repository via the git-lfs file hosting mechanism, and their schemas are documented as described in the \"RTX-KG2 schema and RTX-KG2pre Biolink compliance\" section and in the RTX-KG2 documentation sections that are linked therein . RTX-KG2c can be queried via a REST API that implements the open-standard Translator Reasoner API (TRAPI) specification (github: NCATSTranslator/ReasonerAPI) and that is registered via the SmartAPI framework and therefore discoverable using SmartAPI-associated tooling such as BioThings Explorer . The RTX-KG2 API enables one-hop querying of the knowledge graph; queries are internally serviced by the PloverDB in-memory graph database (see the \"Building RTX-KG2c, the canonicalized version of RTX-KG2\" section).'}},\n",
       " {'id': 646,\n",
       "  'distance': 0.6845976114273071,\n",
       "  'entity': {'section': 'methods',\n",
       "   'pmid': 36175836,\n",
       "   'text': 'The RTX-KG2 build process incorporates multiple layers of quality control, including both automated and manual procedures (see the \"Building RTX-KG2pre from upstream sources\" section for details). As the first step in the build process, scripts validate the consistency of the RTX-KG2 semantic layer with the Biolink model . During knowledge integration, relationships whose subject or object nodes are not present in the knowledge graph are logged for offline investigation . A report of statistics on the RTX-KG2 knowledge graph is generated: including (i) node counts by knowledge source and by semantic type and (ii) edge counts by source and by relationship type: both before and after redundant edges are joined in the merge process . The procedure for RTX-KG2 builds includes a script-facilitated comparison of that report for the new build with the equivalent report for the previous build, in order to enable the build supervisor to recognize anomalously large (e.g., more than three-fold) changes in node or edge counts conditioned on source, category, or predicate . Once a new RTX-KG2c build is installed into the RTX-KG2 query API, functional correctness of the live API is verified using a Python-based test suite . Reproducibility of the RTX-KG2 build is enhanced by the intentional choice of using an ETL approach based on flat-file exports knowledge-source databases . To aid with versioning, most sources have their version stored in the name attribute in the node in RTX-KG2 that represents the source database . RTX-KG2 content and statistics . The latest released version of RTX-KG2pre as of this writing, RTX-KG2.7.3, contains 10.2 million nodes and 54.0 million edges . Each edge is labeled with one of 77 distinct predicates (Biolink relationship types) and each node with one of 56 distinct categories (Biolink concept semantic types). In terms of frequency distribution, there is over six decades of variation across node categories (Fig . 4) and edge predicates (Fig . 5), with the dominant category being OrganismTaxon (reflecting the significant size of the NCBI organism classification ontology) and the dominant predicate being has_participant (reflecting the significant size of the PathBank database). Fig . 5 shows a breakdown of edges in KG2.7.3 by their Biolink predicate . KG2.7.3c contains 6.4 million nodes and 39.3 million edges, which is approximately 62% of the nodes and 73% of the edges of KG2.7.3pre, reflecting an expected reduction in node count due to canonicalization as well as due to post-canonicalization edge merging.'}},\n",
       " {'id': 645,\n",
       "  'distance': 0.6845976114273071,\n",
       "  'entity': {'section': 'methods',\n",
       "   'pmid': 36175836,\n",
       "   'text': 'In our work on RTX-KG2pre, we found that coalescing nodes for semantically equivalent concepts into single nodes facilitates reasoning by reducing the complexity of graph paths that represent answers for common translational questions . Thus, to enhance the utility of RTX-KG2 for translational reasoning, we created a version of RTX-KG2 called RTX-KG2canonicalized (abbreviated in this work as RTX-KG2c) in which semantically equivalent nodes are coalesced to a single concept node . In brief, building RTX-KG2c from RTX-KG2pre is carried out in five steps: For Steps 2-3, the RTX-KG2 build system uses the ARAX system\\'s Node Synonymizer service (see in \"ARAX Node Synonymizer\" section for details). The RTX-KG2c graph is serialized in JSON format (see the \"RTX-KG2 schema and RTX-KG2pre Biolink compliance\" section), archived in a GitHub large file storage (LFS) repository (see Availability of data and materials), and imported into a custom-built, open-source, in-memory graph database, PloverDB (github: RTXteam/PloverDB). The build process for RTX-KG2c is Python-based and has comparable hardware requirements to the RTX-KG2pre build process (see the \"Requirements\" section). Formally, the approach used in building RTX-KG2c is concept-oriented as opposed to the realist methodology underlying OBO Foundry ontologies . RTX-KG2 schema and RTX-KG2pre Biolink compliance . The RTX-KG2.7.3 knowledge graph follows the Biolink model (version 2.1.0) for its semantic layer and (in RTX-KG2pre) its schema . RTX-KG2 uses Biolink\\'s category hierarchy for its concept (node) types (Fig . 2) and Biolink\\'s predicate hierarchy for its relationship (edge) types (Fig . 3). When mapping terms from their original sources to the Biolink terminology, the RTX-KG2 build system consults the Biolink model\\'s internal mappings in order to detect any inconsistencies between the two . Because relationship terms that are highly specific are often mapped to more generalized terminology, the original source\\'s phrasing is preserved in the relation property2 . In addition to mapping upstream source relations to Biolink predicates, the RTX-KG2 build process coalesces edges that have the same end nodes and the same predicate (it does, however, preserve the provenance information from both of the coalesced edges). The schema of the JSON serialization of RTX-KG2pre is documented in detail in the RTX-KG2 project area github: RTXteam/RTX-KG2 and summarized in the \"Detailed schema for RTX-KG2\" section . In brief, RTX-KG2 is serialized as a JSON object with keys nodes and edges, with the nodes object containing a list of serialized objects for the concept nodes in the graph, and with edges containing a list of serialized objects for the subject-object-relationship triples in the graph . Quality control and reproducibility.'}},\n",
       " {'id': 644,\n",
       "  'distance': 0.6845976114273071,\n",
       "  'entity': {'section': 'methods',\n",
       "   'pmid': 36175836,\n",
       "   'text': 'Each knowledge source\\'s concepts are assigned Biolink concept semantic types: which are called \"categories\" in the Biolink model: and relationships are assigned Biolink relationship types at the time that the source is ingested . All but two of the 23 OWL-format sources are ontologies from the OBO Foundry; the remaining two OWL-format sources are the Experimental Factor Ontology (EFO) and Orphanet Rare Disease Ontology . Direct-to-JSON sources . With the direct-to-JSON method, sources are ingested in seven different file formats (in \"Sources and their file formats\" section). One source, KEGG, is queried via an API rather than using a flat file download, due to the lack of a download option for users that do not have a commercial license . For the 20 direct-to-JSON sources, the RTX-KG2 system has one ETL module for each source, with each module producing a source-specific JSON file in the RTX-KG2 JSON schema (see the \"RTX-KG2 schema and RTX-KG2pre Biolink compliance\" section) (in contrast, for the 50 RDF-based sources, the system has a single ETL module for ingesting all sources together). The RDF-based method merges all of the OWL and TTL sources (class-based), without flattening the ontologies (i.e., preserving rdfs: subClassOf relationships) and generates a single JSON file . The hybrid design of RTX-KG2 balances the benefits of modularity (where it is feasible in the direct-to-JSON method) with the need for a monolithic ingestion module for ontologies where inter-ontology axioms are needed for determining semantic types at the ETL stage . Building RTX-KG2c, the canonicalized version of RTX-KG2 . RTX-KG2pre nodes and edges are loaded from the RTX-KG2pre TSV files; . the set of nodes is partitioned into disjoint subsets of equivalent nodes; . from each group of equivalent nodes, a canonical node identifier is chosen, added to RTX-KG2c, and annotated with the identifiers of its synonymous nodes (along with other information merged from the synonymous nodes); . edges from RTX-KG2pre are remapped to refer only to the canonical node identifiers; and . edges with the same subject, object, and Biolink predicate are merged . Because the various ontologies that RTX-KG2pre ingests often represent the same concept using multiple different identifiers, some of the nodes in RTX-KG2pre represent equivalent concepts . For example, Parkinson\\'s disease is represented by several nodes in RTX-KG2pre, such as the nodes with identifiers MONDO:0005180, DOID:14330, EFO:0002508, and MESH: D010300, many of which are connected in RTX-KG2pre with relationships of type biolink: same_as or non-transitive generalizations of that relationship type.'}},\n",
       " {'id': 643,\n",
       "  'distance': 0.6845976114273071,\n",
       "  'entity': {'section': 'methods',\n",
       "   'pmid': 36175836,\n",
       "   'text': 'Construction and content . In this section, we describe how RTX-KG2 is constructed; provide an overview of its graph database schema; and summarize its content in terms of sources, semantic breadth, and size . The overall build process, along with the various outputs of RTX-KG2, is depicted in Fig . 1 . Broadly speaking, the RTX-KG2 build system does four things: it (i) loads information from source databases (blue triangles in Fig . 1) via the World Wide Web as described in the \"Sources and their file formats\" section; (ii) integrates the knowledge into a precursor knowledge graph called RTX-KG2pre (upper green hexagon in Fig . 1) and hosts it in a Neo4j database (upper orange cloud in Fig . 1) as described in the \"Building RTX-KG2pre from upstream sources\" section; (iii) coalesces equivalent concept nodes into a canonicalized knowledge graph called RTX-KG2c (brown circle in Fig . 1) as described in the \"Building RTX-KG2c, the canonicalized version of RTX-KG2\" section, with a schema that is described in the \"RTX-KG2 schema and RTX-KG2pre Biolink compliance\" section; and (iv) provides various knowledge graph artifacts and services as described in the \"RTX-KG2 content and statistics\" section . We provide technical details of the RTX-KG2 build system in the \"RTX-KG2 build system and software\" section . Sources and their file formats . RTX-KG2 integrates 70 sources (Table 1), 50 of them via a resource description framework (RDF) -based ingestion method and 20 of them via a direct-to-JSON ingestion method . Sources are loaded in a specific order controlled by a configuration file, with precedence applying to the assignment of Biolink categories to nodes . RDF-based sources . Of the 50 RDF-based sources, the system ingests 27 in Terse RDF Triple Language (TTL) format and 23 as OWL ontologies in RDF/XML format (which we abbreviate here as \"OWL\"). Of the 27 TTL sources, 26 are from the UMLS, obtained as described in \"umls2rdf and owltools\" section; the remaining source is a TTL representation of the Biolink model, which defines the semantic layer for RTX-KG2, including hierarchies of concept types and relation types (see the \"RTX-KG2 content and statistics\" section). In addition to concept type and relation type hierarchies, the Biolink model provides equivalence mappings of the Biolink types to classes in other high-level ontologies (such as biolink: Gene being equivalent to SIO:010035) and of the Biolink concept types to prioritized lists of identifier types for the concept type1.'}}]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"id\": 593,\n",
      "    \"distance\": 0.7601258754730225,\n",
      "    \"entity\": {\n",
      "        \"section\": \"title\",\n",
      "        \"pmid\": 35840060,\n",
      "        \"text\": \"KG-Predict: A knowledge graph computational framework for drug repurposing.\"\n",
      "    }\n",
      "}\n",
      "{\n",
      "    \"id\": 426,\n",
      "    \"distance\": 0.724739134311676,\n",
      "    \"entity\": {\n",
      "        \"section\": \"title\",\n",
      "        \"pmid\": 34297570,\n",
      "        \"text\": \"Knowledge Graph-Based Approaches to Drug Repurposing for COVID-19.\"\n",
      "    }\n",
      "}\n",
      "{\n",
      "    \"id\": 534,\n",
      "    \"distance\": 0.6885125041007996,\n",
      "    \"entity\": {\n",
      "        \"section\": \"title\",\n",
      "        \"pmid\": 34181736,\n",
      "        \"text\": \"CROssBAR: comprehensive resource of biomedical relations with knowledge graph representations.\"\n",
      "    }\n",
      "}\n",
      "{\n",
      "    \"id\": 649,\n",
      "    \"distance\": 0.6845976114273071,\n",
      "    \"entity\": {\n",
      "        \"section\": \"methods\",\n",
      "        \"pmid\": 36175836,\n",
      "        \"text\": \"To the extent that it incorporates multiple graph structural variations, RTX-KG2 can also be used as a test-bed for evaluating the performance of structurally generalizable graph analysis methods such as a subset of us have done for the case of a structurally generalizable node-node similarity measure . Another application of RTX-KG2 is as training data for drug repurposing models; previous work by our team utilized RTX-KG2's predecessor, RTX-KG1, to train a random forest model that predicts novel drug treatments for diseases . More recent extensions to this work have utilized the canonicalized version of RTX-KG2 as training data . Comparison to other knowledge graphs/providers . To objectively evaluate the size and semantic richness of RTX-KG2 in comparison to other biomedical knowledge graphs, we compared it to four other knowledge graphs that are in active use for translational applications: Hetionet, SPOKE, the SRI Reference Knowledge Graph, and ROBOKOP . For the counts of meta-triples (i.e., counts of edges with a given pattern of subject category, object category, and predicate, which provide quantitative information about the richness of the knowledge graph for providing relationships of particular types), we accessed the ROBOKOP and SPOKE graphs via their SmartAPI-registered Translator API (\\\"TRAPI\\\") endpoints on March 15, 2022 . For Hetionet and ROBOKOP, we used published node and edge counts . For the SRI Reference Knowledge Graph, we downloaded the graph in KGX TSV format from the Biomedical Data Translator Knowledge Graph Exchange (KGE) and analyzed the graph locally . Node and edge counts for various knowledge graphs are shown in Table 2 . The large number of edges in the ROBOKOP knowledge graph versus RTX-KG2 reflects the latter's practice of joining edges that have identical subject node, object node, and predicate type . The large number of nodes in the SRI Reference Knowledge Graph versus RTX-KG2pre is largely due to more InformationContentEntity nodes (3.7 million vs. 144,396) and SequenceVariant nodes (2.4 million vs. 0) in the former . Notably, RTX-KG2 contains a richer set of meta-triples (distinct combinations of subject node category, edge predicate, and object node category) versus the other knowledge graphs (Table 3); RTX-KG2c contains 4.6-fold more meta-triples than the second-ranked knowledge graph and 18.5-fold more meta-triples than the third-ranked knowledge graph, by meta-triple count . In general we would expect a graph with a greater number of meta-triples to be able to provide answers to a wider variety of queries, which is somewhat corroborated by the findings in the following paragraph.\"\n",
      "    }\n",
      "}\n",
      "{\n",
      "    \"id\": 648,\n",
      "    \"distance\": 0.6845976114273071,\n",
      "    \"entity\": {\n",
      "        \"section\": \"methods\",\n",
      "        \"pmid\": 36175836,\n",
      "        \"text\": \"The ARAX API (arax.rtx.ai/api/arax/v1.2/openapi.json), which itself queries the RTX-KG2 API, can be used to achieve multi-hop RTX-KG2 queries . Further, RTX-KG2c is archived in Biolink Knowledge Graph eXchange TSV format (KGX TSV format, documented at github: biolink/kgx) through the Translator Knowledge Graph Exchange (KGE; see Fig . 1) archive and registry system (github: NCATSTranslator/Knowledge_Graph_Exchange_Registry) (currently in testing phase). Utility and discussion . Uptake and adoption . Due to its comprehensiveness and/or its speed, RTX-KG2 is already being used as a core knowledge provider (see github: NCATSTranslator/Translator-All/wiki/KG2) or knowledge graph by five diverse reasoning agents within the Translator system: ARAX, which our team developed and which provides sophisticated workflow capabilities and overlay of virtual edges for associations based on literature co-occurrence or network structural equivalence; mediKanren, which provides sophisticated network motif-finding and path-finding using the miniKanren logic programming language; BioThings Explorer, the engine for autonomous querying of distributed biomedical knowledge, described in the \\\"Background\\\" section; ARAGORN (github: ranking-agent/aragorn), a reasoning agent that has unique capabilities for coalescing and ranking knowledge subgraphs; and the Explanatory Agent (github: NCATSTranslator/Explanatory-Agent), a reasoning agent that uses natural language-understanding models in order to explain and rank results . Use of the RTX-KG2 API appears to be increasing over time, with an average of 1,084 queries per day over the 9 months prior to this writing (September 2021 - June 2022) vs. an average of 1,417 queries per day over the last 3 months (April - June 2022), a 1.3-fold increase . Programs that query RTX-KG2 may optionally identify themselves; of the 51% of queries from the last 9 months in which the submitter was identified, approximately 63% were by the Explanatory Agent, 23% by ARAX, 7% by BioThings Explorer, and 7% by ARAGORN . Rather than using the RTX-KG2 API, the mediKanren reasoning agent uses a bulk download of RTX-KG2pre in conjunction with their own canonicalization algorithm . In addition to its primary intended use-case for on-demand knowledge exploration and concept-specific reasoning, the RTX-KG2 knowledge graph can be used as a structure prior for data-driven network inference, for example, causal network learning . We have recently described a computational method, Kg2Causal, for using a general-purpose biomedical knowledge graph to extract a network structure prior distribution for data-driven causal network inference from multivariate observations . Using the predecessor graph, RTX-KG1, we found that using a general knowledge graph as a prior significantly improved the accuracy of data-driven causal network inference compared to using any of several uninformative network structure priors.\"\n",
      "    }\n",
      "}\n",
      "{\n",
      "    \"id\": 647,\n",
      "    \"distance\": 0.6845976114273071,\n",
      "    \"entity\": {\n",
      "        \"section\": \"methods\",\n",
      "        \"pmid\": 36175836,\n",
      "        \"text\": \"By number of edges, the largest contributing knowledge source for RTX-KG2.7.3pre is SemMedDB, which has 19.3M edges (comprising about a third of the edges in the graph), followed by PathWhiz (13.7M edges), NCBI Taxonomy (3.6M edges), and DrugBank (2.8M edges). In terms of their total (i.e., inout) vertex degree distributions, both KG2pre and KG2c appear to be approximately scale-free (Fig . 6) with a power law exponent of 2.43, meaning that the frequency of concepts with connectivity k decreases as . Figure 7 highlights the frequencies of various combinations of subject node category and object node category appearing together in edges in KG2c, indicating (1) high levels of cross-category axioms among \\\"molecular entity\\\", \\\"small molecule\\\", and \\\"chemical entity\\\" and (2) high levels of connections between \\\"pathway\\\" and \\\"molecular entity\\\", \\\"small molecule\\\", \\\"molecular activity\\\", \\\"organism taxon\\\", \\\"anatomical entity\\\", and \\\"transcript\\\". Note that the category-category frequency heatmap is not expected to be symmetric for a knowledge graph (such as RTX-KG2) with a high proportion of relationship types that have non-reflexive subject-object semantics . RTX-KG2 access channels . The complete software code for building RTX-KG2 and for hosting an indexed RTX-KG2 graph database in Neo4j is publicly available in an open-source repository (see Availability of data and materials). In addition, the content of the latest RTX-KG2 graphs (version KG2.7.3) that we have built can be easily accessed via three different channels (see Availability of data and materials), depending on the use-case: (i) serialized flat-file download (as described below); (ii) REpresentational State Transfer (REST) API (i.e., a web API); or (iii) web browser user interface, through the ARAX system . Tab-separated value (TSV) serializations of RTX-KG2pre and JSON serializations of RTX-KG2c are available in a public GitHub repository via the git-lfs file hosting mechanism, and their schemas are documented as described in the \\\"RTX-KG2 schema and RTX-KG2pre Biolink compliance\\\" section and in the RTX-KG2 documentation sections that are linked therein . RTX-KG2c can be queried via a REST API that implements the open-standard Translator Reasoner API (TRAPI) specification (github: NCATSTranslator/ReasonerAPI) and that is registered via the SmartAPI framework and therefore discoverable using SmartAPI-associated tooling such as BioThings Explorer . The RTX-KG2 API enables one-hop querying of the knowledge graph; queries are internally serviced by the PloverDB in-memory graph database (see the \\\"Building RTX-KG2c, the canonicalized version of RTX-KG2\\\" section).\"\n",
      "    }\n",
      "}\n",
      "{\n",
      "    \"id\": 646,\n",
      "    \"distance\": 0.6845976114273071,\n",
      "    \"entity\": {\n",
      "        \"section\": \"methods\",\n",
      "        \"pmid\": 36175836,\n",
      "        \"text\": \"The RTX-KG2 build process incorporates multiple layers of quality control, including both automated and manual procedures (see the \\\"Building RTX-KG2pre from upstream sources\\\" section for details). As the first step in the build process, scripts validate the consistency of the RTX-KG2 semantic layer with the Biolink model . During knowledge integration, relationships whose subject or object nodes are not present in the knowledge graph are logged for offline investigation . A report of statistics on the RTX-KG2 knowledge graph is generated: including (i) node counts by knowledge source and by semantic type and (ii) edge counts by source and by relationship type: both before and after redundant edges are joined in the merge process . The procedure for RTX-KG2 builds includes a script-facilitated comparison of that report for the new build with the equivalent report for the previous build, in order to enable the build supervisor to recognize anomalously large (e.g., more than three-fold) changes in node or edge counts conditioned on source, category, or predicate . Once a new RTX-KG2c build is installed into the RTX-KG2 query API, functional correctness of the live API is verified using a Python-based test suite . Reproducibility of the RTX-KG2 build is enhanced by the intentional choice of using an ETL approach based on flat-file exports knowledge-source databases . To aid with versioning, most sources have their version stored in the name attribute in the node in RTX-KG2 that represents the source database . RTX-KG2 content and statistics . The latest released version of RTX-KG2pre as of this writing, RTX-KG2.7.3, contains 10.2 million nodes and 54.0 million edges . Each edge is labeled with one of 77 distinct predicates (Biolink relationship types) and each node with one of 56 distinct categories (Biolink concept semantic types). In terms of frequency distribution, there is over six decades of variation across node categories (Fig . 4) and edge predicates (Fig . 5), with the dominant category being OrganismTaxon (reflecting the significant size of the NCBI organism classification ontology) and the dominant predicate being has_participant (reflecting the significant size of the PathBank database). Fig . 5 shows a breakdown of edges in KG2.7.3 by their Biolink predicate . KG2.7.3c contains 6.4 million nodes and 39.3 million edges, which is approximately 62% of the nodes and 73% of the edges of KG2.7.3pre, reflecting an expected reduction in node count due to canonicalization as well as due to post-canonicalization edge merging.\"\n",
      "    }\n",
      "}\n",
      "{\n",
      "    \"id\": 645,\n",
      "    \"distance\": 0.6845976114273071,\n",
      "    \"entity\": {\n",
      "        \"section\": \"methods\",\n",
      "        \"pmid\": 36175836,\n",
      "        \"text\": \"In our work on RTX-KG2pre, we found that coalescing nodes for semantically equivalent concepts into single nodes facilitates reasoning by reducing the complexity of graph paths that represent answers for common translational questions . Thus, to enhance the utility of RTX-KG2 for translational reasoning, we created a version of RTX-KG2 called RTX-KG2canonicalized (abbreviated in this work as RTX-KG2c) in which semantically equivalent nodes are coalesced to a single concept node . In brief, building RTX-KG2c from RTX-KG2pre is carried out in five steps: For Steps 2-3, the RTX-KG2 build system uses the ARAX system's Node Synonymizer service (see in \\\"ARAX Node Synonymizer\\\" section for details). The RTX-KG2c graph is serialized in JSON format (see the \\\"RTX-KG2 schema and RTX-KG2pre Biolink compliance\\\" section), archived in a GitHub large file storage (LFS) repository (see Availability of data and materials), and imported into a custom-built, open-source, in-memory graph database, PloverDB (github: RTXteam/PloverDB). The build process for RTX-KG2c is Python-based and has comparable hardware requirements to the RTX-KG2pre build process (see the \\\"Requirements\\\" section). Formally, the approach used in building RTX-KG2c is concept-oriented as opposed to the realist methodology underlying OBO Foundry ontologies . RTX-KG2 schema and RTX-KG2pre Biolink compliance . The RTX-KG2.7.3 knowledge graph follows the Biolink model (version 2.1.0) for its semantic layer and (in RTX-KG2pre) its schema . RTX-KG2 uses Biolink's category hierarchy for its concept (node) types (Fig . 2) and Biolink's predicate hierarchy for its relationship (edge) types (Fig . 3). When mapping terms from their original sources to the Biolink terminology, the RTX-KG2 build system consults the Biolink model's internal mappings in order to detect any inconsistencies between the two . Because relationship terms that are highly specific are often mapped to more generalized terminology, the original source's phrasing is preserved in the relation property2 . In addition to mapping upstream source relations to Biolink predicates, the RTX-KG2 build process coalesces edges that have the same end nodes and the same predicate (it does, however, preserve the provenance information from both of the coalesced edges). The schema of the JSON serialization of RTX-KG2pre is documented in detail in the RTX-KG2 project area github: RTXteam/RTX-KG2 and summarized in the \\\"Detailed schema for RTX-KG2\\\" section . In brief, RTX-KG2 is serialized as a JSON object with keys nodes and edges, with the nodes object containing a list of serialized objects for the concept nodes in the graph, and with edges containing a list of serialized objects for the subject-object-relationship triples in the graph . Quality control and reproducibility.\"\n",
      "    }\n",
      "}\n",
      "{\n",
      "    \"id\": 644,\n",
      "    \"distance\": 0.6845976114273071,\n",
      "    \"entity\": {\n",
      "        \"section\": \"methods\",\n",
      "        \"pmid\": 36175836,\n",
      "        \"text\": \"Each knowledge source's concepts are assigned Biolink concept semantic types: which are called \\\"categories\\\" in the Biolink model: and relationships are assigned Biolink relationship types at the time that the source is ingested . All but two of the 23 OWL-format sources are ontologies from the OBO Foundry; the remaining two OWL-format sources are the Experimental Factor Ontology (EFO) and Orphanet Rare Disease Ontology . Direct-to-JSON sources . With the direct-to-JSON method, sources are ingested in seven different file formats (in \\\"Sources and their file formats\\\" section). One source, KEGG, is queried via an API rather than using a flat file download, due to the lack of a download option for users that do not have a commercial license . For the 20 direct-to-JSON sources, the RTX-KG2 system has one ETL module for each source, with each module producing a source-specific JSON file in the RTX-KG2 JSON schema (see the \\\"RTX-KG2 schema and RTX-KG2pre Biolink compliance\\\" section) (in contrast, for the 50 RDF-based sources, the system has a single ETL module for ingesting all sources together). The RDF-based method merges all of the OWL and TTL sources (class-based), without flattening the ontologies (i.e., preserving rdfs: subClassOf relationships) and generates a single JSON file . The hybrid design of RTX-KG2 balances the benefits of modularity (where it is feasible in the direct-to-JSON method) with the need for a monolithic ingestion module for ontologies where inter-ontology axioms are needed for determining semantic types at the ETL stage . Building RTX-KG2c, the canonicalized version of RTX-KG2 . RTX-KG2pre nodes and edges are loaded from the RTX-KG2pre TSV files; . the set of nodes is partitioned into disjoint subsets of equivalent nodes; . from each group of equivalent nodes, a canonical node identifier is chosen, added to RTX-KG2c, and annotated with the identifiers of its synonymous nodes (along with other information merged from the synonymous nodes); . edges from RTX-KG2pre are remapped to refer only to the canonical node identifiers; and . edges with the same subject, object, and Biolink predicate are merged . Because the various ontologies that RTX-KG2pre ingests often represent the same concept using multiple different identifiers, some of the nodes in RTX-KG2pre represent equivalent concepts . For example, Parkinson's disease is represented by several nodes in RTX-KG2pre, such as the nodes with identifiers MONDO:0005180, DOID:14330, EFO:0002508, and MESH: D010300, many of which are connected in RTX-KG2pre with relationships of type biolink: same_as or non-transitive generalizations of that relationship type.\"\n",
      "    }\n",
      "}\n",
      "{\n",
      "    \"id\": 643,\n",
      "    \"distance\": 0.6845976114273071,\n",
      "    \"entity\": {\n",
      "        \"section\": \"methods\",\n",
      "        \"pmid\": 36175836,\n",
      "        \"text\": \"Construction and content . In this section, we describe how RTX-KG2 is constructed; provide an overview of its graph database schema; and summarize its content in terms of sources, semantic breadth, and size . The overall build process, along with the various outputs of RTX-KG2, is depicted in Fig . 1 . Broadly speaking, the RTX-KG2 build system does four things: it (i) loads information from source databases (blue triangles in Fig . 1) via the World Wide Web as described in the \\\"Sources and their file formats\\\" section; (ii) integrates the knowledge into a precursor knowledge graph called RTX-KG2pre (upper green hexagon in Fig . 1) and hosts it in a Neo4j database (upper orange cloud in Fig . 1) as described in the \\\"Building RTX-KG2pre from upstream sources\\\" section; (iii) coalesces equivalent concept nodes into a canonicalized knowledge graph called RTX-KG2c (brown circle in Fig . 1) as described in the \\\"Building RTX-KG2c, the canonicalized version of RTX-KG2\\\" section, with a schema that is described in the \\\"RTX-KG2 schema and RTX-KG2pre Biolink compliance\\\" section; and (iv) provides various knowledge graph artifacts and services as described in the \\\"RTX-KG2 content and statistics\\\" section . We provide technical details of the RTX-KG2 build system in the \\\"RTX-KG2 build system and software\\\" section . Sources and their file formats . RTX-KG2 integrates 70 sources (Table 1), 50 of them via a resource description framework (RDF) -based ingestion method and 20 of them via a direct-to-JSON ingestion method . Sources are loaded in a specific order controlled by a configuration file, with precedence applying to the assignment of Biolink categories to nodes . RDF-based sources . Of the 50 RDF-based sources, the system ingests 27 in Terse RDF Triple Language (TTL) format and 23 as OWL ontologies in RDF/XML format (which we abbreviate here as \\\"OWL\\\"). Of the 27 TTL sources, 26 are from the UMLS, obtained as described in \\\"umls2rdf and owltools\\\" section; the remaining source is a TTL representation of the Biolink model, which defines the semantic layer for RTX-KG2, including hierarchies of concept types and relation types (see the \\\"RTX-KG2 content and statistics\\\" section). In addition to concept type and relation type hierarchies, the Biolink model provides equivalence mappings of the Biolink types to classes in other high-level ontologies (such as biolink: Gene being equivalent to SIO:010035) and of the Biolink concept types to prioritized lists of identifier types for the concept type1.\"\n",
      "    }\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "for x in (result[0]):\n",
    "    print(json.dumps(x, indent=4))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pubmed-rag-iB5Z1S4t-py3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
